// @taiga-ui/core/esm2015/components/loader/loader.directive.js
import { __decorate, __param } from "tslib";
import { Directive, ElementRef, Inject, NgZone } from '@angular/core';
import { ANIMATION_FRAME, USER_AGENT } from '@ng-web-apis/common';
import { isEdgeOlderThan, isIE, TuiDestroyService, tuiZonefree } from '@taiga-ui/cdk';
import { Observable } from 'rxjs';
import { pairwise, takeUntil } from 'rxjs/operators';
const PERIMETER = 314;
const COEFFICIENT = 1.5;
let TuiLoaderDirective = class TuiLoaderDirective {
    constructor(animationFrame$, ngZone, destroy$, elementRef, userAgent) {
        this.elementRef = elementRef;
        this.strokeDasharray = PERIMETER;
        this.strokeDashoffset = 0;
        if (!isEdgeOlderThan(17, userAgent) && !isIE(userAgent)) {
            return;
        }
        animationFrame$
            .pipe(tuiZonefree(ngZone), pairwise(), takeUntil(destroy$))
            .subscribe(([cur, prev]) => {
            this.animate(prev - cur);
        });
    }
    animate(delta) {
        if (this.strokeDasharray < 0) {
            this.strokeDasharray = PERIMETER * 2;
            this.strokeDashoffset = 0;
        }
        const strokeDasharray = Math.abs(this.strokeDasharray - PERIMETER);
        const fps = 1000 / delta;
        const offsetStep = 1 + Math.floor(this.strokeDasharray / PERIMETER);
        const { style } = this.elementRef.nativeElement;
        style.strokeDashoffset = this.strokeDashoffset.toString();
        style.strokeDasharray = PERIMETER - strokeDasharray + ' ' + strokeDasharray;
        this.strokeDasharray -= PERIMETER / COEFFICIENT / fps;
        this.strokeDashoffset += (offsetStep * PERIMETER) / COEFFICIENT / fps;
        // this.elementRef.nativeElement.parentElement!.style.animation = 'none';
    }
};
TuiLoaderDirective.ctorParameters = () => [
    { type: Observable, decorators: [{ type: Inject, args: [ANIMATION_FRAME,] }] },
    { type: NgZone, decorators: [{ type: Inject, args: [NgZone,] }] },
    { type: Observable, decorators: [{ type: Inject, args: [TuiDestroyService,] }] },
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
    { type: String, decorators: [{ type: Inject, args: [USER_AGENT,] }] }
];
TuiLoaderDirective = __decorate([
    Directive({
        selector: 'circle',
        providers: [TuiDestroyService],
    }),
    __param(0, Inject(ANIMATION_FRAME)),
    __param(1, Inject(NgZone)),
    __param(2, Inject(TuiDestroyService)),
    __param(3, Inject(ElementRef)),
    __param(4, Inject(USER_AGENT))
], TuiLoaderDirective);
export { TuiLoaderDirective };
