// @taiga-ui/kit/fesm2015/taiga-ui-kit-components-radio-group.js
import { __decorate } from 'tslib';
import { Input, Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';

var TuiRadioGroupComponent_1;
let TuiRadioGroupComponent = TuiRadioGroupComponent_1 = class TuiRadioGroupComponent {
    constructor() {
        this.name = `tui-radio-group-${TuiRadioGroupComponent_1.index++}`;
    }
};
TuiRadioGroupComponent.index = 0;
__decorate([
    Input(),
    tuiDefaultProp()
], TuiRadioGroupComponent.prototype, "name", void 0);
TuiRadioGroupComponent = TuiRadioGroupComponent_1 = __decorate([
    Component({
        selector: 'tui-radio-group',
        template: "<ng-content></ng-content>\n",
        changeDetection: ChangeDetectionStrategy.OnPush
    })
], TuiRadioGroupComponent);

let TuiRadioGroupModule = class TuiRadioGroupModule {
};
TuiRadioGroupModule = __decorate([
    NgModule({
        declarations: [TuiRadioGroupComponent],
        exports: [TuiRadioGroupComponent],
    })
], TuiRadioGroupModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiRadioGroupComponent, TuiRadioGroupModule };
