// @taiga-ui/kit/fesm2015/taiga-ui-kit-components-island.js
import { __decorate } from 'tslib';
import { Input, HostBinding, Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import { CommonModule } from '@angular/common';

let TuiIslandComponent = class TuiIslandComponent {
    constructor() {
        this.size = 'm';
        this.textAlign = "left" /* Left */;
        this.hoverable = false;
        this.transparent = false;
    }
    get sizeS() {
        return this.size === 's';
    }
    get sizeM() {
        return this.size === 'm';
    }
    get sizeL() {
        return this.size === 'l';
    }
    get textAlignLeft() {
        return this.textAlign === 'left';
    }
    get textAlignCenter() {
        return this.textAlign === 'center';
    }
    get textAlignRight() {
        return this.textAlign === 'right';
    }
};
__decorate([
    Input(),
    tuiDefaultProp()
], TuiIslandComponent.prototype, "size", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], TuiIslandComponent.prototype, "textAlign", void 0);
__decorate([
    Input(),
    HostBinding('class.tui-island_hoverable'),
    tuiDefaultProp()
], TuiIslandComponent.prototype, "hoverable", void 0);
__decorate([
    Input(),
    HostBinding('class.tui-island_transparent'),
    tuiDefaultProp()
], TuiIslandComponent.prototype, "transparent", void 0);
__decorate([
    HostBinding('class.tui-island_size_s')
], TuiIslandComponent.prototype, "sizeS", null);
__decorate([
    HostBinding('class.tui-island_size_m')
], TuiIslandComponent.prototype, "sizeM", null);
__decorate([
    HostBinding('class.tui-island_size_l')
], TuiIslandComponent.prototype, "sizeL", null);
__decorate([
    HostBinding('class.tui-island_text-align_left')
], TuiIslandComponent.prototype, "textAlignLeft", null);
__decorate([
    HostBinding('class.tui-island_text-align_center')
], TuiIslandComponent.prototype, "textAlignCenter", null);
__decorate([
    HostBinding('class.tui-island_text-align_right')
], TuiIslandComponent.prototype, "textAlignRight", null);
TuiIslandComponent = __decorate([
    Component({
        selector: 'tui-island, a[tuiIsland]',
        changeDetection: ChangeDetectionStrategy.OnPush,
        template: "<ng-content></ng-content>\n",
        host: {
            class: 'tui-island',
        },
        styles: [":host{display:block}"]
    })
], TuiIslandComponent);

let TuiIslandModule = class TuiIslandModule {
};
TuiIslandModule = __decorate([
    NgModule({
        imports: [CommonModule],
        declarations: [TuiIslandComponent],
        exports: [TuiIslandComponent],
    })
], TuiIslandModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiIslandComponent, TuiIslandModule };
