// @taiga-ui/kit/fesm2015/taiga-ui-kit-providers.js
import { Optional } from '@angular/core';
import { TuiDropdownControllerDirective, TUI_DROPDOWN_CONTROLLER } from '@taiga-ui/core';

function fixedDropdownControllerFactory(directive) {
    directive = directive || new TuiDropdownControllerDirective();
    directive.limitWidth = "fixed" /* Fixed */;
    return directive;
}
const FIXED_DROPDOWN_CONTROLLER_PROVIDER = [
    {
        provide: TUI_DROPDOWN_CONTROLLER,
        deps: [[new Optional(), TuiDropdownControllerDirective]],
        useFactory: fixedDropdownControllerFactory,
    },
];
function leftAlignedDropdownControllerFactory(directive) {
    directive = directive || new TuiDropdownControllerDirective();
    directive.align = 'left';
    return directive;
}
const LEFT_ALIGNED_DROPDOWN_CONTROLLER_PROVIDER = [
    {
        provide: TUI_DROPDOWN_CONTROLLER,
        deps: [[new Optional(), TuiDropdownControllerDirective]],
        useFactory: leftAlignedDropdownControllerFactory,
    },
];

/**
 * Generated bundle index. Do not edit.
 */

export { FIXED_DROPDOWN_CONTROLLER_PROVIDER, LEFT_ALIGNED_DROPDOWN_CONTROLLER_PROVIDER, fixedDropdownControllerFactory, leftAlignedDropdownControllerFactory };
