// @taiga-ui/core/esm2015/directives/hint-controller/hint-controller.directive.js
var TuiHintControllerDirective_1;
import { __decorate } from "tslib";
import { Directive, forwardRef, Input } from '@angular/core';
import { TuiController, tuiDefaultProp } from '@taiga-ui/cdk';
import { TUI_HINT_CONTROLLER } from './hint-controller.token';
// TODO: v2.0 use in Charts
let TuiHintControllerDirective = TuiHintControllerDirective_1 = class TuiHintControllerDirective extends TuiController {
    constructor() {
        super(...arguments);
        this.content = null;
        this.direction = 'bottom-left';
        this.mode = null;
    }
};
__decorate([
    Input('tuiHintContent'),
    tuiDefaultProp()
], TuiHintControllerDirective.prototype, "content", void 0);
__decorate([
    Input('tuiHintDirection'),
    tuiDefaultProp()
], TuiHintControllerDirective.prototype, "direction", void 0);
__decorate([
    Input('tuiHintMode'),
    tuiDefaultProp()
], TuiHintControllerDirective.prototype, "mode", void 0);
TuiHintControllerDirective = TuiHintControllerDirective_1 = __decorate([
    Directive({
        selector: '[tuiHintContent]',
        providers: [
            {
                provide: TUI_HINT_CONTROLLER,
                useExisting: forwardRef(() => TuiHintControllerDirective_1),
            },
        ],
    })
], TuiHintControllerDirective);
export { TuiHintControllerDirective };
