// @taiga-ui/core/esm2015/directives/described-by/described-by.directive.js
import { __decorate, __param } from "tslib";
import { Directive, HostBinding, Inject, Input } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import { TuiHintService } from '@taiga-ui/core/services';
import { Observable } from 'rxjs';
import { TUI_DESCRIBED_BY_PROVIDERS, TUI_DESCRIBED_BY_SHOW, } from './described-by.providers';
export const DESCRIBED_BY = '_described-by';
/**
 * A directive linking focusable elements and hints for accessibility
 */
let TuiDescribedByDirective = class TuiDescribedByDirective {
    constructor(hintService, visibility$) {
        this.tuiDescribedBy = '';
        visibility$.subscribe(visible => {
            if (!this.tuiDescribedBy) {
                return;
            }
            if (visible) {
                hintService.showHintForId(this.tuiDescribedBy);
            }
            else {
                hintService.hideHintForId(this.tuiDescribedBy);
            }
        });
    }
    get computedDescribedBy() {
        return this.tuiDescribedBy ? this.tuiDescribedBy + DESCRIBED_BY : null;
    }
};
TuiDescribedByDirective.ctorParameters = () => [
    { type: TuiHintService, decorators: [{ type: Inject, args: [TuiHintService,] }] },
    { type: Observable, decorators: [{ type: Inject, args: [TUI_DESCRIBED_BY_SHOW,] }] }
];
__decorate([
    Input(),
    tuiDefaultProp()
], TuiDescribedByDirective.prototype, "tuiDescribedBy", void 0);
__decorate([
    HostBinding('attr.aria-describedby')
], TuiDescribedByDirective.prototype, "computedDescribedBy", null);
TuiDescribedByDirective = __decorate([
    Directive({
        selector: '[tuiDescribedBy]:not(ng-container)',
        providers: TUI_DESCRIBED_BY_PROVIDERS,
    }),
    __param(0, Inject(TuiHintService)),
    __param(1, Inject(TUI_DESCRIBED_BY_SHOW))
], TuiDescribedByDirective);
export { TuiDescribedByDirective };
