// @taiga-ui/core/esm2015/directives/described-by/described-by.providers.js
import { ElementRef, InjectionToken } from '@angular/core';
import { TuiDestroyService, TuiFocusVisibleService, typedFromEvent } from '@taiga-ui/cdk';
import { merge, timer } from 'rxjs';
import { distinctUntilChanged, filter, mapTo, startWith, switchMapTo, take, takeUntil, tap, } from 'rxjs/operators';
const DELAY = 1000;
export const TUI_DESCRIBED_BY_SHOW = new InjectionToken('Accessible tooltip visibility stream');
export const TUI_DESCRIBED_BY_PROVIDERS = [
    TuiDestroyService,
    TuiFocusVisibleService,
    {
        provide: TUI_DESCRIBED_BY_SHOW,
        deps: [TuiFocusVisibleService, ElementRef],
        useFactory: describedByFactory,
    },
];
export function describedByFactory(focusVisible$, { nativeElement }) {
    return merge(focusVisible$.pipe(filter(v => v), switchMapTo(timer(DELAY).pipe(mapTo(true), takeUntil(typedFromEvent(nativeElement, 'keydown')))), switchMapTo(typedFromEvent(nativeElement, 'keydown').pipe(filter(({ key }) => key === 'Escape'), take(1), tap(event => {
        event.stopPropagation();
    }), mapTo(false), startWith(true)))), typedFromEvent(nativeElement, 'blur').pipe(mapTo(false))).pipe(distinctUntilChanged());
}
