// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-obscured.js
import { __decorate, __param } from 'tslib';
import { Optional, Inject, Self, Input, Output, Directive, NgModule } from '@angular/core';
import { tuiRequiredSetter } from '@taiga-ui/cdk/decorators';
import { TuiActiveZoneDirective } from '@taiga-ui/cdk/directives/active-zone';
import { TuiObscuredService, TuiParentsScrollService, TuiDestroyService } from '@taiga-ui/cdk/services';
import { Subject, EMPTY } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';

/**
 * Directive that monitors element visibility
 */
let TuiObscuredDirective = class TuiObscuredDirective {
    constructor(activeZone, obscured$) {
        this.enabled$ = new Subject();
        const mapped$ = obscured$.pipe(map(obscuredBy => !!obscuredBy &&
            (!activeZone ||
                !obscuredBy.length ||
                obscuredBy.every(element => !activeZone.contains(element)))));
        this.tuiObscured = this.enabled$.pipe(switchMap(enabled => (enabled ? mapped$ : EMPTY)));
    }
    set tuiObscuredEnabled(enabled) {
        this.enabled$.next(enabled);
    }
};
TuiObscuredDirective.ctorParameters = () => [
    { type: TuiActiveZoneDirective, decorators: [{ type: Optional }, { type: Inject, args: [TuiActiveZoneDirective,] }] },
    { type: TuiObscuredService, decorators: [{ type: Self }, { type: Inject, args: [TuiObscuredService,] }] }
];
__decorate([
    Input(),
    tuiRequiredSetter()
], TuiObscuredDirective.prototype, "tuiObscuredEnabled", null);
__decorate([
    Output()
], TuiObscuredDirective.prototype, "tuiObscured", void 0);
TuiObscuredDirective = __decorate([
    Directive({
        selector: '[tuiObscured]',
        providers: [TuiObscuredService, TuiParentsScrollService, TuiDestroyService],
    }),
    __param(0, Optional()),
    __param(0, Inject(TuiActiveZoneDirective)),
    __param(1, Self()),
    __param(1, Inject(TuiObscuredService))
], TuiObscuredDirective);

let TuiObscuredModule = class TuiObscuredModule {
};
TuiObscuredModule = __decorate([
    NgModule({
        declarations: [TuiObscuredDirective],
        exports: [TuiObscuredDirective],
    })
], TuiObscuredModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiObscuredDirective, TuiObscuredModule };
