// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-media.js
import { __decorate, __param } from 'tslib';
import { EventEmitter, ElementRef, Inject, Input, HostBinding, Output, HostListener, Directive, NgModule } from '@angular/core';
import { tuiDefaultProp, tuiRequiredSetter } from '@taiga-ui/cdk/decorators';

function nonNegativeFiniteAssertion(value) {
    return isFinite(value) && value >= 0;
}
function volumeAssertion(volume) {
    return isFinite(volume) && volume >= 0 && volume <= 1;
}
let TuiMediaDirective = class TuiMediaDirective {
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.volume = 1;
        this.currentTimeChange = new EventEmitter();
        this.pausedChange = new EventEmitter();
        this.volumeChange = new EventEmitter();
        this.playbackRate = 1;
    }
    set playbackRateSetter(playbackRate) {
        this.updatePlaybackRate(playbackRate);
    }
    set currentTime(currentTime) {
        if (Math.abs(currentTime - this.currentTime) > 0.001) {
            this.elementRef.nativeElement.currentTime = currentTime;
        }
    }
    set paused(paused) {
        if (paused) {
            this.elementRef.nativeElement.pause();
        }
        else {
            this.elementRef.nativeElement.play();
            this.updatePlaybackRate(this.playbackRate);
        }
    }
    get paused() {
        return this.elementRef.nativeElement.paused;
    }
    get currentTime() {
        return this.elementRef.nativeElement.currentTime;
    }
    // @bad TODO: Make sure no other events can affect this like network issues etc.
    onPausedChange(paused) {
        this.pausedChange.emit(paused);
        this.updatePlaybackRate(this.playbackRate);
    }
    onVolumeChange() {
        this.volume = this.elementRef.nativeElement.volume;
        this.volumeChange.emit(this.volume);
    }
    onCurrentTimeChange() {
        this.currentTimeChange.emit(this.currentTime);
    }
    changeDetectionTrigger() {
        // @bad TODO: consider if other events need to trigger CD
    }
    updatePlaybackRate(playbackRate) {
        this.playbackRate = playbackRate;
        this.elementRef.nativeElement.playbackRate = this.playbackRate;
    }
};
TuiMediaDirective.ctorParameters = () => [
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] }
];
__decorate([
    Input(),
    HostBinding('volume'),
    tuiDefaultProp(volumeAssertion)
], TuiMediaDirective.prototype, "volume", void 0);
__decorate([
    Input('playbackRate'),
    tuiRequiredSetter(nonNegativeFiniteAssertion)
], TuiMediaDirective.prototype, "playbackRateSetter", null);
__decorate([
    Input(),
    tuiRequiredSetter(nonNegativeFiniteAssertion)
], TuiMediaDirective.prototype, "currentTime", null);
__decorate([
    Input()
], TuiMediaDirective.prototype, "paused", null);
__decorate([
    Output()
], TuiMediaDirective.prototype, "currentTimeChange", void 0);
__decorate([
    Output()
], TuiMediaDirective.prototype, "pausedChange", void 0);
__decorate([
    Output()
], TuiMediaDirective.prototype, "volumeChange", void 0);
__decorate([
    HostListener('ended', ['true']),
    HostListener('pause', ['true']),
    HostListener('play', ['false'])
], TuiMediaDirective.prototype, "onPausedChange", null);
__decorate([
    HostListener('volumechange')
], TuiMediaDirective.prototype, "onVolumeChange", null);
__decorate([
    HostListener('timeupdate'),
    HostListener('seeking'),
    HostListener('seeked')
], TuiMediaDirective.prototype, "onCurrentTimeChange", null);
__decorate([
    HostListener('durationchange')
], TuiMediaDirective.prototype, "changeDetectionTrigger", null);
TuiMediaDirective = __decorate([
    Directive({
        selector: 'video[tuiMedia], audio[tuiMedia]',
        exportAs: 'tuiMedia',
    }),
    __param(0, Inject(ElementRef))
], TuiMediaDirective);

let TuiMediaModule = class TuiMediaModule {
};
TuiMediaModule = __decorate([
    NgModule({
        declarations: [TuiMediaDirective],
        exports: [TuiMediaDirective],
    })
], TuiMediaModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiMediaDirective, TuiMediaModule, nonNegativeFiniteAssertion, volumeAssertion };
