// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-components-dialog-host.js
import { __decorate, __param } from 'tslib';
import { Inject, Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { TUI_PARENT_ANIMATION } from '@taiga-ui/cdk/constants';
import { TUI_DIALOGS } from '@taiga-ui/cdk/tokens';
import { merge } from 'rxjs';
import { CommonModule } from '@angular/common';
import { TuiFocusTrapModule } from '@taiga-ui/cdk/directives/focus-trap';
import { TuiOverscrollModule } from '@taiga-ui/cdk/directives/overscroll';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';

// @dynamic
let TuiDialogHostComponent = class TuiDialogHostComponent {
    constructor(dialogs) {
        this.dialogs = dialogs;
        this.dialogs$ = merge(...this.dialogs);
    }
};
TuiDialogHostComponent.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [TUI_DIALOGS,] }] }
];
TuiDialogHostComponent = __decorate([
    Component({
        selector: 'tui-dialog-host',
        template: "<section\n    *ngFor=\"let item of dialogs$ | async\"\n    polymorpheus-outlet\n    tuiFocusTrap\n    tuiOverscroll=\"all\"\n    class=\"dialog\"\n    role=\"dialog\"\n    aria-modal=\"true\"\n    [attr.aria-labelledby]=\"item.id\"\n    [content]=\"item.component\"\n    [context]=\"item\"\n    [@tuiParentAnimation]\n></section>\n<div class=\"overlay\"></div>\n",
        changeDetection: ChangeDetectionStrategy.OnPush,
        animations: [TUI_PARENT_ANIMATION],
        styles: [":host{position:fixed;left:0;bottom:0;width:100%;height:0}.dialog,.overlay{position:fixed;top:0;left:0;right:0;bottom:0;display:flex;align-items:flex-start;outline:0;overflow:auto;overflow:overlay}.overlay{pointer-events:none;opacity:0;transition:opacity .3s;background:rgba(0,0,0,.5)}.dialog+.overlay{opacity:1}.dialog:last-of-type{z-index:1}"]
    }),
    __param(0, Inject(TUI_DIALOGS))
], TuiDialogHostComponent);

let TuiDialogHostModule = class TuiDialogHostModule {
};
TuiDialogHostModule = __decorate([
    NgModule({
        imports: [CommonModule, PolymorpheusModule, TuiOverscrollModule, TuiFocusTrapModule],
        declarations: [TuiDialogHostComponent],
        exports: [TuiDialogHostComponent],
    })
], TuiDialogHostModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiDialogHostComponent, TuiDialogHostModule };
