// @taiga-ui/cdk/fesm2015/taiga-ui-cdk-directives-resize.js
import { __decorate, __param } from 'tslib';
import { Inject, Directive, NgModule } from '@angular/core';
import { TuiResizeService, TuiDestroyService } from '@taiga-ui/cdk/services';
import { Observable } from 'rxjs';

// @dynamic
let TuiResizeDirective = class TuiResizeDirective {
    constructor(tuiResize) {
        this.tuiResize = tuiResize;
    }
};
TuiResizeDirective.ctorParameters = () => [
    { type: Observable, decorators: [{ type: Inject, args: [TuiResizeService,] }] }
];
TuiResizeDirective = __decorate([
    Directive({
        selector: '[tuiResize]',
        outputs: ['tuiResize'],
        providers: [TuiDestroyService, TuiResizeService],
    }),
    __param(0, Inject(TuiResizeService))
], TuiResizeDirective);

/**
 * @experimental
 */
let TuiResizeModule = class TuiResizeModule {
};
TuiResizeModule = __decorate([
    NgModule({
        declarations: [TuiResizeDirective],
        exports: [TuiResizeDirective],
    })
], TuiResizeModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiResizeDirective, TuiResizeModule };
