// @taiga-ui/cdk/esm2015/directives/checked/checked.directive.js
import { __decorate, __param } from "tslib";
import { Directive, ElementRef, EventEmitter, HostListener, Inject, Input, Output, Renderer2, } from '@angular/core';
let TuiCheckedDirective = class TuiCheckedDirective {
    constructor(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.tuiCheckedChange = new EventEmitter();
        this.updateProperty('checked', false);
    }
    set tuiChecked(checked) {
        this.updateProperty('checked', checked || false);
        this.updateProperty('indeterminate', checked === null);
    }
    onChange({ checked }) {
        this.updateProperty('indeterminate', false);
        this.tuiCheckedChange.emit(checked);
    }
    updateProperty(property, value) {
        this.renderer.setProperty(this.element.nativeElement, property, value);
    }
};
TuiCheckedDirective.ctorParameters = () => [
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
    { type: Renderer2, decorators: [{ type: Inject, args: [Renderer2,] }] }
];
__decorate([
    Input()
], TuiCheckedDirective.prototype, "tuiChecked", null);
__decorate([
    Output()
], TuiCheckedDirective.prototype, "tuiCheckedChange", void 0);
__decorate([
    HostListener('change', ['$event.target'])
], TuiCheckedDirective.prototype, "onChange", null);
TuiCheckedDirective = __decorate([
    Directive({
        selector: 'input[tuiChecked], input[tuiCheckedChange]',
    }),
    __param(0, Inject(ElementRef)),
    __param(1, Inject(Renderer2))
], TuiCheckedDirective);
export { TuiCheckedDirective };
