// @taiga-ui/core/esm5/components/notification/notification.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, HostBinding, Inject, Input, Output, } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import { TUI_CLOSE_WORD } from '@taiga-ui/core/tokens';
import { Observable } from 'rxjs';
export var STATUS_ICON = {
    info: 'tuiIconInfo',
    success: 'tuiIconCheckCircle',
    error: 'tuiIconCancel',
    warning: 'tuiIconAttention',
};
// @bad TODO: Think about moving to kit
var TuiNotificationComponent = /** @class */ (function () {
    function TuiNotificationComponent(closeWord$) {
        this.closeWord$ = closeWord$;
        this.hasIcon = true;
        this.status = "info" /* Info */;
        this.close = new EventEmitter();
    }
    Object.defineProperty(TuiNotificationComponent.prototype, "icon", {
        get: function () {
            return STATUS_ICON[this.status];
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiNotificationComponent.prototype, "hasClose", {
        get: function () {
            return !!this.close.observers.length;
        },
        enumerable: true,
        configurable: true
    });
    TuiNotificationComponent.ctorParameters = function () { return [
        { type: Observable, decorators: [{ type: Inject, args: [TUI_CLOSE_WORD,] }] }
    ]; };
    __decorate([
        Input(),
        HostBinding('class._has-icon'),
        tuiDefaultProp()
    ], TuiNotificationComponent.prototype, "hasIcon", void 0);
    __decorate([
        Input(),
        HostBinding('attr.data-tui-host-status'),
        tuiDefaultProp()
    ], TuiNotificationComponent.prototype, "status", void 0);
    __decorate([
        Output()
    ], TuiNotificationComponent.prototype, "close", void 0);
    __decorate([
        HostBinding('class._has-close-button')
    ], TuiNotificationComponent.prototype, "hasClose", null);
    TuiNotificationComponent = __decorate([
        Component({
            selector: 'tui-notification',
            template: "<tui-svg\n    *ngIf=\"hasIcon\"\n    automation-id=\"tui-notification__icon\"\n    class=\"icon\"\n    [src]=\"icon\"\n></tui-svg>\n<div class=\"content\">\n    <ng-content></ng-content>\n</div>\n<button\n    *ngIf=\"hasClose\"\n    tuiIconButton\n    type=\"button\"\n    automation-id=\"tui-notification__close\"\n    size=\"xs\"\n    class=\"close\"\n    icon=\"tuiIconCloseLarge\"\n    appearance=\"icon\"\n    [title]=\"closeWord$ | async\"\n    (click)=\"close.emit()\"\n></button>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [":host{font:var(--tui-font-text-s);color:var(--tui-text-01);position:relative;display:block;padding:13px 16px 11px;border-radius:var(--tui-radius-m);background:#fff;background:var(--tui-base-01);overflow:hidden}:host:after{position:absolute;top:0;left:0;width:100%;height:100%;content:'';pointer-events:none}:host._has-close-button{padding-right:32px}:host._has-icon{padding-left:44px}:host[data-tui-host-status=info]{color:var(--tui-info-fill)}:host[data-tui-host-status=info]:after{background:var(--tui-info-bg)}:host[data-tui-host-status=success]{color:var(--tui-success-fill)}:host[data-tui-host-status=success]:after{background:var(--tui-success-bg)}:host[data-tui-host-status=error]{color:var(--tui-error-fill)}:host[data-tui-host-status=error]:after{background:var(--tui-error-bg)}:host[data-tui-host-status=warning]{color:var(--tui-warning-fill)}:host[data-tui-host-status=warning]:after{background:var(--tui-warning-bg)}.content{word-wrap:break-word;color:var(--tui-text-01)}.icon{position:absolute;top:0;left:16px;display:flex;width:16px;height:100%;max-height:48px;align-items:center}.close{position:absolute;top:10px;right:8px}"]
        }),
        __param(0, Inject(TUI_CLOSE_WORD))
    ], TuiNotificationComponent);
    return TuiNotificationComponent;
}());
export { TuiNotificationComponent };
