// @taiga-ui/kit/esm2015/components/island/island.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, HostBinding, Input } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
let TuiIslandComponent = class TuiIslandComponent {
    constructor() {
        this.size = 'm';
        this.textAlign = "left" /* Left */;
        this.hoverable = false;
        this.transparent = false;
    }
    get sizeS() {
        return this.size === 's';
    }
    get sizeM() {
        return this.size === 'm';
    }
    get sizeL() {
        return this.size === 'l';
    }
    get textAlignLeft() {
        return this.textAlign === 'left';
    }
    get textAlignCenter() {
        return this.textAlign === 'center';
    }
    get textAlignRight() {
        return this.textAlign === 'right';
    }
};
__decorate([
    Input(),
    tuiDefaultProp()
], TuiIslandComponent.prototype, "size", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], TuiIslandComponent.prototype, "textAlign", void 0);
__decorate([
    Input(),
    HostBinding('class.tui-island_hoverable'),
    tuiDefaultProp()
], TuiIslandComponent.prototype, "hoverable", void 0);
__decorate([
    Input(),
    HostBinding('class.tui-island_transparent'),
    tuiDefaultProp()
], TuiIslandComponent.prototype, "transparent", void 0);
__decorate([
    HostBinding('class.tui-island_size_s')
], TuiIslandComponent.prototype, "sizeS", null);
__decorate([
    HostBinding('class.tui-island_size_m')
], TuiIslandComponent.prototype, "sizeM", null);
__decorate([
    HostBinding('class.tui-island_size_l')
], TuiIslandComponent.prototype, "sizeL", null);
__decorate([
    HostBinding('class.tui-island_text-align_left')
], TuiIslandComponent.prototype, "textAlignLeft", null);
__decorate([
    HostBinding('class.tui-island_text-align_center')
], TuiIslandComponent.prototype, "textAlignCenter", null);
__decorate([
    HostBinding('class.tui-island_text-align_right')
], TuiIslandComponent.prototype, "textAlignRight", null);
TuiIslandComponent = __decorate([
    Component({
        selector: 'tui-island, a[tuiIsland]',
        changeDetection: ChangeDetectionStrategy.OnPush,
        template: "<ng-content></ng-content>\n",
        host: {
            class: 'tui-island',
        },
        styles: [":host{display:block}"]
    })
], TuiIslandComponent);
export { TuiIslandComponent };
