// @taiga-ui/core/esm2015/modules/notifications/notification-alert/notification-alert.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, Inject, Input } from '@angular/core';
import { TuiDestroyService, tuiPure } from '@taiga-ui/cdk';
import { timer } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
export const ALERT_AUTOCLOSE_TIMEOUT = 3000;
let TuiNotificationAlertComponent = class TuiNotificationAlertComponent {
    constructor(destroy$) {
        timer(ALERT_AUTOCLOSE_TIMEOUT)
            .pipe(takeUntil(destroy$))
            .subscribe(() => {
            if (this.safeItem.autoClose) {
                this.closeDialog();
            }
        });
    }
    get safeItem() {
        if (!this.item) {
            throw new Error('Notification was created as undefined');
        }
        return this.item;
    }
    get context() {
        return this.calculateContext(this.safeItem);
    }
    closeDialog() {
        this.safeItem.observer.complete();
    }
    calculateContext({ status, data, label, observer, }) {
        return {
            $implicit: status,
            data: data,
            label: label,
            closeHook: () => {
                observer.complete();
            },
            emitHook: (data) => {
                observer.next(data);
            },
            emitAndCloseHook: (data) => {
                observer.next(data);
                observer.complete();
            },
        };
    }
};
TuiNotificationAlertComponent.ctorParameters = () => [
    { type: TuiDestroyService, decorators: [{ type: Inject, args: [TuiDestroyService,] }] }
];
__decorate([
    Input()
], TuiNotificationAlertComponent.prototype, "item", void 0);
__decorate([
    tuiPure
], TuiNotificationAlertComponent.prototype, "calculateContext", null);
TuiNotificationAlertComponent = __decorate([
    Component({
        selector: 'tui-notification-alert',
        changeDetection: ChangeDetectionStrategy.OnPush,
        template: "<tui-notification\n    *ngIf=\"safeItem.hasCloseButton else noClose\"\n    [status]=\"safeItem.status\"\n    [hasIcon]=\"safeItem.hasIcon\"\n    (close)=\"closeDialog()\"\n>\n    <label\n        *ngIf=\"safeItem.label\"\n        automation-id=\"tui-notification-alert__heading\"\n        class=\"heading\"\n    >\n        {{safeItem.label}}\n    </label>\n    <div\n        polymorpheus-outlet\n        automation-id=\"tui-notification-alert__content\"\n        class=\"content\"\n        [content]=\"safeItem.content\"\n        [context]=\"context\"\n    ></div>\n</tui-notification>\n<ng-template #noClose>\n    <tui-notification [status]=\"safeItem.status\" [hasIcon]=\"safeItem.hasIcon\">\n        <label\n            *ngIf=\"safeItem.label\"\n            automation-id=\"tui-notification-alert__heading\"\n            class=\"heading\"\n        >\n            {{safeItem.label}}\n        </label>\n        <div\n            polymorpheus-outlet\n            automation-id=\"tui-notification-alert__content\"\n            class=\"content\"\n            [content]=\"safeItem.content\"\n            [context]=\"context\"\n        ></div>\n    </tui-notification>\n</ng-template>\n",
        providers: [TuiDestroyService],
        styles: [":host{box-shadow:0 8px 16px rgba(51,51,51,.2);display:block;border-radius:var(--tui-radius-m)}.heading{font:var(--tui-font-text-m);line-height:20px;font-weight:700;margin:0}.content{color:var(--tui-text-01);word-wrap:break-word}.content:empty{display:none}"]
    }),
    __param(0, Inject(TuiDestroyService))
], TuiNotificationAlertComponent);
export { TuiNotificationAlertComponent };
