// @taiga-ui/core/esm5/components/hints-host/hints-host.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, ElementRef, Inject, } from '@angular/core';
import { TuiDestroyService } from '@taiga-ui/cdk';
import { tuiFadeIn } from '@taiga-ui/core/animations';
import { TuiHintDirective } from '@taiga-ui/core/directives/hint';
import { TuiHintService } from '@taiga-ui/core/services';
import { Observable } from 'rxjs';
import { skip, takeUntil } from 'rxjs/operators';
var TuiHintsHostComponent = /** @class */ (function () {
    function TuiHintsHostComponent(elementRef, hints$, changeDetectorRef, destroy$) {
        var _this = this;
        this.elementRef = elementRef;
        this.changeDetectorRef = changeDetectorRef;
        /**
         * @awful TODO: remove hints variable
         * and use hints$ after resolving issue with Change Detection
         */
        this.hints = [];
        hints$.pipe(skip(1), takeUntil(destroy$)).subscribe(function (hints) {
            _this.hints = hints;
            _this.changeDetectorRef.detectChanges();
        });
    }
    Object.defineProperty(TuiHintsHostComponent.prototype, "clientRect", {
        get: function () {
            return this.elementRef.nativeElement.getBoundingClientRect();
        },
        enumerable: true,
        configurable: true
    });
    TuiHintsHostComponent.prototype.onHovered = function (hovered, directive) {
        if (directive instanceof TuiHintDirective) {
            directive.componentHovered$.next(hovered);
        }
    };
    TuiHintsHostComponent.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: TuiHintService, decorators: [{ type: Inject, args: [TuiHintService,] }] },
        { type: ChangeDetectorRef, decorators: [{ type: Inject, args: [ChangeDetectorRef,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TuiDestroyService,] }] }
    ]; };
    TuiHintsHostComponent = __decorate([
        Component({
            selector: 'tui-hints-host',
            template: "<tui-hint-box\n    *ngFor=\"let hint of hints\"\n    role=\"tooltip\"\n    [attr.id]=\"hint.id\"\n    [@tuiFadeIn]\n    [mode]=\"hint.mode\"\n    [direction]=\"hint.direction\"\n    [hint]=\"hint\"\n    (tuiHoveredChange)=\"onHovered($event, hint)\"\n>\n    <div polymorpheus-outlet [content]=\"hint.content\">\n        <ng-template let-text>\n            <span class=\"text\">{{text}}</span>\n        </ng-template>\n    </div>\n</tui-hint-box>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            animations: [tuiFadeIn],
            providers: [TuiDestroyService],
            host: {
                'aria-live': 'polite',
            },
            styles: [":host{position:fixed;bottom:0;left:0;width:100%;height:0}.text{white-space:pre-wrap}"]
        }),
        __param(0, Inject(ElementRef)),
        __param(1, Inject(TuiHintService)),
        __param(2, Inject(ChangeDetectorRef)),
        __param(3, Inject(TuiDestroyService))
    ], TuiHintsHostComponent);
    return TuiHintsHostComponent;
}());
export { TuiHintsHostComponent };
