// @taiga-ui/cdk/bundles/taiga-ui-cdk-directives-active-zone.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('@ng-web-apis/common'), require('@taiga-ui/cdk/decorators'), require('@taiga-ui/cdk/observables'), require('@taiga-ui/cdk/utils/dom'), require('@taiga-ui/cdk/utils/focus'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/directives/active-zone', ['exports', '@angular/common', '@angular/core', '@ng-web-apis/common', '@taiga-ui/cdk/decorators', '@taiga-ui/cdk/observables', '@taiga-ui/cdk/utils/dom', '@taiga-ui/cdk/utils/focus', 'rxjs', 'rxjs/operators'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].cdk = global['taiga-ui'].cdk || {}, global['taiga-ui'].cdk.directives = global['taiga-ui'].cdk.directives || {}, global['taiga-ui'].cdk.directives['active-zone'] = {}), global.ng.common, global.ng.core, global.common$1, global['taiga-ui'].cdk.decorators, global['taiga-ui'].cdk.observables, global['taiga-ui'].cdk.utils.dom, global['taiga-ui'].cdk.utils.focus, global.rxjs, global.rxjs.operators));
}(this, (function (exports, common, core, common$1, decorators, observables, dom, focus, rxjs, operators) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    // @dynamic
    var TuiActiveZoneDirective = /** @class */ (function () {
        function TuiActiveZoneDirective(element, directParentActiveZone, ngZone, windowRef, documentRef) {
            var _this = this;
            this.element = element;
            this.directParentActiveZone = directParentActiveZone;
            this.subActiveZones = [];
            this.tuiActiveZoneParent = null;
            var skipNextFocusOut = false;
            if (this.directParentActiveZone) {
                this.directParentActiveZone.addSubActiveZone(this);
            }
            this.tuiActiveZoneChange = rxjs.merge(observables.typedFromEvent(windowRef, 'focusout').pipe(operators.filter(function (event) {
                var actualTarget = dom.getActualTarget(event);
                return (!skipNextFocusOut &&
                    _this.contains(actualTarget) &&
                    !focus.isNativeFocused(actualTarget) &&
                    isValidFocusoutEvent(event) &&
                    !_this.contains(event.relatedTarget));
            }), operators.mapTo(false)), observables.typedFromEvent(windowRef, 'focusin').pipe(operators.map(function (event) { return _this.contains(dom.getActualTarget(event)); })), observables.typedFromEvent(windowRef, 'mousedown').pipe(operators.filter(function (event) {
                return documentRef.body === event.target ||
                    !focus.isNativeFocused(dom.getActualTarget(event));
            }), operators.switchMap(function (event) {
                var actualTarget = dom.getActualTarget(event);
                var targetInZone = _this.contains(actualTarget);
                var activeElement = focus.getNativeFocused(documentRef);
                var focusInZone = activeElement !== null && _this.contains(activeElement);
                var focusNowhere = activeElement === null || activeElement === documentRef.body;
                // If default behavior is prevented — focus either remained
                // where it was or it was moved manually so we just check
                // if the focused element is within the zone or target is
                // within the zone and focus is nowhere
                if (event.defaultPrevented) {
                    return rxjs.of(focusInZone || (focusNowhere && targetInZone));
                }
                // If mouseDown happened inside the zone and focus is outside we
                // return true if target is not focusable or wait for focusIn
                if (targetInZone && !focusInZone && actualTarget instanceof Element) {
                    return !focus.isNativeMouseFocusable(actualTarget)
                        ? rxjs.of(true)
                        : observables.typedFromEvent(windowRef, 'focusin').pipe(operators.take(1), operators.mapTo(targetInZone));
                }
                // If focus is in the zone we wait for the next focusOut event and
                // map it to either true or false, depending on if the mouseDown
                // target is within the zone or not
                if (focusInZone) {
                    // @bad TODO: Think of a way to handle this without side-effects
                    skipNextFocusOut = true;
                    return observables.typedFromEvent(windowRef, 'focusout').pipe(operators.take(1), operators.mapTo(targetInZone));
                }
                return rxjs.of(false);
            }))).pipe(operators.tap(function () {
                skipNextFocusOut = false;
            }), operators.startWith(false), operators.distinctUntilChanged(), operators.skip(1), observables.tuiZoneOptimized(ngZone));
        }
        TuiActiveZoneDirective_1 = TuiActiveZoneDirective;
        Object.defineProperty(TuiActiveZoneDirective.prototype, "tuiActiveZoneParentSetter", {
            set: function (zone) {
                if (this.tuiActiveZoneParent) {
                    this.tuiActiveZoneParent.removeSubActiveZone(this);
                }
                if (zone) {
                    zone.addSubActiveZone(this);
                }
                this.tuiActiveZoneParent = zone;
            },
            enumerable: true,
            configurable: true
        });
        TuiActiveZoneDirective.prototype.ngOnDestroy = function () {
            if (!!this.directParentActiveZone) {
                this.directParentActiveZone.removeSubActiveZone(this);
            }
            if (!!this.tuiActiveZoneParent) {
                this.tuiActiveZoneParent.removeSubActiveZone(this);
            }
        };
        TuiActiveZoneDirective.prototype.contains = function (node) {
            return (!!node &&
                (this.element.nativeElement.contains(node) ||
                    this.subActiveZones.some(function (item, index, array) {
                        return array.indexOf(item) === index && item.contains(node);
                    })));
        };
        TuiActiveZoneDirective.prototype.addSubActiveZone = function (activeZone) {
            this.subActiveZones = __spread(this.subActiveZones, [activeZone]);
        };
        TuiActiveZoneDirective.prototype.removeSubActiveZone = function (activeZone) {
            var index = this.subActiveZones.findIndex(function (item) { return item === activeZone; });
            this.subActiveZones = __spread(this.subActiveZones.slice(0, index), this.subActiveZones.slice(index + 1));
        };
        var TuiActiveZoneDirective_1;
        TuiActiveZoneDirective.ctorParameters = function () { return [
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: TuiActiveZoneDirective, decorators: [{ type: core.Inject, args: [TuiActiveZoneDirective_1,] }, { type: core.Optional }, { type: core.SkipSelf }] },
            { type: core.NgZone, decorators: [{ type: core.Inject, args: [core.NgZone,] }] },
            { type: Window, decorators: [{ type: core.Inject, args: [common$1.WINDOW,] }] },
            { type: Document, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] }
        ]; };
        __decorate([
            core.Input('tuiActiveZoneParent'),
            decorators.tuiDefaultProp()
        ], TuiActiveZoneDirective.prototype, "tuiActiveZoneParentSetter", null);
        __decorate([
            core.Output()
        ], TuiActiveZoneDirective.prototype, "tuiActiveZoneChange", void 0);
        TuiActiveZoneDirective = TuiActiveZoneDirective_1 = __decorate([
            core.Directive({
                selector: '[tuiActiveZone]:not(ng-container), [tuiActiveZoneChange]:not(ng-container), [tuiActiveZoneParent]:not(ng-container)',
                exportAs: 'tuiActiveZone',
            }),
            __param(0, core.Inject(core.ElementRef)),
            __param(1, core.Inject(TuiActiveZoneDirective_1)),
            __param(1, core.Optional()),
            __param(1, core.SkipSelf()),
            __param(2, core.Inject(core.NgZone)),
            __param(3, core.Inject(common$1.WINDOW)),
            __param(4, core.Inject(common.DOCUMENT))
        ], TuiActiveZoneDirective);
        return TuiActiveZoneDirective;
    }());
    // Chrome workaround for triggering `focusout` event upon element removal
    function isValidFocusoutEvent(_a) {
        var relatedTarget = _a.relatedTarget, sourceCapabilities = _a.sourceCapabilities;
        return sourceCapabilities !== null || relatedTarget !== null;
    }

    var TuiActiveZoneModule = /** @class */ (function () {
        function TuiActiveZoneModule() {
        }
        TuiActiveZoneModule = __decorate([
            core.NgModule({
                declarations: [TuiActiveZoneDirective],
                exports: [TuiActiveZoneDirective],
            })
        ], TuiActiveZoneModule);
        return TuiActiveZoneModule;
    }());

    exports.TuiActiveZoneDirective = TuiActiveZoneDirective;
    exports.TuiActiveZoneModule = TuiActiveZoneModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
