// @taiga-ui/cdk/bundles/taiga-ui-cdk-classes.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@taiga-ui/cdk/constants')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/cdk/classes', ['exports', '@taiga-ui/cdk/constants'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].cdk = global['taiga-ui'].cdk || {}, global['taiga-ui'].cdk.classes = {}), global['taiga-ui'].cdk.constants));
}(this, (function (exports, constants) { 'use strict';

    var tuiAssert = {
        enabled: false,
        get assert() {
            return this.enabled
                ? Function.prototype.bind.call(console.assert, console)
                : constants.EMPTY_FUNCTION;
        },
    };

    var TuiValidationError = /** @class */ (function () {
        function TuiValidationError(message, context) {
            this.message = message;
            this.context = context;
        }
        return TuiValidationError;
    }());

    exports.TuiValidationError = TuiValidationError;
    exports.tuiAssert = tuiAssert;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
