// @taiga-ui/core/esm2015/abstract/abstract-hint.js
import { __decorate } from "tslib";
import { Directive, ElementRef, Input, OnDestroy } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import { TuiHintService } from '@taiga-ui/core/services';
let AbstractTuiHint = class AbstractTuiHint {
    constructor(elementRef, hintService) {
        this.elementRef = elementRef;
        this.hintService = hintService;
        this.mode = null;
        this.direction = 'bottom-left';
        this.content = '';
    }
    ngOnDestroy() {
        this.hideTooltip();
    }
    showTooltip() {
        this.hintService.add(this);
    }
    hideTooltip() {
        this.hintService.remove(this);
    }
};
AbstractTuiHint.ctorParameters = () => [
    { type: ElementRef },
    { type: TuiHintService }
];
__decorate([
    Input('tuiHintMode'),
    tuiDefaultProp()
], AbstractTuiHint.prototype, "mode", void 0);
__decorate([
    Input('tuiHintDirection'),
    tuiDefaultProp()
], AbstractTuiHint.prototype, "direction", void 0);
AbstractTuiHint = __decorate([
    Directive()
], AbstractTuiHint);
export { AbstractTuiHint };
