// @taiga-ui/cdk/esm5/components/portal-host/portal.service.js
import { __decorate } from "tslib";
import { Injectable, } from '@angular/core';
import * as i0 from "@angular/core";
var NO_HOST = 'Portals cannot be used without TuiPortalHostComponent';
/**
 * Service for displaying portals
 */
var TuiPortalService = /** @class */ (function () {
    function TuiPortalService() {
    }
    Object.defineProperty(TuiPortalService.prototype, "safeHost", {
        get: function () {
            if (!this.host) {
                throw new Error(NO_HOST);
            }
            return this.host;
        },
        enumerable: true,
        configurable: true
    });
    TuiPortalService.prototype.attach = function (host) {
        this.host = host;
    };
    TuiPortalService.prototype.add = function (componentFactory, injector) {
        return this.safeHost.addComponentChild(componentFactory, injector);
    };
    TuiPortalService.prototype.remove = function (_a) {
        var hostView = _a.hostView;
        hostView.destroy();
    };
    TuiPortalService.prototype.addTemplate = function (templateRef, context) {
        return this.safeHost.addTemplateChild(templateRef, context);
    };
    TuiPortalService.prototype.removeTemplate = function (viewRef) {
        viewRef.destroy();
    };
    TuiPortalService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TuiPortalService_Factory() { return new TuiPortalService(); }, token: TuiPortalService, providedIn: "root" });
    TuiPortalService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], TuiPortalService);
    return TuiPortalService;
}());
export { TuiPortalService };
