// @taiga-ui/kit/esm5/components/tabs/tab/tab.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, HostBinding, Inject, Optional, } from '@angular/core';
import { RouterLinkActive } from '@angular/router';
import { TUI_IS_ANDROID, TUI_IS_IOS, TuiFocusVisibleService } from '@taiga-ui/cdk';
import { TUI_MODE } from '@taiga-ui/core';
import { TUI_MOBILE_AWARE } from '@taiga-ui/kit/tokens';
import { Observable } from 'rxjs';
import { TUI_TAB_EVENT, TUI_TAB_PROVIDERS } from './tab.providers';
var TuiTabComponent = /** @class */ (function () {
    function TuiTabComponent(routerLinkActive, mode$, mobileAware, isIos, isAndroid, event$, _a, focusVisible$) {
        var _this = this;
        var nativeElement = _a.nativeElement;
        this.routerLinkActive = routerLinkActive;
        this.mode$ = mode$;
        this.focusVisible = false;
        this.isIos = mobileAware && isIos;
        this.isAndroid = mobileAware && isAndroid;
        event$.subscribe(function (event) {
            nativeElement.dispatchEvent(event);
        });
        focusVisible$.subscribe(function (visible) {
            _this.focusVisible = visible;
        });
    }
    Object.defineProperty(TuiTabComponent.prototype, "isActive", {
        get: function () {
            return !!this.routerLinkActive && this.routerLinkActive.isActive;
        },
        enumerable: true,
        configurable: true
    });
    TuiTabComponent.ctorParameters = function () { return [
        { type: RouterLinkActive, decorators: [{ type: Optional }, { type: Inject, args: [RouterLinkActive,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TUI_MODE,] }] },
        { type: Boolean, decorators: [{ type: Inject, args: [TUI_MOBILE_AWARE,] }] },
        { type: Boolean, decorators: [{ type: Inject, args: [TUI_IS_IOS,] }] },
        { type: Boolean, decorators: [{ type: Inject, args: [TUI_IS_ANDROID,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TUI_TAB_EVENT,] }] },
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: TuiFocusVisibleService, decorators: [{ type: Inject, args: [TuiFocusVisibleService,] }] }
    ]; };
    __decorate([
        HostBinding('class._ios')
    ], TuiTabComponent.prototype, "isIos", void 0);
    __decorate([
        HostBinding('class._android')
    ], TuiTabComponent.prototype, "isAndroid", void 0);
    __decorate([
        HostBinding('class._focus-visible')
    ], TuiTabComponent.prototype, "focusVisible", void 0);
    __decorate([
        HostBinding('class._active')
    ], TuiTabComponent.prototype, "isActive", null);
    TuiTabComponent = __decorate([
        Component({
            selector: "a[tuiTab]:not([routerLink]), a[tuiTab][routerLink][routerLinkActive], button[tuiTab]",
            template: "<span class=\"highlight\"><ng-content></ng-content></span>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: TUI_TAB_PROVIDERS,
            host: {
                '($.data-mode.attr)': 'mode$',
            },
            styles: [":host-context(._underline):hover{box-shadow:inset 0 -2px var(--tui-base-03)}:host{transition-duration:.3s;transition-timing-function:ease-in-out;-webkit-appearance:none;-moz-appearance:none;appearance:none;padding:0;border:0;background:0 0;font-size:inherit;line-height:inherit;position:relative;display:flex;flex-shrink:0;height:100%;justify-content:space-between;align-items:center;white-space:nowrap;cursor:pointer;outline:0;text-decoration:none;color:inherit;transition-property:color,box-shadow,opacity}:host:not(:last-child){margin-right:24px}:host:hover{color:var(--tui-text-01)}:host:disabled{opacity:var(--tui-disabled-opacity);pointer-events:none}:host:not([data-mode])._active{opacity:1;color:var(--tui-text-01);box-shadow:none}:host[data-mode=onDark]{color:rgba(255,255,255,.72)}:host[data-mode=onDark]._active,:host[data-mode=onDark]:hover{color:#fff}:host[data-mode=onDark]._focus-visible .highlight{background:var(--tui-clear-inverse);color:var(--tui-text-01)}:host[data-mode=onLight]{color:rgba(0,0,0,.56)}:host[data-mode=onLight]._focus-visible .highlight{background:var(--tui-text-01);color:#fff}:host._focus-visible .highlight{background:var(--tui-selection)}:host._android{transition-property:opacity;transition-duration:.3s;transition-timing-function:ease-in-out;height:48px;flex:1;flex-shrink:0;justify-content:center;margin:0;text-transform:uppercase;color:var(--tui-base-09);opacity:.5;font-size:14px;font-weight:500;letter-spacing:-.13px;font-family:-apple-system,BlinkMacSystemFont,Roboto,'Helvetica Neue',sans-serif}:host._android:hover{box-shadow:none}:host._android._active{opacity:1;color:var(--tui-base-09)}:host._android .highlight{width:auto}:host._ios{height:28px;flex:1;flex-shrink:0;justify-content:center;margin:0;color:var(--tui-base-01);opacity:1;font-size:13px;font-weight:500;letter-spacing:-.08px;font-family:-apple-system,BlinkMacSystemFont,Roboto,'Helvetica Neue',sans-serif}:host._ios:hover{box-shadow:none}:host._ios._active{font-size:14px;font-weight:600;color:var(--tui-link)}:host._ios .highlight{width:auto}.highlight{display:inline-flex;width:100%;align-items:center;justify-content:inherit}.highlight ::ng-deep>*{flex-shrink:0}"]
        }),
        __param(0, Optional()),
        __param(0, Inject(RouterLinkActive)),
        __param(1, Inject(TUI_MODE)),
        __param(2, Inject(TUI_MOBILE_AWARE)),
        __param(3, Inject(TUI_IS_IOS)),
        __param(4, Inject(TUI_IS_ANDROID)),
        __param(5, Inject(TUI_TAB_EVENT)),
        __param(6, Inject(ElementRef)),
        __param(7, Inject(TuiFocusVisibleService))
    ], TuiTabComponent);
    return TuiTabComponent;
}());
export { TuiTabComponent };
