// @taiga-ui/core/esm5/components/dialog/dialog.service.js
import { __decorate, __extends } from "tslib";
import { Injectable } from '@angular/core';
import { AbstractTuiDialogService } from '@taiga-ui/cdk';
import { PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { TuiDialogComponent } from './dialog.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk/services";
var DIALOG = new PolymorpheusComponent(TuiDialogComponent);
var ɵ0 = undefined;
var DEFAULT_OPTIONS = {
    size: 'm',
    required: false,
    closeable: true,
    dismissible: true,
    label: '',
    header: '',
    data: ɵ0,
};
var TuiDialogService = /** @class */ (function (_super) {
    __extends(TuiDialogService, _super);
    function TuiDialogService() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.component = DIALOG;
        _this.defaultOptions = DEFAULT_OPTIONS;
        return _this;
    }
    TuiDialogService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TuiDialogService_Factory() { return new TuiDialogService(i0.ɵɵinject(i1.TuiIdService)); }, token: TuiDialogService, providedIn: "root" });
    TuiDialogService = __decorate([
        Injectable({
            providedIn: 'root',
        })
    ], TuiDialogService);
    return TuiDialogService;
}(AbstractTuiDialogService));
export { TuiDialogService };
export { ɵ0 };
