// @taiga-ui/core/bundles/taiga-ui-core-components-dialog.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core/animations'), require('@taiga-ui/core/tokens'), require('@tinkoff/ng-polymorpheus'), require('rxjs'), require('@angular/common'), require('@ng-web-apis/common'), require('rxjs/operators'), require('@taiga-ui/core/components/button'), require('@taiga-ui/cdk/services')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/dialog', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core/animations', '@taiga-ui/core/tokens', '@tinkoff/ng-polymorpheus', 'rxjs', '@angular/common', '@ng-web-apis/common', 'rxjs/operators', '@taiga-ui/core/components/button', '@taiga-ui/cdk/services'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.components = global['taiga-ui'].core.components || {}, global['taiga-ui'].core.components.dialog = {}), global.ng.core, global.cdk, global['taiga-ui'].core.animations, global['taiga-ui'].core.tokens, global.ngPolymorpheus, global.rxjs, global.ng.common, global.common$1, global.rxjs.operators, global['taiga-ui'].core.components.button, global.services));
}(this, (function (exports, core, cdk, animations, tokens, ngPolymorpheus, rxjs, common, common$1, operators, button, services) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var TUI_DIALOGS_CLOSE = new core.InjectionToken('A stream to close dialogs', {
        factory: function () { return rxjs.EMPTY; },
    });
    var SCROLLBAR_PLACEHOLDER = 17;
    function dialogCloseStreamFactory(documentRef, windowRef, _a, close$, destroy$, _b) {
        var nativeElement = _a.nativeElement;
        var dismissible = _b.dismissible;
        return dismissible
            ? rxjs.merge(cdk.typedFromEvent(nativeElement, 'click').pipe(operators.filter(cdk.isCurrentTarget)), cdk.typedFromEvent(documentRef, 'keydown').pipe(operators.filter(function (_a) {
                var key = _a.key, target = _a.target;
                return key === 'Escape' &&
                    target instanceof Element &&
                    (!cdk.containsOrAfter(nativeElement, target) ||
                        nativeElement.contains(target));
            })), cdk.typedFromEvent(documentRef, 'mousedown').pipe(operators.filter(function (_a) {
                var target = _a.target, clientX = _a.clientX;
                return target instanceof Element &&
                    windowRef.innerWidth - clientX > SCROLLBAR_PLACEHOLDER &&
                    !cdk.containsOrAfter(nativeElement, target);
            }), operators.switchMapTo(cdk.typedFromEvent(documentRef, 'mouseup').pipe(operators.take(1), operators.filter(function (_a) {
                var target = _a.target;
                return target instanceof Element &&
                    !cdk.containsOrAfter(nativeElement, target);
            })))), close$).pipe(operators.takeUntil(destroy$))
            : close$;
    }
    var TUI_DIALOG_CLOSE_STREAM = new core.InjectionToken('Dialogs closing stream');
    var TUI_DIALOG_PROVIDERS = [
        cdk.TuiDestroyService,
        {
            provide: TUI_DIALOG_CLOSE_STREAM,
            deps: [
                common.DOCUMENT,
                common$1.WINDOW,
                core.ElementRef,
                TUI_DIALOGS_CLOSE,
                cdk.TuiDestroyService,
                ngPolymorpheus.POLYMORPHEUS_CONTEXT,
            ],
            useFactory: dialogCloseStreamFactory,
        },
    ];

    var SMALL_DIALOGS_ANIMATION = { value: '', params: { start: '40px' } };
    var FULLSCREEN_DIALOGS_ANIMATION = { value: '', params: { start: '100vh' } };
    var REQUIRED_ERROR = new Error('Required dialog was dismissed');
    // @dynamic
    var TuiDialogComponent = /** @class */ (function () {
        function TuiDialogComponent(isMobile, context, close$, closeWord$) {
            var _this = this;
            this.isMobile = isMobile;
            this.context = context;
            this.closeWord$ = closeWord$;
            close$.subscribe(function () {
                _this.close();
            });
        }
        Object.defineProperty(TuiDialogComponent.prototype, "size", {
            get: function () {
                return this.context.size;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiDialogComponent.prototype, "header", {
            get: function () {
                return this.context.header;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiDialogComponent.prototype, "h", {
            get: function () {
                if (this.isMobile) {
                    return 'h5';
                }
                switch (this.size) {
                    case 's':
                        return 'h5';
                    case 'm':
                        return 'h4';
                    default:
                        return 'h3';
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiDialogComponent.prototype, "slideInTop", {
            get: function () {
                return this.size === 'fullscreen' || this.size === 'page' || this.isMobile
                    ? FULLSCREEN_DIALOGS_ANIMATION
                    : SMALL_DIALOGS_ANIMATION;
            },
            enumerable: true,
            configurable: true
        });
        TuiDialogComponent.prototype.close = function () {
            if (this.context.required) {
                this.context.$implicit.error(REQUIRED_ERROR);
            }
            else {
                this.context.$implicit.complete();
            }
        };
        TuiDialogComponent.ctorParameters = function () { return [
            { type: Boolean, decorators: [{ type: core.Inject, args: [cdk.TUI_IS_MOBILE,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [ngPolymorpheus.POLYMORPHEUS_CONTEXT,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [TUI_DIALOG_CLOSE_STREAM,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_CLOSE_WORD,] }] }
        ]; };
        __decorate([
            core.HostBinding('attr.data-size')
        ], TuiDialogComponent.prototype, "size", null);
        __decorate([
            core.HostBinding('class._centered')
        ], TuiDialogComponent.prototype, "header", null);
        __decorate([
            core.HostBinding('@tuiSlideInTop')
        ], TuiDialogComponent.prototype, "slideInTop", null);
        TuiDialogComponent = __decorate([
            core.Component({
                selector: 'tui-dialog',
                template: "<header\n    *ngIf=\"header\"\n    polymorpheus-outlet\n    class=\"header\"\n    [content]=\"header\"\n    [context]=\"context\"\n></header>\n<div class=\"content\">\n    <h2 class=\"heading tui-text_{{h}}\" [id]=\"context.id\">{{context.label}}</h2>\n    <section\n        polymorpheus-outlet\n        [context]=\"context\"\n        [content]=\"context.content\"\n    ></section>\n</div>\n<div *ngIf=\"context.closeable\" class=\"wrapper\">\n    <button\n        tuiIconButton\n        tuiPreventDefault=\"mousedown\"\n        automation-id=\"tui-dialog__close\"\n        class=\"close\"\n        size=\"s\"\n        appearance=\"\"\n        icon=\"tuiIconCloseLarge\"\n        shape=\"rounded\"\n        [title]=\"closeWord$ | async\"\n        (click)=\"close()\"\n    ></button>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: TUI_DIALOG_PROVIDERS,
                animations: [animations.tuiSlideInTop, animations.tuiFadeIn],
                host: {
                    '[@tuiFadeIn]': 'true',
                },
                styles: [":host{font:var(--tui-font-text-m);position:relative;display:flex;flex-direction:column;box-sizing:border-box;margin:auto;border-radius:16px;border:40px solid transparent}:host:after{position:absolute;top:0;left:0;width:100%;height:100%;box-shadow:0 18px 30px rgba(51,51,51,.52);content:'';border-radius:inherit;pointer-events:none}:host[data-size='s']{width:480px}:host[data-size='s'] .content{padding:24px}:host[data-size='s'] .heading{margin-bottom:8px}:host[data-size='m']{width:680px}:host[data-size='m'] .heading{margin-bottom:12px}:host[data-size='l']{width:880px}:host[data-size='l'] .heading{margin-bottom:16px}:host[data-size=fullscreen],:host[data-size=page]{width:100vw;min-height:100vh;border-radius:0;border:none;background:var(--tui-base-01)}:host[data-size=fullscreen] .content,:host[data-size=page] .content{padding:48px calc(50vw - 720px / 2)}:host[data-size=fullscreen] .heading,:host[data-size=page] .heading{margin-bottom:16px}:host[data-size=page] .content{padding:0}:host._centered{text-align:center}@media screen and (max-width:599px){:host[data-size]{min-width:100vw;width:100vw;max-width:100vw;border-radius:0;border:none;margin:auto 0 0}:host[data-size] .content{padding:32px 24px}:host[data-size] .heading{margin-bottom:8px}:host[data-size=fullscreen] .header{flex:1}}.heading{margin:0;word-wrap:break-word}.heading:empty{display:none}.header{display:flex;border-top-left-radius:inherit;border-top-right-radius:inherit;overflow:hidden}.content{border-radius:inherit;padding:32px;background:var(--tui-base-01)}.content:not(:first-child){border-top-left-radius:0;border-top-right-radius:0}.wrapper{position:-webkit-sticky;position:sticky;top:0;order:-1}.close{transition-property:background;transition-duration:.3s;transition-timing-function:ease-in-out;position:absolute;top:0;right:0;color:var(--tui-base-01)}@media screen and (max-width:599px){.close{background:0 0;color:var(--tui-base-06)}.close:hover{color:var(--tui-base-07)}}@media screen and (min-width:600px){.close{position:fixed;top:24px;right:24px;background:rgba(104,104,104,.96)}.close:hover{background:rgba(159,159,159,.86)}:host:not([data-size=fullscreen]) .close{-webkit-animation:.3s tuiFadeIn;animation:.3s tuiFadeIn}:host:not([data-size=fullscreen]).ng-animating .close{display:none}}"]
            }),
            __param(0, core.Inject(cdk.TUI_IS_MOBILE)),
            __param(1, core.Inject(ngPolymorpheus.POLYMORPHEUS_CONTEXT)),
            __param(2, core.Inject(TUI_DIALOG_CLOSE_STREAM)),
            __param(3, core.Inject(tokens.TUI_CLOSE_WORD))
        ], TuiDialogComponent);
        return TuiDialogComponent;
    }());

    var DIALOG = new ngPolymorpheus.PolymorpheusComponent(TuiDialogComponent);
    var ɵ0 = undefined;
    var DEFAULT_OPTIONS = {
        size: 'm',
        required: false,
        closeable: true,
        dismissible: true,
        label: '',
        header: '',
        data: ɵ0,
    };
    var TuiDialogService = /** @class */ (function (_super) {
        __extends(TuiDialogService, _super);
        function TuiDialogService() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            _this.component = DIALOG;
            _this.defaultOptions = DEFAULT_OPTIONS;
            return _this;
        }
        TuiDialogService.ɵprov = core.ɵɵdefineInjectable({ factory: function TuiDialogService_Factory() { return new TuiDialogService(core.ɵɵinject(services.TuiIdService)); }, token: TuiDialogService, providedIn: "root" });
        TuiDialogService = __decorate([
            core.Injectable({
                providedIn: 'root',
            })
        ], TuiDialogService);
        return TuiDialogService;
    }(cdk.AbstractTuiDialogService));

    var TuiDialogModule = /** @class */ (function () {
        function TuiDialogModule() {
        }
        TuiDialogModule = __decorate([
            core.NgModule({
                declarations: [TuiDialogComponent],
                exports: [TuiDialogComponent],
                entryComponents: [TuiDialogComponent],
                providers: [
                    {
                        provide: cdk.TUI_DIALOGS,
                        useExisting: TuiDialogService,
                        multi: true,
                    },
                ],
                imports: [ngPolymorpheus.PolymorpheusModule, button.TuiButtonModule, common.CommonModule, cdk.TuiPreventDefaultModule],
            })
        ], TuiDialogModule);
        return TuiDialogModule;
    }());

    exports.TUI_DIALOGS_CLOSE = TUI_DIALOGS_CLOSE;
    exports.TUI_DIALOG_CLOSE_STREAM = TUI_DIALOG_CLOSE_STREAM;
    exports.TUI_DIALOG_PROVIDERS = TUI_DIALOG_PROVIDERS;
    exports.TuiDialogComponent = TuiDialogComponent;
    exports.TuiDialogModule = TuiDialogModule;
    exports.TuiDialogService = TuiDialogService;
    exports.dialogCloseStreamFactory = dialogCloseStreamFactory;
    exports.ɵ0 = ɵ0;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
