// @taiga-ui/core/bundles/taiga-ui-core-constants.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@taiga-ui/cdk')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/constants', ['exports', '@taiga-ui/cdk'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.constants = {}), global.cdk));
}(this, (function (exports, cdk) { 'use strict';

    var DEFAULT_MARGIN = 4;
    var DEFAULT_MIN_HEIGHT = 80;
    var DEFAULT_MAX_HEIGHT = 400;
    var DEFAULT_MAX_WIDTH = 600;

    var TUI_ANIMATION_DURATION = '300ms';

    var DEFAULT_ICONS_PATH = function (name) {
        return name.includes('.svg#') ? name : "#" + name;
    };

    var TUI_DEFAULT_MARKER_HANDLER = function () { return cdk.EMPTY_ARRAY; };

    var tuiEditingKeys = [
        'Spacebar',
        'Backspace',
        'Delete',
        'ArrowLeft',
        'ArrowRight',
        'Left',
        'Right',
        'End',
        'Home',
    ];

    /**
     * An event indicating that async data for expand has finished loading.
     * Dispatch to finish loading states for {@link TuiExpandComponent}.
     */
    var TUI_EXPAND_LOADED = 'tui-expand-loaded';
    /**
     * An event for scrolling an element into view within {@link TuiScrollbarComponent}.
     */
    var TUI_SCROLL_INTO_VIEW = 'tui-scroll-into-view';
    /**
     * An event indicating and error during icon loading in {@link TuiSvgComponent}.
     */
    var TUI_ICON_ERROR = 'tui-icon-error';

    var MASK_CARET_TRAP = '[]';

    /**
     * Media constants are exact the same as css media breakpoints
     */
    var MEDIA = {
        mobile: 320,
        tablet: 600,
        desktopSmall: 1024,
        desktopLarge: 1360,
    };

    var TUI_DIGIT_REGEXP = /\d/;
    var TUI_NON_DIGIT_REGEXP = /\D/;
    var TUI_NON_DIGITS_REGEXP = /\D+/g;
    var TUI_LAST_PUNCTUATION_MARK_REGEXP = /[.,\\/#!$%\\^&\\*;:{}=\\-_\`~()]$/;
    var TUI_LATIN_REGEXP = /[A-z]/;
    var TUI_LATIN_AND_NUMBERS_REGEXP = /[A-z|0-9]/;

    var VERSION = '2.7.0';

    exports.DEFAULT_ICONS_PATH = DEFAULT_ICONS_PATH;
    exports.DEFAULT_MARGIN = DEFAULT_MARGIN;
    exports.DEFAULT_MAX_HEIGHT = DEFAULT_MAX_HEIGHT;
    exports.DEFAULT_MAX_WIDTH = DEFAULT_MAX_WIDTH;
    exports.DEFAULT_MIN_HEIGHT = DEFAULT_MIN_HEIGHT;
    exports.MASK_CARET_TRAP = MASK_CARET_TRAP;
    exports.MEDIA = MEDIA;
    exports.TUI_ANIMATION_DURATION = TUI_ANIMATION_DURATION;
    exports.TUI_DEFAULT_MARKER_HANDLER = TUI_DEFAULT_MARKER_HANDLER;
    exports.TUI_DIGIT_REGEXP = TUI_DIGIT_REGEXP;
    exports.TUI_EXPAND_LOADED = TUI_EXPAND_LOADED;
    exports.TUI_ICON_ERROR = TUI_ICON_ERROR;
    exports.TUI_LAST_PUNCTUATION_MARK_REGEXP = TUI_LAST_PUNCTUATION_MARK_REGEXP;
    exports.TUI_LATIN_AND_NUMBERS_REGEXP = TUI_LATIN_AND_NUMBERS_REGEXP;
    exports.TUI_LATIN_REGEXP = TUI_LATIN_REGEXP;
    exports.TUI_NON_DIGITS_REGEXP = TUI_NON_DIGITS_REGEXP;
    exports.TUI_NON_DIGIT_REGEXP = TUI_NON_DIGIT_REGEXP;
    exports.TUI_SCROLL_INTO_VIEW = TUI_SCROLL_INTO_VIEW;
    exports.VERSION = VERSION;
    exports.tuiEditingKeys = tuiEditingKeys;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
