// @taiga-ui/core/bundles/taiga-ui-core-components-svg.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('@angular/platform-browser'), require('@ng-web-apis/common'), require('@taiga-ui/cdk'), require('@taiga-ui/core/constants'), require('@taiga-ui/core/services'), require('@taiga-ui/core/tokens'), require('@taiga-ui/core/utils/miscellaneous'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/svg', ['exports', '@angular/common', '@angular/core', '@angular/platform-browser', '@ng-web-apis/common', '@taiga-ui/cdk', '@taiga-ui/core/constants', '@taiga-ui/core/services', '@taiga-ui/core/tokens', '@taiga-ui/core/utils/miscellaneous', 'rxjs', 'rxjs/operators'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.components = global['taiga-ui'].core.components || {}, global['taiga-ui'].core.components.svg = {}), global.ng.common, global.ng.core, global.ng.platformBrowser, global.common$1, global.cdk, global['taiga-ui'].core.constants, global['taiga-ui'].core.services, global['taiga-ui'].core.tokens, global['taiga-ui'].core.utils.miscellaneous, global.rxjs, global.rxjs.operators));
}(this, (function (exports, common, core, platformBrowser, common$1, cdk, constants, services, tokens, miscellaneous, rxjs, operators) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var UNDEFINED_NAMED_ICON = 'Attempted to use undefined named icon';
    var MISSING_EXTERNAL_ICON = 'External icon is missing on the given URL';
    var FAILED_EXTERNAL_ICON = 'Failed to load external SVG';
    // TODO: Consider moving to CDK along with SvgService and SvgDefsHostComponent
    // @dynamic
    var TuiSvgComponent = /** @class */ (function () {
        function TuiSvgComponent(documentRef, windowRef, iconsPath, tuiSanitizer, svgService, staticRequestService, sanitizer, elementRef, userAgent) {
            var _this = this;
            this.documentRef = documentRef;
            this.windowRef = windowRef;
            this.iconsPath = iconsPath;
            this.tuiSanitizer = tuiSanitizer;
            this.svgService = svgService;
            this.staticRequestService = staticRequestService;
            this.sanitizer = sanitizer;
            this.elementRef = elementRef;
            this.userAgent = userAgent;
            this.icon = '';
            this.src$ = new rxjs.ReplaySubject(1);
            this.isIE = cdk.isIE(this.userAgent);
            this.innerHTML$ = this.src$.pipe(operators.switchMap(function () {
                return _this.isExternal
                    ? _this.getExternalIcon(_this.icon)
                    : rxjs.of(_this.getSafeHtml(_this.icon));
            }), operators.startWith(''));
        }
        Object.defineProperty(TuiSvgComponent.prototype, "src", {
            set: function (src) {
                this.icon = src;
                this.src$.next();
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "use", {
            get: function () {
                return this.icon.includes('.svg#')
                    ? this.icon
                    : this.resolveName(this.icon, this.iconsPath);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isInnerHTML", {
            get: function () {
                return this.isSrc || this.isExternal || (this.isName && this.isShadowDOM);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isShadowDOM", {
            get: function () {
                return (cdk.getDocumentOrShadowRoot(this.elementRef.nativeElement) !== this.documentRef);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isUse", {
            get: function () {
                return this.use.includes('.svg#');
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isExternal", {
            get: function () {
                return this.isUrl || (this.isIE && this.isUse) || this.isCrossDomain;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isUrl", {
            get: function () {
                return this.icon.endsWith('.svg');
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isSrc", {
            get: function () {
                return miscellaneous.isPresumedHTMLString(this.icon);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isName", {
            get: function () {
                return !this.isUrl && !this.isUse && !this.isSrc;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiSvgComponent.prototype, "isCrossDomain", {
            get: function () {
                var _a = this, use = _a.use, isUse = _a.isUse, windowRef = _a.windowRef;
                return (isUse &&
                    use.startsWith('http') &&
                    !!windowRef.origin &&
                    !use.startsWith(windowRef.origin));
            },
            enumerable: true,
            configurable: true
        });
        TuiSvgComponent.prototype.onError = function (message) {
            if (message === void 0) { message = MISSING_EXTERNAL_ICON; }
            var icon = this.icon;
            var event = cdk.tuiCustomEvent(constants.TUI_ICON_ERROR, {
                bubbles: true,
                detail: {
                    message: message,
                    icon: icon,
                },
            }, this.documentRef);
            cdk.tuiAssert.assert(false, message, icon);
            this.elementRef.nativeElement.dispatchEvent(event);
        };
        TuiSvgComponent.prototype.resolveName = function (name, iconsPath) {
            return iconsPath(name);
        };
        TuiSvgComponent.prototype.getSafeHtml = function (src) {
            return this.isSrc ? this.sanitize(src) : this.process(src);
        };
        TuiSvgComponent.prototype.process = function (src) {
            var icon = this.svgService.getOriginal(src);
            if (this.isName && !icon && !!src) {
                this.onError(UNDEFINED_NAMED_ICON);
            }
            // Empty line for innerHTML when icon is shown through USE tag
            return !this.isShadowDOM || !this.isName ? '' : this.sanitize(icon || '');
        };
        TuiSvgComponent.prototype.sanitize = function (src) {
            return this.tuiSanitizer
                ? this.sanitizer.bypassSecurityTrustHtml(this.tuiSanitizer.sanitize(core.SecurityContext.HTML, src) || '')
                : src;
        };
        // @bad TODO: Create a simple XMLHttpRequest to Observable service
        TuiSvgComponent.prototype.getExternalIcon = function (src) {
            var _this = this;
            var url = src.includes('.svg') ? src : this.use;
            return this.staticRequestService.request(url).pipe(operators.catchError(function () {
                _this.onError(FAILED_EXTERNAL_ICON);
                return rxjs.of('');
            }), operators.map(function (response) {
                return _this.sanitize(response.replace('<svg', '<svg focusable="false"'));
            }));
        };
        TuiSvgComponent.ctorParameters = function () { return [
            { type: Document, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] },
            { type: Window, decorators: [{ type: core.Inject, args: [common$1.WINDOW,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [tokens.TUI_ICONS_PATH,] }] },
            { type: core.Sanitizer, decorators: [{ type: core.Optional }, { type: core.Inject, args: [tokens.TUI_SANITIZER,] }] },
            { type: services.TuiSvgService, decorators: [{ type: core.Inject, args: [services.TuiSvgService,] }] },
            { type: cdk.TuiStaticRequestService, decorators: [{ type: core.Inject, args: [cdk.TuiStaticRequestService,] }] },
            { type: platformBrowser.DomSanitizer, decorators: [{ type: core.Inject, args: [platformBrowser.DomSanitizer,] }] },
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: String, decorators: [{ type: core.Inject, args: [common$1.USER_AGENT,] }] }
        ]; };
        __decorate([
            core.Input(),
            cdk.tuiRequiredSetter()
        ], TuiSvgComponent.prototype, "src", null);
        __decorate([
            cdk.tuiPure
        ], TuiSvgComponent.prototype, "resolveName", null);
        TuiSvgComponent = __decorate([
            core.Component({
                selector: 'tui-svg',
                template: "<ng-container *tuiLet=\"innerHTML$ | async as innerHTML\">\n    <div\n        *ngIf=\"isInnerHTML; else useTemplate\"\n        class=\"src\"\n        [innerHTML]=\"innerHTML\"\n    ></div>\n    <ng-template #useTemplate>\n        <svg\n            version=\"1.1\"\n            xmlns=\"http://www.w3.org/2000/svg\"\n            xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n            focusable=\"false\"\n            width=\"100%\"\n            height=\"100%\"\n            (error)=\"onError()\"\n        >\n            <use [attr.xlink:href]=\"use\"></use>\n        </svg>\n    </ng-template>\n</ng-container>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [":host{display:inline-block;vertical-align:middle;flex-shrink:0;line-height:0;height:24px;width:24px;fill:currentColor}.src{display:flex;height:100%;align-items:center;justify-content:center}"]
            }),
            __param(0, core.Inject(common.DOCUMENT)),
            __param(1, core.Inject(common$1.WINDOW)),
            __param(2, core.Inject(tokens.TUI_ICONS_PATH)),
            __param(3, core.Optional()),
            __param(3, core.Inject(tokens.TUI_SANITIZER)),
            __param(4, core.Inject(services.TuiSvgService)),
            __param(5, core.Inject(cdk.TuiStaticRequestService)),
            __param(6, core.Inject(platformBrowser.DomSanitizer)),
            __param(7, core.Inject(core.ElementRef)),
            __param(8, core.Inject(common$1.USER_AGENT))
        ], TuiSvgComponent);
        return TuiSvgComponent;
    }());

    var TuiSvgModule = /** @class */ (function () {
        function TuiSvgModule() {
        }
        TuiSvgModule = __decorate([
            core.NgModule({
                imports: [common.CommonModule, cdk.TuiLetModule],
                declarations: [TuiSvgComponent],
                exports: [TuiSvgComponent],
            })
        ], TuiSvgModule);
        return TuiSvgModule;
    }());

    exports.TuiSvgComponent = TuiSvgComponent;
    exports.TuiSvgModule = TuiSvgModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
