// @taiga-ui/core/bundles/taiga-ui-core-components-loader.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('@ng-web-apis/common'), require('@taiga-ui/cdk'), require('@taiga-ui/core/utils/miscellaneous'), require('rxjs'), require('rxjs/operators'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/loader', ['exports', '@angular/common', '@angular/core', '@ng-web-apis/common', '@taiga-ui/cdk', '@taiga-ui/core/utils/miscellaneous', 'rxjs', 'rxjs/operators', '@tinkoff/ng-polymorpheus'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.components = global['taiga-ui'].core.components || {}, global['taiga-ui'].core.components.loader = {}), global.ng.common, global.ng.core, global.common$1, global.cdk, global['taiga-ui'].core.utils.miscellaneous, global.rxjs, global.rxjs.operators, global.ngPolymorpheus));
}(this, (function (exports, common, core, common$1, cdk, miscellaneous, rxjs, operators, ngPolymorpheus) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    // @dynamic
    var TuiLoaderComponent = /** @class */ (function () {
        function TuiLoaderComponent(documentRef, elementRef, userAgent, isIos) {
            this.documentRef = documentRef;
            this.elementRef = elementRef;
            this.userAgent = userAgent;
            this.isIos = isIos;
            this.size = 'm';
            this.inheritColor = false;
            this.overlay = false;
            this.textContent = null;
            this.loading = true;
            this.animatedWithJs = cdk.isEdgeOlderThan(17, this.userAgent) || cdk.isIE(this.userAgent);
            this.isApple = cdk.isSafari(this.elementRef.nativeElement) || this.isIos;
        }
        Object.defineProperty(TuiLoaderComponent.prototype, "showLoader", {
            set: function (value) {
                // @bad TODO: https://github.com/angular/angular/issues/32083 think of a better way
                if (value && this.focused) {
                    cdk.blurNativeFocused(this.documentRef);
                }
                this.loading = value;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiLoaderComponent.prototype, "hasOverlay", {
            get: function () {
                return this.overlay && this.loading;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiLoaderComponent.prototype, "hasText", {
            get: function () {
                return !!this.textContent;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiLoaderComponent.prototype, "isHorizontal", {
            get: function () {
                return !miscellaneous.sizeBigger(this.size);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiLoaderComponent.prototype, "focused", {
            get: function () {
                return cdk.isNativeFocusedIn(this.elementRef.nativeElement);
            },
            enumerable: true,
            configurable: true
        });
        TuiLoaderComponent.ctorParameters = function () { return [
            { type: Document, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] },
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: String, decorators: [{ type: core.Inject, args: [common$1.USER_AGENT,] }] },
            { type: Boolean, decorators: [{ type: core.Inject, args: [cdk.TUI_IS_IOS,] }] }
        ]; };
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiLoaderComponent.prototype, "size", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiLoaderComponent.prototype, "inheritColor", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiLoaderComponent.prototype, "overlay", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiLoaderComponent.prototype, "textContent", void 0);
        __decorate([
            core.Input(),
            cdk.tuiRequiredSetter()
        ], TuiLoaderComponent.prototype, "showLoader", null);
        __decorate([
            core.HostBinding('class._loading')
        ], TuiLoaderComponent.prototype, "loading", void 0);
        __decorate([
            core.HostBinding('class._animated-with-js')
        ], TuiLoaderComponent.prototype, "animatedWithJs", void 0);
        TuiLoaderComponent = __decorate([
            core.Component({
                selector: 'tui-loader',
                template: "<fieldset\n    class=\"content\"\n    [class.content_has-overlay]=\"hasOverlay\"\n    [class.content_loading]=\"loading\"\n    [disabled]=\"loading && !isApple\"\n>\n    <ng-content></ng-content>\n</fieldset>\n\n<div\n    *ngIf=\"loading\"\n    class=\"loader\"\n    [attr.data-tui-element-size]=\"size\"\n    [class.loader_horizontal]=\"isHorizontal\"\n    [class.loader_inherit-color]=\"inheritColor\"\n>\n    <svg\n        automation-id=\"tui-loader__loader\"\n        class=\"icon\"\n        focusable=\"false\"\n        viewBox=\"0 0 100 100\"\n        xmlns=\"http://www.w3.org/2000/svg\"\n    >\n        <circle\n            class=\"circle\"\n            cx=\"50\"\n            cy=\"50\"\n            r=\"50\"\n            stroke-dasharray=\"314\"\n        ></circle>\n    </svg>\n\n    <div\n        polymorpheus-outlet\n        *ngIf=\"hasText\"\n        automation-id=\"tui-loader__text\"\n        class=\"text\"\n        [class.text_horizontal]=\"isHorizontal\"\n        [content]=\"textContent\"\n    ></div>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: ["@-webkit-keyframes tuiLoaderDashOffset{0%,100%{stroke-dashoffset:-66}50%{stroke-dashoffset:-314}}@keyframes tuiLoaderDashOffset{0%,100%{stroke-dashoffset:-66}50%{stroke-dashoffset:-314}}@-webkit-keyframes tuiLoaderRotate{0%{transform:rotate(0)}50%{transform:rotate(1turn)}100%{transform:rotate(3turn)}}@keyframes tuiLoaderRotate{0%{transform:rotate(0)}50%{transform:rotate(1turn)}100%{transform:rotate(3turn)}}:host{position:relative;display:flex}:host._loading{overflow:hidden}.content{z-index:0;min-width:100%;height:100%;padding:0;margin:0;border:none}.content_has-overlay{opacity:.3}.content_loading{pointer-events:none}.loader{position:relative;left:-100%;display:flex;flex-direction:column;align-items:center;justify-content:center;min-width:100%;flex-shrink:0;-ms-grid-row-align:center;align-self:center;color:var(--tui-text-01);stroke:var(--tui-primary);-webkit-animation:.3s tuiFadeIn;animation:.3s tuiFadeIn}.loader.loader_horizontal{flex-direction:row}.loader.loader_inherit-color{color:inherit;stroke:currentColor}.loader[data-tui-element-size=xs]{font-size:12px;stroke-width:38}.loader[data-tui-element-size='s']{font-size:16px;stroke-width:25}.loader[data-tui-element-size='m']{font-size:24px;stroke-width:17}.loader[data-tui-element-size='l']{font-size:40px;stroke-width:15}.loader[data-tui-element-size=xl]{font-size:56px;stroke-width:14}.loader[data-tui-element-size=xxl]{font-size:80px;stroke-width:10}.text{font:var(--tui-font-text-s);color:var(--tui-text-01);white-space:nowrap;overflow:hidden;text-overflow:ellipsis;margin-top:16px;color:inherit;max-width:100%;text-align:center;stroke-width:0}.text_horizontal{margin:0 0 0 16px}.icon{display:block;min-width:1em;max-width:1em;min-height:1em;max-height:1em;margin:4px -.5em;border-radius:100%;overflow:hidden;transform:scale(1,-1)}:host:not(._animated-with-js) .icon{-webkit-animation:3s linear infinite tuiLoaderRotate;animation:3s linear infinite tuiLoaderRotate}.circle{fill:none;stroke:inherit;stroke-width:inherit;-webkit-animation:3s linear infinite tuiLoaderDashOffset;animation:3s linear infinite tuiLoaderDashOffset}"]
            }),
            __param(0, core.Inject(common.DOCUMENT)),
            __param(1, core.Inject(core.ElementRef)),
            __param(2, core.Inject(common$1.USER_AGENT)),
            __param(3, core.Inject(cdk.TUI_IS_IOS))
        ], TuiLoaderComponent);
        return TuiLoaderComponent;
    }());

    var PERIMETER = 314;
    var COEFFICIENT = 1.5;
    var TuiLoaderDirective = /** @class */ (function () {
        function TuiLoaderDirective(animationFrame$, ngZone, destroy$, elementRef, userAgent) {
            var _this = this;
            this.elementRef = elementRef;
            this.strokeDasharray = PERIMETER;
            this.strokeDashoffset = 0;
            if (!cdk.isEdgeOlderThan(17, userAgent) && !cdk.isIE(userAgent)) {
                return;
            }
            animationFrame$
                .pipe(cdk.tuiZonefree(ngZone), operators.pairwise(), operators.takeUntil(destroy$))
                .subscribe(function (_a) {
                var _b = __read(_a, 2), cur = _b[0], prev = _b[1];
                _this.animate(prev - cur);
            });
        }
        TuiLoaderDirective.prototype.animate = function (delta) {
            if (this.strokeDasharray < 0) {
                this.strokeDasharray = PERIMETER * 2;
                this.strokeDashoffset = 0;
            }
            var strokeDasharray = Math.abs(this.strokeDasharray - PERIMETER);
            var fps = 1000 / delta;
            var offsetStep = 1 + Math.floor(this.strokeDasharray / PERIMETER);
            var style = this.elementRef.nativeElement.style;
            style.strokeDashoffset = this.strokeDashoffset.toString();
            style.strokeDasharray = PERIMETER - strokeDasharray + ' ' + strokeDasharray;
            this.strokeDasharray -= PERIMETER / COEFFICIENT / fps;
            this.strokeDashoffset += (offsetStep * PERIMETER) / COEFFICIENT / fps;
            // this.elementRef.nativeElement.parentElement!.style.animation = 'none';
        };
        TuiLoaderDirective.ctorParameters = function () { return [
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [common$1.ANIMATION_FRAME,] }] },
            { type: core.NgZone, decorators: [{ type: core.Inject, args: [core.NgZone,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [cdk.TuiDestroyService,] }] },
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: String, decorators: [{ type: core.Inject, args: [common$1.USER_AGENT,] }] }
        ]; };
        TuiLoaderDirective = __decorate([
            core.Directive({
                selector: 'circle',
                providers: [cdk.TuiDestroyService],
            }),
            __param(0, core.Inject(common$1.ANIMATION_FRAME)),
            __param(1, core.Inject(core.NgZone)),
            __param(2, core.Inject(cdk.TuiDestroyService)),
            __param(3, core.Inject(core.ElementRef)),
            __param(4, core.Inject(common$1.USER_AGENT))
        ], TuiLoaderDirective);
        return TuiLoaderDirective;
    }());

    var TuiLoaderModule = /** @class */ (function () {
        function TuiLoaderModule() {
        }
        TuiLoaderModule = __decorate([
            core.NgModule({
                imports: [common.CommonModule, ngPolymorpheus.PolymorpheusModule],
                declarations: [TuiLoaderComponent, TuiLoaderDirective],
                exports: [TuiLoaderComponent],
            })
        ], TuiLoaderModule);
        return TuiLoaderModule;
    }());

    exports.TuiLoaderComponent = TuiLoaderComponent;
    exports.TuiLoaderDirective = TuiLoaderDirective;
    exports.TuiLoaderModule = TuiLoaderModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
