// @taiga-ui/core/bundles/taiga-ui-core-modules-notifications.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('@taiga-ui/core/components/notification'), require('@tinkoff/ng-polymorpheus'), require('@taiga-ui/core/animations')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/modules/notifications', ['exports', '@angular/core', '@taiga-ui/cdk', 'rxjs', 'rxjs/operators', '@angular/common', '@taiga-ui/core/components/notification', '@tinkoff/ng-polymorpheus', '@taiga-ui/core/animations'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.modules = global['taiga-ui'].core.modules || {}, global['taiga-ui'].core.modules.notifications = {}), global.ng.core, global.cdk, global.rxjs, global.rxjs.operators, global.ng.common, global['taiga-ui'].core.components.notification, global.ngPolymorpheus, global['taiga-ui'].core.animations));
}(this, (function (exports, core, cdk, rxjs, operators, common, notification, ngPolymorpheus, animations) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var ALERT_AUTOCLOSE_TIMEOUT = 3000;
    var TuiNotificationAlertComponent = /** @class */ (function () {
        function TuiNotificationAlertComponent(destroy$) {
            var _this = this;
            rxjs.timer(ALERT_AUTOCLOSE_TIMEOUT)
                .pipe(operators.takeUntil(destroy$))
                .subscribe(function () {
                if (_this.safeItem.autoClose) {
                    _this.closeDialog();
                }
            });
        }
        Object.defineProperty(TuiNotificationAlertComponent.prototype, "safeItem", {
            get: function () {
                if (!this.item) {
                    throw new Error('Notification was created as undefined');
                }
                return this.item;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiNotificationAlertComponent.prototype, "context", {
            get: function () {
                return this.calculateContext(this.safeItem);
            },
            enumerable: true,
            configurable: true
        });
        TuiNotificationAlertComponent.prototype.closeDialog = function () {
            this.safeItem.observer.complete();
        };
        TuiNotificationAlertComponent.prototype.calculateContext = function (_a) {
            var status = _a.status, data = _a.data, label = _a.label, observer = _a.observer;
            return {
                $implicit: status,
                data: data,
                label: label,
                closeHook: function () {
                    observer.complete();
                },
                emitHook: function (data) {
                    observer.next(data);
                },
                emitAndCloseHook: function (data) {
                    observer.next(data);
                    observer.complete();
                },
            };
        };
        TuiNotificationAlertComponent.ctorParameters = function () { return [
            { type: cdk.TuiDestroyService, decorators: [{ type: core.Inject, args: [cdk.TuiDestroyService,] }] }
        ]; };
        __decorate([
            core.Input()
        ], TuiNotificationAlertComponent.prototype, "item", void 0);
        __decorate([
            cdk.tuiPure
        ], TuiNotificationAlertComponent.prototype, "calculateContext", null);
        TuiNotificationAlertComponent = __decorate([
            core.Component({
                selector: 'tui-notification-alert',
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                template: "<tui-notification\n    *ngIf=\"safeItem.hasCloseButton else noClose\"\n    [status]=\"safeItem.status\"\n    [hasIcon]=\"safeItem.hasIcon\"\n    (close)=\"closeDialog()\"\n>\n    <label\n        *ngIf=\"safeItem.label\"\n        automation-id=\"tui-notification-alert__heading\"\n        class=\"heading\"\n    >\n        {{safeItem.label}}\n    </label>\n    <div\n        polymorpheus-outlet\n        automation-id=\"tui-notification-alert__content\"\n        class=\"content\"\n        [content]=\"safeItem.content\"\n        [context]=\"context\"\n    ></div>\n</tui-notification>\n<ng-template #noClose>\n    <tui-notification [status]=\"safeItem.status\" [hasIcon]=\"safeItem.hasIcon\">\n        <label\n            *ngIf=\"safeItem.label\"\n            automation-id=\"tui-notification-alert__heading\"\n            class=\"heading\"\n        >\n            {{safeItem.label}}\n        </label>\n        <div\n            polymorpheus-outlet\n            automation-id=\"tui-notification-alert__content\"\n            class=\"content\"\n            [content]=\"safeItem.content\"\n            [context]=\"context\"\n        ></div>\n    </tui-notification>\n</ng-template>\n",
                providers: [cdk.TuiDestroyService],
                styles: [":host{box-shadow:0 8px 16px rgba(51,51,51,.2);display:block;border-radius:var(--tui-radius-m)}.heading{font:var(--tui-font-text-m);line-height:20px;font-weight:700;margin:0}.content{color:var(--tui-text-01);word-wrap:break-word}.content:empty{display:none}"]
            }),
            __param(0, core.Inject(cdk.TuiDestroyService))
        ], TuiNotificationAlertComponent);
        return TuiNotificationAlertComponent;
    }());

    var TuiNotificationAlertModule = /** @class */ (function () {
        function TuiNotificationAlertModule() {
        }
        TuiNotificationAlertModule = __decorate([
            core.NgModule({
                imports: [common.CommonModule, ngPolymorpheus.PolymorpheusModule, notification.TuiNotificationModule],
                declarations: [TuiNotificationAlertComponent],
                exports: [TuiNotificationAlertComponent],
            })
        ], TuiNotificationAlertModule);
        return TuiNotificationAlertModule;
    }());

    var NotificationAlert = /** @class */ (function () {
        function NotificationAlert(observer, content, options) {
            var _a = options.label, label = _a === void 0 ? '' : _a, _b = options.status, status = _b === void 0 ? "info" /* Info */ : _b, _c = options.hasIcon, hasIcon = _c === void 0 ? true : _c, _d = options.autoClose, autoClose = _d === void 0 ? true : _d, _e = options.hasCloseButton, hasCloseButton = _e === void 0 ? true : _e;
            this.observer = observer;
            this.content = content;
            this.label = label;
            this.status = status;
            this.hasIcon = hasIcon;
            this.autoClose = autoClose;
            this.hasCloseButton = hasCloseButton;
            if (options && this.withData(options)) {
                this.data = options.data;
            }
        }
        NotificationAlert.prototype.withData = function (options) {
            return 'data' in options;
        };
        return NotificationAlert;
    }());

    var NO_HOST = 'Notifications are disabled, enable support by adding TuiNotificationsModule to your main app module';
    var TuiNotificationsService = /** @class */ (function () {
        function TuiNotificationsService() {
            /** @internal */
            this.items$ = new rxjs.BehaviorSubject([]);
        }
        TuiNotificationsService.prototype.show = function (content, options) {
            var _this = this;
            if (options === void 0) { options = {}; }
            cdk.tuiAssert.assert(!!this.items$.observers.length, NO_HOST);
            return new rxjs.Observable(function (observer) {
                var notification = new NotificationAlert(observer, content, options);
                _this.items$.next(__spread(_this.items$.value, [notification]));
                return function () {
                    _this.items$.next(_this.items$.value.filter(function (item) { return item !== notification; }));
                };
            });
        };
        TuiNotificationsService.ɵprov = core.ɵɵdefineInjectable({ factory: function TuiNotificationsService_Factory() { return new TuiNotificationsService(); }, token: TuiNotificationsService, providedIn: "root" });
        TuiNotificationsService = __decorate([
            core.Injectable({
                providedIn: 'root',
            })
        ], TuiNotificationsService);
        return TuiNotificationsService;
    }());

    var TuiNotificationsHostComponent = /** @class */ (function () {
        function TuiNotificationsHostComponent(service) {
            this.service = service;
        }
        TuiNotificationsHostComponent.ctorParameters = function () { return [
            { type: TuiNotificationsService, decorators: [{ type: core.Inject, args: [TuiNotificationsService,] }] }
        ]; };
        TuiNotificationsHostComponent = __decorate([
            core.Component({
                selector: 'tui-notifications-host',
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                template: "<ng-container *tuiLet=\"service.items$ | async as items\">\n    <div\n        *ngIf=\"items.length\"\n        class=\"wrapper\"\n        [@tuiFadeInList]=\"items.length\"\n        [@tuiHeightCollapseList]=\"items.length\"\n        [@tuiSlideInRightList]=\"items.length\"\n    >\n        <tui-notification-alert\n            *ngFor=\"let item of items\"\n            class=\"alert\"\n            role=\"alert\"\n            [item]=\"item\"\n        ></tui-notification-alert>\n    </div>\n</ng-container>\n",
                animations: [animations.tuiFadeInList, animations.tuiSlideInRightList, animations.tuiHeightCollapseList],
                styles: [":host{position:fixed;top:0;left:0;width:100%;height:0}.wrapper{width:288px;margin-left:auto;padding:32px 48px 0 0}@media screen and (max-width:1023px){.wrapper{padding:16px 16px 0 0}}@media screen and (min-width:1360px){.wrapper{width:368px}}.alert:not(:first-child){margin-top:12px}"]
            }),
            __param(0, core.Inject(TuiNotificationsService))
        ], TuiNotificationsHostComponent);
        return TuiNotificationsHostComponent;
    }());

    var ɵ0 = TuiNotificationsHostComponent;
    var TuiNotificationsModule = /** @class */ (function () {
        function TuiNotificationsModule() {
        }
        TuiNotificationsModule = __decorate([
            core.NgModule({
                imports: [common.CommonModule, TuiNotificationAlertModule, cdk.TuiLetModule],
                declarations: [TuiNotificationsHostComponent],
                exports: [TuiNotificationsHostComponent],
                entryComponents: [TuiNotificationsHostComponent],
                providers: [
                    {
                        provide: TuiNotificationsHostComponent,
                        useValue: ɵ0,
                    },
                ],
            })
        ], TuiNotificationsModule);
        return TuiNotificationsModule;
    }());

    exports.ALERT_AUTOCLOSE_TIMEOUT = ALERT_AUTOCLOSE_TIMEOUT;
    exports.NotificationAlert = NotificationAlert;
    exports.TuiNotificationAlertComponent = TuiNotificationAlertComponent;
    exports.TuiNotificationAlertModule = TuiNotificationAlertModule;
    exports.TuiNotificationsHostComponent = TuiNotificationsHostComponent;
    exports.TuiNotificationsModule = TuiNotificationsModule;
    exports.TuiNotificationsService = TuiNotificationsService;
    exports.ɵ0 = ɵ0;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
