// @taiga-ui/core/bundles/taiga-ui-core-services.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs'), require('@ng-web-apis/common'), require('rxjs/operators'), require('@angular/router'), require('@taiga-ui/cdk'), require('@angular/platform-browser'), require('@taiga-ui/core/tokens'), require('@taiga-ui/core/utils/dom')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/services', ['exports', '@angular/core', 'rxjs', '@ng-web-apis/common', 'rxjs/operators', '@angular/router', '@taiga-ui/cdk', '@angular/platform-browser', '@taiga-ui/core/tokens', '@taiga-ui/core/utils/dom'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.services = {}), global.ng.core, global.rxjs, global.common, global.rxjs.operators, global.ng.router, global.cdk, global.ng.platformBrowser, global['taiga-ui'].core.tokens, global['taiga-ui'].core.utils.dom));
}(this, (function (exports, core, rxjs, common, operators, router, cdk, platformBrowser, tokens, dom) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    /**
     * Service for displaying hints/tooltips
     */
    var TuiHintService = /** @class */ (function (_super) {
        __extends(TuiHintService, _super);
        function TuiHintService() {
            var _this = _super.call(this, function (observer) { return _this.hints$.subscribe(observer); }) || this;
            _this.hints$ = new rxjs.BehaviorSubject([]);
            /**
             * TODO: v2.0
             * We need the following logic for desribedBy
             * move it into another service that can register hints and
             * manage it using TuiHintService inside
             */
            _this.directives = [];
            return _this;
        }
        TuiHintService.prototype.add = function (directive) {
            this.hints$.next(__spread(this.hints$.value, [directive]));
        };
        TuiHintService.prototype.remove = function (directive) {
            this.hints$.next(this.hints$.value.filter(function (hint) { return hint !== directive; }));
        };
        TuiHintService.prototype.register = function (directive) {
            this.directives = __spread(this.directives, [directive]);
        };
        TuiHintService.prototype.unregister = function (directive) {
            this.remove(directive);
            this.directives = this.directives.filter(function (dir) { return dir !== directive; });
        };
        TuiHintService.prototype.showHintForId = function (id) {
            var directive = this.findDirectiveWithHintId(id);
            if (directive) {
                this.add(directive);
            }
        };
        TuiHintService.prototype.hideHintForId = function (id) {
            var directive = this.findDirectiveWithHintId(id);
            if (directive) {
                this.remove(directive);
            }
        };
        TuiHintService.prototype.findDirectiveWithHintId = function (id) {
            return this.directives.find(function (directive) { return directive.tuiHintId === id; });
        };
        TuiHintService.ɵprov = core.ɵɵdefineInjectable({ factory: function TuiHintService_Factory() { return new TuiHintService(); }, token: TuiHintService, providedIn: "root" });
        TuiHintService = __decorate([
            core.Injectable({
                providedIn: 'root',
            })
        ], TuiHintService);
        return TuiHintService;
    }(rxjs.Observable));

    // @dynamic
    var TuiNightThemeService = /** @class */ (function (_super) {
        __extends(TuiNightThemeService, _super);
        function TuiNightThemeService(windowRef) {
            var _this = this;
            var media = windowRef.matchMedia('(prefers-color-scheme: dark)');
            var media$ = rxjs.fromEvent(media, 'change').pipe(operators.startWith(null), operators.map(function () { return media.matches; }), operators.share());
            _this = _super.call(this, function (subscriber) { return media$.subscribe(subscriber); }) || this;
            return _this;
        }
        TuiNightThemeService.ctorParameters = function () { return [
            { type: Window, decorators: [{ type: core.Inject, args: [common.WINDOW,] }] }
        ]; };
        TuiNightThemeService.ɵprov = core.ɵɵdefineInjectable({ factory: function TuiNightThemeService_Factory() { return new TuiNightThemeService(core.ɵɵinject(common.WINDOW)); }, token: TuiNightThemeService, providedIn: "root" });
        TuiNightThemeService = __decorate([
            core.Injectable({
                providedIn: 'root',
            }),
            __param(0, core.Inject(common.WINDOW))
        ], TuiNightThemeService);
        return TuiNightThemeService;
    }(rxjs.Observable));

    var TuiRouterLinkActiveService = /** @class */ (function (_super) {
        __extends(TuiRouterLinkActiveService, _super);
        function TuiRouterLinkActiveService(routerLinkActive, ngZone, animationFrame$, destroy$) {
            var _this = this;
            var stream$ = routerLinkActive
                ? animationFrame$.pipe(operators.map(function () { return routerLinkActive.isActive; }), operators.distinctUntilChanged(), cdk.tuiZoneOptimized(ngZone), operators.takeUntil(destroy$))
                : rxjs.EMPTY;
            _this = _super.call(this, function (subscriber) { return stream$.subscribe(subscriber); }) || this;
            return _this;
        }
        TuiRouterLinkActiveService.ctorParameters = function () { return [
            { type: router.RouterLinkActive, decorators: [{ type: core.Optional }, { type: core.Inject, args: [router.RouterLinkActive,] }] },
            { type: core.NgZone, decorators: [{ type: core.Inject, args: [core.NgZone,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [common.ANIMATION_FRAME,] }] },
            { type: cdk.TuiDestroyService, decorators: [{ type: core.Inject, args: [cdk.TuiDestroyService,] }] }
        ]; };
        TuiRouterLinkActiveService = __decorate([
            core.Injectable(),
            __param(0, core.Optional()),
            __param(0, core.Inject(router.RouterLinkActive)),
            __param(1, core.Inject(core.NgZone)),
            __param(2, core.Inject(common.ANIMATION_FRAME)),
            __param(3, core.Inject(cdk.TuiDestroyService))
        ], TuiRouterLinkActiveService);
        return TuiRouterLinkActiveService;
    }(rxjs.Observable));

    /**
     * Service for reusing SVGs without inlining each instance
     */
    // @dynamic
    var TuiSvgService = /** @class */ (function () {
        function TuiSvgService(tuiSanitizer, sanitizer, icons) {
            this.tuiSanitizer = tuiSanitizer;
            this.sanitizer = sanitizer;
            this.items$ = new rxjs.BehaviorSubject(new Map());
            this.originals = {};
            this.define(icons);
        }
        TuiSvgService.prototype.define = function (icons) {
            var _this = this;
            var value = this.items$.value;
            Object.keys(icons).forEach(function (key) {
                _this.defineIcon(key, icons[key], value);
            });
            this.items$.next(value);
        };
        TuiSvgService.prototype.getOriginal = function (name) {
            return this.originals[name] || null;
        };
        TuiSvgService.prototype.defineIcon = function (name, src, map) {
            var _a;
            if (map.has(name)) {
                return;
            }
            var parsed = this.parseSrc(name, src);
            if (!parsed) {
                cdk.tuiAssert.assert(false, 'Unable to parse given SVG src');
                return;
            }
            map.set(name, parsed);
            this.originals = __assign(__assign({}, this.originals), (_a = {}, _a[name] = src, _a));
        };
        TuiSvgService.prototype.parseSrc = function (name, src) {
            return this.sanitize(dom.processIcon(src, name));
        };
        TuiSvgService.prototype.sanitize = function (src) {
            return this.sanitizer.bypassSecurityTrustHtml((this.tuiSanitizer
                ? this.tuiSanitizer.sanitize(core.SecurityContext.HTML, src)
                : this.sanitizer.sanitize(core.SecurityContext.HTML, src)) || '');
        };
        TuiSvgService.ctorParameters = function () { return [
            { type: core.Sanitizer, decorators: [{ type: core.Optional }, { type: core.Inject, args: [tokens.TUI_SANITIZER,] }] },
            { type: platformBrowser.DomSanitizer, decorators: [{ type: core.Inject, args: [platformBrowser.DomSanitizer,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [tokens.TUI_ICONS,] }] }
        ]; };
        TuiSvgService.ɵprov = core.ɵɵdefineInjectable({ factory: function TuiSvgService_Factory() { return new TuiSvgService(core.ɵɵinject(tokens.TUI_SANITIZER, 8), core.ɵɵinject(platformBrowser.DomSanitizer), core.ɵɵinject(tokens.TUI_ICONS)); }, token: TuiSvgService, providedIn: "root" });
        TuiSvgService = __decorate([
            core.Injectable({
                providedIn: 'root',
            }),
            __param(0, core.Optional()),
            __param(0, core.Inject(tokens.TUI_SANITIZER)),
            __param(1, core.Inject(platformBrowser.DomSanitizer)),
            __param(2, core.Inject(tokens.TUI_ICONS))
        ], TuiSvgService);
        return TuiSvgService;
    }());

    exports.TuiHintService = TuiHintService;
    exports.TuiNightThemeService = TuiNightThemeService;
    exports.TuiRouterLinkActiveService = TuiRouterLinkActiveService;
    exports.TuiSvgService = TuiSvgService;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
