// @taiga-ui/core/bundles/taiga-ui-core-components-scroll-controls.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('@ng-web-apis/common'), require('@taiga-ui/cdk'), require('@taiga-ui/core/tokens'), require('rxjs'), require('rxjs/operators'), require('@taiga-ui/core/animations'), require('@taiga-ui/core/providers')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/scroll-controls', ['exports', '@angular/common', '@angular/core', '@ng-web-apis/common', '@taiga-ui/cdk', '@taiga-ui/core/tokens', 'rxjs', 'rxjs/operators', '@taiga-ui/core/animations', '@taiga-ui/core/providers'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.components = global['taiga-ui'].core.components || {}, global['taiga-ui'].core.components['scroll-controls'] = {}), global.ng.common, global.ng.core, global.common$1, global.cdk, global['taiga-ui'].core.tokens, global.rxjs, global.rxjs.operators, global['taiga-ui'].core.animations, global['taiga-ui'].core.providers));
}(this, (function (exports, common, core, common$1, cdk, tokens, rxjs, operators, animations, providers) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var MIN_WIDTH = 24;
    // @bad TODO: add support for window scroll control
    // @dynamic
    var TuiScrollbarDirective = /** @class */ (function () {
        function TuiScrollbarDirective(ngZone, renderer, destroy$, container, documentRef, windowRef, elementRef, viewportScroller) {
            var _this = this;
            this.container = container;
            this.documentRef = documentRef;
            this.windowRef = windowRef;
            this.elementRef = elementRef;
            this.viewportScroller = viewportScroller;
            this.tuiScrollbar = "vertical" /* Vertical */;
            var nativeElement = this.elementRef.nativeElement;
            var mousedown$ = cdk.typedFromEvent(nativeElement, 'mousedown');
            var mousemove$ = cdk.typedFromEvent(this.documentRef, 'mousemove');
            var mouseup$ = cdk.typedFromEvent(this.documentRef, 'mouseup');
            mousedown$
                .pipe(cdk.preventDefault(), operators.switchMap(function (event) {
                var rect = event.currentTarget.getBoundingClientRect();
                var vertical = getOffsetVertical(event, rect);
                var horizontal = getOffsetHorizontal(event, rect);
                return mousemove$.pipe(operators.map(function (event) { return _this.getScrolled(event, vertical, horizontal); }), operators.takeUntil(mouseup$));
            }), operators.takeUntil(destroy$), cdk.tuiZonefree(ngZone))
                .subscribe(function (_a) {
                var _b = __read(_a, 2), scrollTop = _b[0], scrollLeft = _b[1];
                var _c = __read(_this.viewportScroller.getScrollPosition(), 2), x = _c[0], y = _c[1];
                if (!_this.container) {
                    _this.viewportScroller.scrollToPosition([
                        _this.tuiScrollbar === "vertical" /* Vertical */ ? x : scrollLeft,
                        _this.tuiScrollbar === "vertical" /* Vertical */ ? scrollTop : y,
                    ]);
                    return;
                }
                if (_this.tuiScrollbar === "vertical" /* Vertical */) {
                    renderer.setProperty(_this.container.nativeElement, 'scrollTop', scrollTop);
                }
                else {
                    renderer.setProperty(_this.container.nativeElement, 'scrollLeft', scrollLeft);
                }
            });
            rxjs.merge(rxjs.fromEvent(this.container ? this.container.nativeElement : this.windowRef, 'scroll'), rxjs.interval(cdk.POLLING_TIME))
                .pipe(operators.takeUntil(destroy$), cdk.tuiZonefree(ngZone))
                .subscribe(function () {
                if (_this.tuiScrollbar === "vertical" /* Vertical */) {
                    renderer.setStyle(nativeElement, 'top', _this.thumb * 100 + "%");
                    renderer.setStyle(nativeElement, 'height', _this.view * 100 + "%");
                }
                else {
                    renderer.setStyle(nativeElement, 'left', _this.thumb * 100 + "%");
                    renderer.setStyle(nativeElement, 'width', _this.view * 100 + "%");
                }
            });
        }
        Object.defineProperty(TuiScrollbarDirective.prototype, "scrolled", {
            get: function () {
                var _a = this.computedContainer, scrollTop = _a.scrollTop, scrollHeight = _a.scrollHeight, clientHeight = _a.clientHeight, scrollLeft = _a.scrollLeft, scrollWidth = _a.scrollWidth, clientWidth = _a.clientWidth;
                return this.tuiScrollbar === "vertical" /* Vertical */
                    ? scrollTop / (scrollHeight - clientHeight)
                    : scrollLeft / (scrollWidth - clientWidth);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiScrollbarDirective.prototype, "compensation", {
            get: function () {
                var _a = this.computedContainer, clientHeight = _a.clientHeight, scrollHeight = _a.scrollHeight, clientWidth = _a.clientWidth, scrollWidth = _a.scrollWidth;
                if (((clientHeight * clientHeight) / scrollHeight > MIN_WIDTH &&
                    this.tuiScrollbar === "vertical" /* Vertical */) ||
                    ((clientWidth * clientWidth) / scrollWidth > MIN_WIDTH &&
                        this.tuiScrollbar === "horizontal" /* Horizontal */)) {
                    return 0;
                }
                return this.tuiScrollbar === "vertical" /* Vertical */
                    ? MIN_WIDTH / clientHeight
                    : MIN_WIDTH / clientWidth;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiScrollbarDirective.prototype, "thumb", {
            get: function () {
                var compensation = this.compensation || this.view;
                return this.scrolled * (1 - compensation);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiScrollbarDirective.prototype, "view", {
            get: function () {
                var _a = this.computedContainer, clientHeight = _a.clientHeight, scrollHeight = _a.scrollHeight, clientWidth = _a.clientWidth, scrollWidth = _a.scrollWidth;
                return this.tuiScrollbar === "vertical" /* Vertical */
                    ? Math.ceil((clientHeight / scrollHeight) * 100) / 100
                    : Math.ceil((clientWidth / scrollWidth) * 100) / 100;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiScrollbarDirective.prototype, "computedContainer", {
            get: function () {
                return this.container
                    ? this.container.nativeElement
                    : this.documentRef.documentElement;
            },
            enumerable: true,
            configurable: true
        });
        TuiScrollbarDirective.prototype.getScrolled = function (_a, offsetVertical, offsetHorizontal) {
            var clientY = _a.clientY, clientX = _a.clientX;
            var _b = this.windowRef, innerWidth = _b.innerWidth, innerHeight = _b.innerHeight;
            var _c = this.elementRef.nativeElement, offsetHeight = _c.offsetHeight, offsetWidth = _c.offsetWidth;
            var _d = this
                .container
                ? this.container.nativeElement.getBoundingClientRect()
                : {}, _e = _d.top, top = _e === void 0 ? 0 : _e, _f = _d.left, left = _f === void 0 ? 0 : _f, _g = _d.width, width = _g === void 0 ? innerWidth : _g, _h = _d.height, height = _h === void 0 ? innerHeight : _h;
            var maxTop = this.computedContainer.scrollHeight - height;
            var maxLeft = this.computedContainer.scrollWidth - width;
            var scrolledTop = (clientY - top - offsetHeight * offsetVertical) / (height - offsetHeight);
            var scrolledLeft = (clientX - left - offsetWidth * offsetHorizontal) / (width - offsetWidth);
            return [maxTop * scrolledTop, maxLeft * scrolledLeft];
        };
        TuiScrollbarDirective.ctorParameters = function () { return [
            { type: core.NgZone, decorators: [{ type: core.Inject, args: [core.NgZone,] }] },
            { type: core.Renderer2, decorators: [{ type: core.Inject, args: [core.Renderer2,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [cdk.TuiDestroyService,] }] },
            { type: core.ElementRef, decorators: [{ type: core.Optional }, { type: core.Inject, args: [tokens.TUI_SCROLL_REF,] }] },
            { type: Document, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] },
            { type: Window, decorators: [{ type: core.Inject, args: [common$1.WINDOW,] }] },
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: common.ViewportScroller, decorators: [{ type: core.Inject, args: [common.ViewportScroller,] }] }
        ]; };
        __decorate([
            core.Input()
        ], TuiScrollbarDirective.prototype, "tuiScrollbar", void 0);
        TuiScrollbarDirective = __decorate([
            core.Directive({
                selector: '[tuiScrollbar]',
                providers: [cdk.TuiDestroyService],
            }),
            __param(0, core.Inject(core.NgZone)),
            __param(1, core.Inject(core.Renderer2)),
            __param(2, core.Inject(cdk.TuiDestroyService)),
            __param(3, core.Optional()),
            __param(3, core.Inject(tokens.TUI_SCROLL_REF)),
            __param(4, core.Inject(common.DOCUMENT)),
            __param(5, core.Inject(common$1.WINDOW)),
            __param(6, core.Inject(core.ElementRef)),
            __param(7, core.Inject(common.ViewportScroller))
        ], TuiScrollbarDirective);
        return TuiScrollbarDirective;
    }());
    function getOffsetVertical(_a, _b) {
        var clientY = _a.clientY;
        var top = _b.top, height = _b.height;
        return (clientY - top) / height;
    }
    function getOffsetHorizontal(_a, _b) {
        var clientX = _a.clientX;
        var left = _b.left, width = _b.width;
        return (clientX - left) / width;
    }

    // @bad TODO: handle click on bar to scroll to that position
    // @dynamic
    var TuiScrollControlsComponent = /** @class */ (function () {
        function TuiScrollControlsComponent(ngZone, documentRef, scrollRef, mode$) {
            var _this = this;
            this.ngZone = ngZone;
            this.documentRef = documentRef;
            this.scrollRef = scrollRef;
            this.mode$ = mode$;
            this.refresh$ = rxjs.interval(300).pipe(operators.map(function () { return _this.scrollbars; }), operators.startWith([false, false]), operators.distinctUntilChanged(function (a, b) { return a[0] === b[0] && a[1] === b[1]; }), cdk.tuiZoneOptimized(this.ngZone));
        }
        Object.defineProperty(TuiScrollControlsComponent.prototype, "scrollbars", {
            get: function () {
                var _a = this.scrollRef
                    ? this.scrollRef.nativeElement
                    : this.documentRef.documentElement, clientHeight = _a.clientHeight, scrollHeight = _a.scrollHeight, clientWidth = _a.clientWidth, scrollWidth = _a.scrollWidth;
                return [
                    Math.ceil((clientHeight / scrollHeight) * 100) < 100,
                    Math.ceil((clientWidth / scrollWidth) * 100) < 100,
                ];
            },
            enumerable: true,
            configurable: true
        });
        TuiScrollControlsComponent.ctorParameters = function () { return [
            { type: core.NgZone, decorators: [{ type: core.Inject, args: [core.NgZone,] }] },
            { type: Document, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] },
            { type: core.ElementRef, decorators: [{ type: core.Optional }, { type: core.Inject, args: [tokens.TUI_SCROLL_REF,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_MODE,] }] }
        ]; };
        TuiScrollControlsComponent = __decorate([
            core.Component({
                selector: 'tui-scroll-controls',
                template: "<ng-container *tuiLet=\"refresh$ | async as bars\">\n    <div\n        *ngIf=\"bars[0]\"\n        class=\"bar bar_vertical\"\n        [class.bar_has-horizontal]=\"bars[1]\"\n        [@tuiFadeIn]\n    >\n        <div tuiScrollbar=\"vertical\" class=\"thumb\"></div>\n    </div>\n    <div\n        *ngIf=\"bars[1]\"\n        class=\"bar bar_horizontal\"\n        [class.bar_has-vertical]=\"bars[0]\"\n        [@tuiFadeIn]\n    >\n        <div tuiScrollbar=\"horizontal\" class=\"thumb\"></div>\n    </div>\n</ng-container>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                animations: [animations.tuiFadeIn],
                providers: [providers.MODE_PROVIDER],
                host: {
                    '($.data-mode.attr)': 'mode$',
                },
                styles: [":host{position:-webkit-sticky;position:sticky;top:0;left:0;z-index:100;min-width:calc(100% - 1px);min-height:calc(100% - 1px);float:left;margin-right:calc(-100% + 1px);pointer-events:none}.bar{position:absolute;right:0;bottom:0}.bar_vertical{top:0}.bar_horizontal{left:0}.bar_has-horizontal{bottom:8px}.bar_has-vertical{right:8px}.thumb{transition-duration:.15s;transition-timing-function:ease-in-out;position:absolute;border-radius:100px;border:4px solid transparent;cursor:pointer;pointer-events:auto;background:var(--tui-text-01);background-clip:content-box;box-sizing:border-box;transition-property:width,height;opacity:.2}.thumb:hover{opacity:.24}.thumb:active{opacity:.48}:host[data-mode=onDark] .thumb{background-color:var(--tui-text-01-night)}.bar_vertical .thumb{right:0;width:12px;min-height:24px}.bar_vertical .thumb:active,.bar_vertical:hover .thumb{width:14px}.bar_horizontal .thumb{bottom:0;height:12px;min-width:24px}.bar_horizontal .thumb:active,.bar_horizontal:hover .thumb{height:14px}"]
            }),
            __param(0, core.Inject(core.NgZone)),
            __param(1, core.Inject(common.DOCUMENT)),
            __param(2, core.Optional()),
            __param(2, core.Inject(tokens.TUI_SCROLL_REF)),
            __param(3, core.Inject(tokens.TUI_MODE))
        ], TuiScrollControlsComponent);
        return TuiScrollControlsComponent;
    }());

    var TuiScrollControlsModule = /** @class */ (function () {
        function TuiScrollControlsModule() {
        }
        TuiScrollControlsModule = __decorate([
            core.NgModule({
                imports: [common.CommonModule, cdk.TuiLetModule],
                declarations: [TuiScrollbarDirective, TuiScrollControlsComponent],
                exports: [TuiScrollControlsComponent],
            })
        ], TuiScrollControlsModule);
        return TuiScrollControlsModule;
    }());

    exports.TuiScrollControlsComponent = TuiScrollControlsComponent;
    exports.TuiScrollControlsModule = TuiScrollControlsModule;
    exports.TuiScrollbarDirective = TuiScrollbarDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
