// @taiga-ui/core/bundles/taiga-ui-core-components-primitive-calendar.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core/constants'), require('@taiga-ui/core/tokens'), require('rxjs'), require('@angular/common'), require('@taiga-ui/core/directives/color')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/core/components/primitive-calendar', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core/constants', '@taiga-ui/core/tokens', 'rxjs', '@angular/common', '@taiga-ui/core/directives/color'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].core = global['taiga-ui'].core || {}, global['taiga-ui'].core.components = global['taiga-ui'].core.components || {}, global['taiga-ui'].core.components['primitive-calendar'] = {}), global.ng.core, global.cdk, global['taiga-ui'].core.constants, global['taiga-ui'].core.tokens, global.rxjs, global.ng.common, global['taiga-ui'].core.directives.color));
}(this, (function (exports, core, cdk, constants, tokens, rxjs, common, color) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var ROWS_COUNT = 6;
    // @dynamic
    var TuiPrimitiveCalendarComponent = /** @class */ (function () {
        function TuiPrimitiveCalendarComponent(weekDays$) {
            var _this = this;
            this.weekDays$ = weekDays$;
            this.disabledItemHandler = cdk.ALWAYS_FALSE_HANDLER;
            this.markerHandler = constants.TUI_DEFAULT_MARKER_HANDLER;
            this.value = null;
            this.hoveredItem = null;
            this.showAdjacent = true;
            this.hoveredItemChange = new core.EventEmitter();
            this.dayClick = new core.EventEmitter();
            this.sheet = [];
            this.pressedItem = null;
            this.currentMonth = null;
            this.today = cdk.TuiDay.currentLocal();
            this.toMarkers = function (day, today, inRange) {
                if (today || inRange) {
                    return null;
                }
                var markers = _this.markerHandler(day);
                return markers.length === 0 ? null : markers;
            };
        }
        Object.defineProperty(TuiPrimitiveCalendarComponent.prototype, "month", {
            set: function (month) {
                if (this.currentMonth !== null && this.currentMonth.monthSame(month)) {
                    return;
                }
                var sheet = [];
                for (var rowIndex = 0; rowIndex < ROWS_COUNT; rowIndex++) {
                    var row = [];
                    for (var colIndex = 0; colIndex < cdk.DAYS_IN_WEEK; colIndex++) {
                        row.push(cdk.TuiDay.getDayFromMonthRowCol(month, rowIndex, colIndex));
                    }
                    sheet.push(row);
                }
                this.sheet = sheet;
                this.currentMonth = month;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPrimitiveCalendarComponent.prototype, "isSingle", {
            get: function () {
                return (this.value !== null &&
                    (this.value instanceof cdk.TuiDay || this.value.isSingleDay));
            },
            enumerable: true,
            configurable: true
        });
        TuiPrimitiveCalendarComponent.prototype.getItemState = function (item) {
            var _a = this, disabledItemHandler = _a.disabledItemHandler, pressedItem = _a.pressedItem, hoveredItem = _a.hoveredItem;
            if (disabledItemHandler(item)) {
                return "disabled" /* Disabled */;
            }
            if (pressedItem !== null && pressedItem.daySame(item)) {
                return "pressed" /* Pressed */;
            }
            if (hoveredItem !== null && hoveredItem.daySame(item)) {
                return "hovered" /* Hovered */;
            }
            return null;
        };
        TuiPrimitiveCalendarComponent.prototype.getItemRange = function (item) {
            var _a = this, value = _a.value, hoveredItem = _a.hoveredItem;
            if (value === null) {
                return null;
            }
            if (value instanceof cdk.TuiDay) {
                return value.daySame(item) ? "single" /* Single */ : null;
            }
            if ((value.from.daySame(item) && !value.isSingleDay) ||
                (hoveredItem !== null &&
                    hoveredItem.dayAfter(value.from) &&
                    value.from.daySame(item) &&
                    value.isSingleDay) ||
                (hoveredItem !== null &&
                    hoveredItem.daySame(item) &&
                    hoveredItem.dayBefore(value.from) &&
                    value.isSingleDay)) {
                return "start" /* Start */;
            }
            if ((value.to.daySame(item) && !value.isSingleDay) ||
                (hoveredItem !== null &&
                    hoveredItem.dayBefore(value.from) &&
                    value.from.daySame(item) &&
                    value.isSingleDay) ||
                (hoveredItem !== null &&
                    hoveredItem.daySame(item) &&
                    hoveredItem.dayAfter(value.from) &&
                    value.isSingleDay)) {
                return "end" /* End */;
            }
            return value.isSingleDay && value.from.daySame(item)
                ? "single" /* Single */
                : null;
        };
        TuiPrimitiveCalendarComponent.prototype.itemIsDisabled = function (day) {
            var disabledItemHandler = this.disabledItemHandler;
            return disabledItemHandler(day);
        };
        TuiPrimitiveCalendarComponent.prototype.itemIsToday = function (item) {
            return this.today.daySame(item);
        };
        TuiPrimitiveCalendarComponent.prototype.itemIsUnavailable = function (item) {
            return this.currentMonth === null || !this.currentMonth.monthSame(item);
        };
        TuiPrimitiveCalendarComponent.prototype.itemIsInterval = function (day) {
            var _a = this, value = _a.value, hoveredItem = _a.hoveredItem;
            if (value === null || value instanceof cdk.TuiDay) {
                return false;
            }
            if (!value.isSingleDay) {
                return value.from.daySameOrBefore(day) && value.to.dayAfter(day);
            }
            if (hoveredItem === null) {
                return false;
            }
            var range = cdk.TuiDayRange.sort(value.from, hoveredItem);
            return range.from.daySameOrBefore(day) && range.to.dayAfter(day);
        };
        TuiPrimitiveCalendarComponent.prototype.onItemHovered = function (hovered, item) {
            this.updateHoveredItem(hovered ? item : null);
        };
        TuiPrimitiveCalendarComponent.prototype.onItemPressed = function (pressed, item) {
            this.updatePressedItem(pressed, item);
        };
        TuiPrimitiveCalendarComponent.prototype.onItemClick = function (event, item) {
            if (this.disabledItemHandler(item)) {
                return;
            }
            event.preventDefault();
            this.dayClick.emit(item);
        };
        TuiPrimitiveCalendarComponent.prototype.updateHoveredItem = function (day) {
            if (cdk.nullableSame(this.hoveredItem, day, function (a, b) { return a.daySame(b); })) {
                return;
            }
            this.hoveredItem = day;
            this.hoveredItemChange.emit(day);
        };
        TuiPrimitiveCalendarComponent.prototype.updatePressedItem = function (pressed, item) {
            this.pressedItem = pressed ? item : null;
        };
        TuiPrimitiveCalendarComponent.ctorParameters = function () { return [
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_SHORT_WEEK_DAYS,] }] }
        ]; };
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPrimitiveCalendarComponent.prototype, "disabledItemHandler", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPrimitiveCalendarComponent.prototype, "markerHandler", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPrimitiveCalendarComponent.prototype, "value", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPrimitiveCalendarComponent.prototype, "hoveredItem", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPrimitiveCalendarComponent.prototype, "showAdjacent", void 0);
        __decorate([
            core.Input(),
            cdk.tuiRequiredSetter()
        ], TuiPrimitiveCalendarComponent.prototype, "month", null);
        __decorate([
            core.Output()
        ], TuiPrimitiveCalendarComponent.prototype, "hoveredItemChange", void 0);
        __decorate([
            core.Output()
        ], TuiPrimitiveCalendarComponent.prototype, "dayClick", void 0);
        __decorate([
            core.HostBinding('class._single')
        ], TuiPrimitiveCalendarComponent.prototype, "isSingle", null);
        TuiPrimitiveCalendarComponent = __decorate([
            core.Component({
                selector: 'tui-primitive-calendar',
                template: "<div class=\"row row_weekday\">\n    <div *ngFor=\"let day of (weekDays$ | async)\" class=\"cell\">{{day}}</div>\n</div>\n<div>\n    <div\n        *tuiRepeatTimes=\"let rowIndex of sheet.length\"\n        automation-id=\"tui-primitive-calendar__row\"\n        class=\"row\"\n    >\n        <ng-container *tuiRepeatTimes=\"let colIndex of sheet[rowIndex].length\">\n            <ng-container *tuiLet=\"sheet[rowIndex][colIndex] as item\">\n                <div\n                    *ngIf=\"!itemIsUnavailable(item) || showAdjacent\"\n                    automation-id=\"tui-primitive-calendar__cell\"\n                    class=\"cell\"\n                    [class.cell_today]=\"itemIsToday(item)\"\n                    [class.cell_interval]=\"itemIsInterval(item)\"\n                    [attr.data-tui-element-range]=\"getItemRange(item)\"\n                    [attr.data-tui-element-state]=\"getItemState(item)\"\n                    (tuiHoveredChange)=\"onItemHovered($event, item)\"\n                    (tuiPressedChange)=\"onItemPressed($event, item)\"\n                    (click)=\"onItemClick($event, item)\"\n                >\n                    <div\n                        automation-id=\"tui-primitive-calendar__item\"\n                        class=\"item\"\n                        [class.item_weekend]=\"colIndex > 4\"\n                        [class.item_unavailable]=\"itemIsUnavailable(item)\"\n                    >\n                        {{item.day}}\n                        <div\n                            *ngIf=\"item | tuiMapper : toMarkers : itemIsToday(item) : !!getItemRange(item) as markers\"\n                            class=\"dots\"\n                        >\n                            <div class=\"dot\" [tuiBackground]=\"markers[0]\"></div>\n                            <div\n                                *ngIf=\"markers.length > 1\"\n                                class=\"dot\"\n                                [tuiBackground]=\"markers[1]\"\n                            ></div>\n                        </div>\n                    </div>\n                </div>\n            </ng-container>\n        </ng-container>\n    </div>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [":host{font:var(--tui-font-text-m);display:block}.row{position:relative;z-index:0;display:flex;justify-content:space-between;height:36px}.item{position:relative;flex:1;line-height:32px;border-radius:var(--tui-radius-m)}.item:after,.item:before{content:'';position:absolute;top:0;left:0;right:0;bottom:0;z-index:-1;border-radius:var(--tui-radius-m)}.cell{position:relative;display:flex;align-items:center;justify-content:center;width:36px;text-align:center;outline:0;cursor:pointer;background-clip:content-box;box-sizing:border-box;border:2px solid transparent}.cell:before{content:'';position:absolute;top:0;left:0;right:0;bottom:0;z-index:-1;border-radius:var(--tui-radius-m)}.cell_today:after{position:absolute;left:50%;transform:translate(-50%,0);content:'';bottom:5px;height:2px;width:12px;border-radius:6px;background-color:var(--tui-text-02)}.cell_interval:before{background:var(--tui-base-02)}:host._single .cell_interval:before{background:var(--tui-secondary-hover)}.cell_interval:not(:last-child):before{right:-36px}.cell_interval:last-child:first-child:before{right:0}.cell_interval:first-child>.item{border-top-left-radius:var(--tui-radius-m);border-bottom-left-radius:var(--tui-radius-m)}.cell_interval:last-child>.item{border-top-right-radius:var(--tui-radius-m);border-bottom-right-radius:var(--tui-radius-m)}.cell_interval>.item{border-radius:0}.cell[data-tui-element-range]:after{background-color:var(--tui-primary-text)}.cell[data-tui-element-range]>.item{color:var(--tui-primary-text)}.cell[data-tui-element-range]>.item:after,.cell[data-tui-element-range]>.item:before{background-color:var(--tui-primary)}.cell[data-tui-element-range][data-tui-element-state=hovered]>.item:after,.cell[data-tui-element-range][data-tui-element-state=hovered]>.item:before{background-color:var(--tui-primary-hover)}.cell[data-tui-element-range][data-tui-element-state=pressed]>.item:after,.cell[data-tui-element-range][data-tui-element-state=pressed]>.item:before{background-color:var(--tui-primary-active)}.cell[data-tui-element-range=end]>.item:before{left:4px}.cell[data-tui-element-range=end]>.item:after{left:-32px;right:100%;transform:translateX(23px) scaleY(.6) scaleX(.4) rotate(45deg)}.cell[data-tui-element-range=start]>.item:before{right:4px}.cell[data-tui-element-range=start]>.item:after{left:100%;right:-32px;transform:translateX(-23px) scaleY(.6) scaleX(.4) rotate(45deg)}.cell[data-tui-element-state=disabled]{pointer-events:none}.cell[data-tui-element-state=disabled]>.item{opacity:.36}.cell[data-tui-element-state=hovered]:hover:not([data-tui-element-range])>.item{background-color:var(--tui-secondary-hover)}.cell[data-tui-element-state=pressed]:hover:not([data-tui-element-range])>.item{background-color:var(--tui-secondary-active)}:host{width:252px}.row{justify-content:flex-start}.row:first-child{justify-content:flex-end}.row_weekday{font:var(--tui-font-text-s);color:var(--tui-text-02);pointer-events:none}.item{display:flex;flex-direction:column}.item_weekend{color:var(--tui-negative)}.item_unavailable{opacity:var(--tui-disabled-opacity)}.dots{display:flex;justify-content:center;margin-top:-8px;padding-bottom:4px}.dot{display:inline-block;width:4px;height:4px;border-radius:100%;margin:0 1px}"]
            }),
            __param(0, core.Inject(tokens.TUI_SHORT_WEEK_DAYS))
        ], TuiPrimitiveCalendarComponent);
        return TuiPrimitiveCalendarComponent;
    }());

    var TuiPrimitiveCalendarModule = /** @class */ (function () {
        function TuiPrimitiveCalendarModule() {
        }
        TuiPrimitiveCalendarModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    cdk.TuiLetModule,
                    cdk.TuiMapperPipeModule,
                    cdk.TuiRepeatTimesModule,
                    cdk.TuiHoveredModule,
                    cdk.TuiPressedModule,
                    color.TuiColorModule,
                ],
                declarations: [TuiPrimitiveCalendarComponent],
                exports: [TuiPrimitiveCalendarComponent],
            })
        ], TuiPrimitiveCalendarModule);
        return TuiPrimitiveCalendarModule;
    }());

    exports.TuiPrimitiveCalendarComponent = TuiPrimitiveCalendarComponent;
    exports.TuiPrimitiveCalendarModule = TuiPrimitiveCalendarModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
