// @taiga-ui/kit/bundles/taiga-ui-kit-providers.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/providers', ['exports', '@angular/core', '@taiga-ui/core'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.providers = {}), global.ng.core, global.core$1));
}(this, (function (exports, core, core$1) { 'use strict';

    function fixedDropdownControllerFactory(directive) {
        directive = directive || new core$1.TuiDropdownControllerDirective();
        directive.limitWidth = "fixed" /* Fixed */;
        return directive;
    }
    var FIXED_DROPDOWN_CONTROLLER_PROVIDER = [
        {
            provide: core$1.TUI_DROPDOWN_CONTROLLER,
            deps: [[new core.Optional(), core$1.TuiDropdownControllerDirective]],
            useFactory: fixedDropdownControllerFactory,
        },
    ];
    function leftAlignedDropdownControllerFactory(directive) {
        directive = directive || new core$1.TuiDropdownControllerDirective();
        directive.align = 'left';
        return directive;
    }
    var LEFT_ALIGNED_DROPDOWN_CONTROLLER_PROVIDER = [
        {
            provide: core$1.TUI_DROPDOWN_CONTROLLER,
            deps: [[new core.Optional(), core$1.TuiDropdownControllerDirective]],
            useFactory: leftAlignedDropdownControllerFactory,
        },
    ];

    exports.FIXED_DROPDOWN_CONTROLLER_PROVIDER = FIXED_DROPDOWN_CONTROLLER_PROVIDER;
    exports.LEFT_ALIGNED_DROPDOWN_CONTROLLER_PROVIDER = LEFT_ALIGNED_DROPDOWN_CONTROLLER_PROVIDER;
    exports.fixedDropdownControllerFactory = fixedDropdownControllerFactory;
    exports.leftAlignedDropdownControllerFactory = leftAlignedDropdownControllerFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
