// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-number.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('angular2-text-mask')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-number', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', 'angular2-text-mask'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['input-number'] = {}), global.ng.core, global.ng.forms, global.cdk, global.core$1, global.angular2TextMask));
}(this, (function (exports, core, forms, cdk, core$1, angular2TextMask) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var DEFAULT_MAX_LENGTH = 18;
    // @dynamic
    var TuiInputNumberComponent = /** @class */ (function (_super) {
        __extends(TuiInputNumberComponent, _super);
        function TuiInputNumberComponent(control, changeDetectorRef) {
            var _this = _super.call(this, control, changeDetectorRef) || this;
            _this.min = -Infinity;
            _this.max = Infinity;
            _this.decimal = "not-zero" /* NotZero */;
            _this.precision = 2;
            _this.postfix = '';
            _this.mask = function (allowNegative, decimal, precision, nativeFocusableElement) { return ({
                mask: core$1.tuiCreateNumberMask({
                    allowNegative: allowNegative,
                    allowDecimal: decimal !== 'never',
                    decimalLimit: precision,
                    requireDecimal: decimal === 'always',
                }),
                pipe: core$1.tuiCreateAutoCorrectedNumberPipe(decimal === 'always' ? precision : 0, ',', nativeFocusableElement || undefined),
                guide: false,
            }); };
            return _this;
        }
        TuiInputNumberComponent_1 = TuiInputNumberComponent;
        Object.defineProperty(TuiInputNumberComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return !this.primitiveTextfield || this.computedDisabled
                    ? null
                    : this.primitiveTextfield.nativeFocusableElement;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "focused", {
            get: function () {
                return !!this.primitiveTextfield && this.primitiveTextfield.focused;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "isNegativeAllowed", {
            get: function () {
                return this.min < 0;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "inputMode", {
            get: function () {
                return this.decimal === 'never' ? "numeric" /* Numeric */ : "decimal" /* Decimal */;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "calculatedMaxLength", {
            get: function () {
                return (DEFAULT_MAX_LENGTH +
                    (this.decimal !== "never" /* Never */ && this.nativeValue.includes(',')
                        ? this.precision + 1
                        : 0));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "formattedValue", {
            get: function () {
                var value = this.value || 0;
                var absValue = Math.abs(value);
                var hasFraction = absValue % 1 > 0;
                var limit = this.decimal === 'always' || hasFraction ? this.precision : 0;
                var fraction = hasFraction
                    ? value.toString().split('.')[1].substr(0, this.precision)
                    : '';
                if (this.focused && this.decimal !== 'always') {
                    limit = fraction.length;
                }
                return core$1.formatNumber(value, limit);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "computedValue", {
            get: function () {
                if (this.focused || !this.isNativeValueInLimit) {
                    return this.nativeValue;
                }
                if (this.value === null) {
                    return core$1.maskedMoneyValueIsEmpty(this.nativeValue) ? this.nativeValue : '';
                }
                return this.formattedValue;
            },
            enumerable: true,
            configurable: true
        });
        TuiInputNumberComponent.prototype.onValue = function (value) {
            if (core$1.maskedMoneyValueIsEmpty(value)) {
                this.updateValue(null);
                return;
            }
            if (this.isNativeValueNotFinished) {
                return;
            }
            var capped = this.absoluteCapInputValue(value);
            if (capped === null || isNaN(capped)) {
                return;
            }
            this.updateValue(capped);
            if (capped !== core$1.maskedNumberStringToNumber(value)) {
                this.nativeValue = this.formattedValue;
            }
        };
        TuiInputNumberComponent.prototype.onKeyDown = function (event) {
            if (event.key !== ',' && event.key !== '.') {
                return;
            }
            if (this.decimal === 'never') {
                event.preventDefault();
                return;
            }
            if (this.nativeValue.includes(',')) {
                event.preventDefault();
                this.setCaretAfterComma();
            }
        };
        TuiInputNumberComponent.prototype.onFocused = function (focused) {
            this.updateFocused(focused);
            if (focused) {
                return;
            }
            var nativeNumberValue = core$1.maskedNumberStringToNumber(this.nativeValue);
            if (isNaN(nativeNumberValue)) {
                this.clear();
                return;
            }
            var clamped = Math.min(this.max, Math.max(this.min, nativeNumberValue));
            this.updateValue(clamped);
            this.nativeValue = this.formattedValue;
        };
        TuiInputNumberComponent.prototype.onHovered = function (hovered) {
            this.updateHovered(hovered);
        };
        TuiInputNumberComponent.prototype.onPressed = function (pressed) {
            this.updatePressed(pressed);
        };
        Object.defineProperty(TuiInputNumberComponent.prototype, "isNativeValueInLimit", {
            get: function () {
                if (this.nativeValue === '') {
                    return true;
                }
                var nativeNumberValue = core$1.maskedNumberStringToNumber(this.nativeValue);
                return nativeNumberValue >= this.min && nativeNumberValue <= this.max;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "isNativeValueNotFinished", {
            get: function () {
                var nativeNumberValue = core$1.maskedNumberStringToNumber(this.nativeValue);
                return nativeNumberValue < 0
                    ? nativeNumberValue > this.max
                    : nativeNumberValue < this.min;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputNumberComponent.prototype, "nativeValue", {
            get: function () {
                return this.nativeFocusableElement ? this.nativeFocusableElement.value : '';
            },
            set: function (value) {
                if (!this.primitiveTextfield || !this.nativeFocusableElement) {
                    return;
                }
                this.primitiveTextfield.value = value;
                this.nativeFocusableElement.value = value;
            },
            enumerable: true,
            configurable: true
        });
        TuiInputNumberComponent.prototype.clear = function () {
            this.nativeValue = '';
            this.updateValue(null);
        };
        TuiInputNumberComponent.prototype.absoluteCapInputValue = function (inputValue) {
            var value = core$1.maskedNumberStringToNumber(inputValue);
            var capped = value < 0 ? Math.max(this.min, value) : Math.min(value, this.max);
            var ineligibleValue = isNaN(capped) || capped < this.min || capped > this.max;
            return ineligibleValue ? null : capped;
        };
        TuiInputNumberComponent.prototype.setCaretAfterComma = function () {
            if (!this.nativeFocusableElement) {
                return;
            }
            var afterCommaPosition = this.nativeValue.length - this.precision;
            this.nativeFocusableElement.setSelectionRange(afterCommaPosition, afterCommaPosition);
        };
        var TuiInputNumberComponent_1;
        TuiInputNumberComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] }
        ]; };
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputNumberComponent.prototype, "min", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputNumberComponent.prototype, "max", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputNumberComponent.prototype, "decimal", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputNumberComponent.prototype, "precision", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputNumberComponent.prototype, "postfix", void 0);
        __decorate([
            core.ViewChild(core$1.TuiPrimitiveTextfieldComponent)
        ], TuiInputNumberComponent.prototype, "primitiveTextfield", void 0);
        TuiInputNumberComponent = TuiInputNumberComponent_1 = __decorate([
            core.Component({
                selector: 'tui-input-number',
                template: "<tui-primitive-textfield\n    class=\"textfield\"\n    tuiTextfieldInputMode=\"decimal\"\n    [pseudoHovered]=\"pseudoHovered\"\n    [pseudoFocused]=\"computedFocused\"\n    [invalid]=\"computedInvalid\"\n    [tuiTextfieldMaxLength]=\"calculatedMaxLength\"\n    [readOnly]=\"readOnly\"\n    [disabled]=\"computedDisabled\"\n    [textMask]=\"isNegativeAllowed | tuiMapper: mask:decimal:precision:nativeFocusableElement\"\n    [value]=\"computedValue\"\n    [postfix]=\"postfix\"\n    [focusable]=\"focusable\"\n    (valueChange)=\"onValue($event)\"\n    (hoveredChange)=\"onHovered($event)\"\n    (focusedChange)=\"onFocused($event)\"\n    (pressedChange)=\"onPressed($event)\"\n    (keydown)=\"onKeyDown($event)\"\n>\n    <ng-content></ng-content>\n</tui-primitive-textfield>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [
                    {
                        provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TuiInputNumberComponent_1; }),
                    },
                ],
                styles: [":host{display:block;border-radius:var(--tui-radius-m)}.textfield{border-radius:inherit}"]
            }),
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef))
        ], TuiInputNumberComponent);
        return TuiInputNumberComponent;
    }(cdk.AbstractTuiNullableControl));

    var TuiInputNumberModule = /** @class */ (function () {
        function TuiInputNumberModule() {
        }
        TuiInputNumberModule = __decorate([
            core.NgModule({
                imports: [
                    angular2TextMask.TextMaskModule,
                    cdk.TuiMapperPipeModule,
                    core$1.TuiPrimitiveTextfieldModule,
                    core$1.TuiTextfieldControllerModule,
                ],
                declarations: [TuiInputNumberComponent],
                exports: [TuiInputNumberComponent],
            })
        ], TuiInputNumberModule);
        return TuiInputNumberModule;
    }());

    exports.TuiInputNumberComponent = TuiInputNumberComponent;
    exports.TuiInputNumberModule = TuiInputNumberModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
