// @taiga-ui/kit/bundles/taiga-ui-kit-components-line-clamp.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@tinkoff/ng-polymorpheus'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('@taiga-ui/core'), require('@ng-web-apis/common'), require('@taiga-ui/kit/utils/dom')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/line-clamp', ['exports', '@angular/core', '@taiga-ui/cdk', '@tinkoff/ng-polymorpheus', 'rxjs', 'rxjs/operators', '@angular/common', '@taiga-ui/core', '@ng-web-apis/common', '@taiga-ui/kit/utils/dom'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['line-clamp'] = {}), global.ng.core, global.cdk, global.ngPolymorpheus, global.rxjs, global.rxjs.operators, global.ng.common, global.core$1, global.common$1, global['taiga-ui'].kit.utils.dom));
}(this, (function (exports, core, cdk, ngPolymorpheus, rxjs, operators, common, core$1, common$1, dom) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var TuiLineClampComponent = /** @class */ (function () {
        function TuiLineClampComponent(elementRef, renderer) {
            var _this = this;
            this.elementRef = elementRef;
            this.lineHeight = 24;
            this.content = '';
            this.linesLimit$ = new rxjs.BehaviorSubject(1);
            this.initialized = false;
            // Skipping initial transition
            setTimeout(function () {
                renderer.addClass(_this.elementRef.nativeElement, '_initialized');
            });
        }
        Object.defineProperty(TuiLineClampComponent.prototype, "linesLimit", {
            set: function (linesLimit) {
                this.linesLimit$.next(linesLimit);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiLineClampComponent.prototype, "lineClamp$", {
            get: function () {
                var _this = this;
                return this.linesLimit$.pipe(operators.startWith(1), operators.pairwise(), operators.switchMap(function (_a) {
                    var _b = __read(_a, 2), prev = _b[0], next = _b[1];
                    return next >= prev
                        ? rxjs.of(next)
                        : cdk.typedFromEvent(_this.elementRef.nativeElement, 'transitionend').pipe(operators.filter(cdk.isCurrentTarget), operators.mapTo(next));
                }));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiLineClampComponent.prototype, "overflown", {
            get: function () {
                if (!this.outlet) {
                    return false;
                }
                var _a = this.outlet.nativeElement, scrollHeight = _a.scrollHeight, scrollWidth = _a.scrollWidth;
                var _b = this.elementRef.nativeElement, clientHeight = _b.clientHeight, clientWidth = _b.clientWidth;
                // 4px buffer for IE/Edge incorrectly rounding scrollHeight
                return scrollHeight - clientHeight > 4 || scrollWidth - clientWidth > 0;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiLineClampComponent.prototype, "computedContent", {
            get: function () {
                return this.overflown ? this.content : '';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiLineClampComponent.prototype, "oneLine", {
            get: function () {
                return this.linesLimit$.value === 1;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiLineClampComponent.prototype, "maxHeight", {
            get: function () {
                return this.initialized ? this.lineHeight * this.linesLimit$.value : null;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiLineClampComponent.prototype, "height", {
            get: function () {
                return !this.outlet ? 0 : this.outlet.nativeElement.scrollHeight + 4 || null;
            },
            enumerable: true,
            configurable: true
        });
        TuiLineClampComponent.prototype.ngAfterViewInit = function () {
            this.initialized = true;
        };
        // Change detection
        TuiLineClampComponent.prototype.markForCheck = function () { };
        TuiLineClampComponent.ctorParameters = function () { return [
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: core.Renderer2, decorators: [{ type: core.Inject, args: [core.Renderer2,] }] }
        ]; };
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiLineClampComponent.prototype, "linesLimit", null);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiLineClampComponent.prototype, "lineHeight", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiLineClampComponent.prototype, "content", void 0);
        __decorate([
            core.ViewChild(ngPolymorpheus.PolymorpheusOutletComponent, { read: core.ElementRef })
        ], TuiLineClampComponent.prototype, "outlet", void 0);
        __decorate([
            cdk.tuiPure
        ], TuiLineClampComponent.prototype, "lineClamp$", null);
        __decorate([
            core.HostBinding('style.maxHeight.px')
        ], TuiLineClampComponent.prototype, "maxHeight", null);
        __decorate([
            core.HostBinding('style.height.px')
        ], TuiLineClampComponent.prototype, "height", null);
        __decorate([
            core.HostListener('mouseenter')
        ], TuiLineClampComponent.prototype, "markForCheck", null);
        TuiLineClampComponent = __decorate([
            core.Component({
                selector: 'tui-line-clamp',
                template: "<div\n    polymorpheus-outlet\n    class=\"wrapper\"\n    lineClamp\n    tuiHintId=\"unnecessary\"\n    tuiHintMode=\"overflow\"\n    [tuiHint]=\"computedContent\"\n    [class.wrapper_ellipsis]=\"oneLine\"\n    [style.-webkit-line-clamp]=\"lineClamp$ | async\"\n    [content]=\"content\"\n    (tuiResize)=\"markForCheck()\"\n></div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [":host{position:relative;display:block;overflow:hidden}:host._initialized{transition-property:max-height;transition-duration:.3s;transition-timing-function:ease-in-out}.wrapper{display:-webkit-box;-webkit-box-orient:vertical;overflow:hidden}.wrapper_ellipsis{white-space:nowrap;overflow:hidden;text-overflow:ellipsis;display:block}"]
            }),
            __param(0, core.Inject(core.ElementRef)),
            __param(1, core.Inject(core.Renderer2))
        ], TuiLineClampComponent);
        return TuiLineClampComponent;
    }());

    var HIDDEN_LEFT = -200;
    // @dynamic
    var TuiLineClampDirective = /** @class */ (function () {
        function TuiLineClampDirective(component, viewContainerRef, documentRef, renderer, ngZone, _a, destroy$, animationFrame$, 
        /**
         * TODO: remove "any" in new TS version; https://github.com/ng-web-apis/common/pull/6
         */
        cssRef) {
            var nativeElement = _a.nativeElement;
            if (cssRef.supports('-webkit-line-clamp', '1') ||
                !(nativeElement.parentElement instanceof HTMLElement)) {
                return;
            }
            var range = documentRef.createRange();
            var ellipsis = renderer.createElement('div');
            var parentElement = nativeElement.parentElement;
            var style = ellipsis.style;
            style.position = 'absolute';
            style.bottom = '0';
            style.left = HIDDEN_LEFT + "px";
            ellipsis.textContent = cdk.CHAR_ELLIPSIS;
            viewContainerRef.element.nativeElement.appendChild(ellipsis);
            animationFrame$
                .pipe(operators.map(function () {
                // 4px buffer for IE/Edge incorrectly rounding scrollHeight
                if (component.oneLine ||
                    nativeElement.scrollHeight - parentElement.clientHeight < 4) {
                    return HIDDEN_LEFT;
                }
                var result = 0;
                var length = nativeElement.textContent
                    ? nativeElement.textContent.length
                    : 0;
                for (var char = 0; char < length - 2; char++) {
                    dom.setRangeOffset(range, nativeElement, char, 'setStart');
                    dom.setRangeOffset(range, nativeElement, char + 1, 'setEnd');
                    var rangeRect = range.getBoundingClientRect();
                    var clientRect = parentElement.getBoundingClientRect();
                    if (Math.round(rangeRect.top - clientRect.bottom) >= 0) {
                        break;
                    }
                    result = rangeRect.right - clientRect.left;
                }
                return result;
            }), cdk.tuiZonefree(ngZone), operators.takeUntil(destroy$))
                .subscribe(function (left) {
                style.left = left + "px";
            });
        }
        TuiLineClampDirective.ctorParameters = function () { return [
            { type: TuiLineClampComponent, decorators: [{ type: core.Inject, args: [TuiLineClampComponent,] }] },
            { type: core.ViewContainerRef, decorators: [{ type: core.Inject, args: [core.ViewContainerRef,] }] },
            { type: Document, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] },
            { type: core.Renderer2, decorators: [{ type: core.Inject, args: [core.Renderer2,] }] },
            { type: core.NgZone, decorators: [{ type: core.Inject, args: [core.NgZone,] }] },
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: cdk.TuiDestroyService, decorators: [{ type: core.Inject, args: [cdk.TuiDestroyService,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [common$1.ANIMATION_FRAME,] }] },
            { type: undefined, decorators: [{ type: core.Inject, args: [common$1.CSS,] }] }
        ]; };
        TuiLineClampDirective = __decorate([
            core.Directive({
                selector: '[lineClamp]',
                providers: [cdk.TuiDestroyService],
            }),
            __param(0, core.Inject(TuiLineClampComponent)),
            __param(1, core.Inject(core.ViewContainerRef)),
            __param(2, core.Inject(common.DOCUMENT)),
            __param(3, core.Inject(core.Renderer2)),
            __param(4, core.Inject(core.NgZone)),
            __param(5, core.Inject(core.ElementRef)),
            __param(6, core.Inject(cdk.TuiDestroyService)),
            __param(7, core.Inject(common$1.ANIMATION_FRAME)),
            __param(8, core.Inject(common$1.CSS))
        ], TuiLineClampDirective);
        return TuiLineClampDirective;
    }());

    var TuiLineClampModule = /** @class */ (function () {
        function TuiLineClampModule() {
        }
        TuiLineClampModule = __decorate([
            core.NgModule({
                imports: [common.CommonModule, ngPolymorpheus.PolymorpheusModule, core$1.TuiHintModule, cdk.TuiResizeModule],
                declarations: [TuiLineClampComponent, TuiLineClampDirective],
                exports: [TuiLineClampComponent],
            })
        ], TuiLineClampModule);
        return TuiLineClampModule;
    }());

    exports.TuiLineClampComponent = TuiLineClampComponent;
    exports.TuiLineClampDirective = TuiLineClampDirective;
    exports.TuiLineClampModule = TuiLineClampModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
