// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-range.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/abstract'), require('@angular/common'), require('@taiga-ui/kit/components/range'), require('angular2-text-mask')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-range', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/abstract', '@angular/common', '@taiga-ui/kit/components/range', 'angular2-text-mask'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['input-range'] = {}), global.ng.core, global.ng.forms, global.cdk, global.core$1, global['taiga-ui'].kit.abstract, global.ng.common, global['taiga-ui'].kit.components.range, global.angular2TextMask));
}(this, (function (exports, core, forms, cdk, core$1, abstract, common, range, angular2TextMask) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    // @dynamic
    var TuiInputRangeComponent = /** @class */ (function (_super) {
        __extends(TuiInputRangeComponent, _super);
        function TuiInputRangeComponent(control, changeDetectorRef, modeDirective, isMobile) {
            var _this = _super.call(this, control, changeDetectorRef) || this;
            _this.modeDirective = modeDirective;
            _this.isMobile = isMobile;
            return _this;
        }
        TuiInputRangeComponent_1 = TuiInputRangeComponent;
        Object.defineProperty(TuiInputRangeComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return !this.nativeLeft || this.disabled ? null : this.nativeLeft.nativeElement;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "focused", {
            get: function () {
                return this.focusedLeft || this.focusedRight;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "focusedLeft", {
            get: function () {
                return !!this.nativeLeft && cdk.isNativeFocused(this.nativeLeft.nativeElement);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "focusedRight", {
            get: function () {
                return !!this.nativeRight && cdk.isNativeFocused(this.nativeRight.nativeElement);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "showMinLabel", {
            get: function () {
                return !this.focusedLeft && !!this.minLabel && this.value[0] === this.min;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "showMaxLabel", {
            get: function () {
                return !this.focusedRight && !!this.maxLabel && this.value[1] === this.max;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "inputValueLeft", {
            get: function () {
                return this.nativeLeft ? this.nativeLeft.nativeElement.value : '';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "inputValueRight", {
            get: function () {
                return this.nativeRight ? this.nativeRight.nativeElement.value : '';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "computedValueLeft", {
            get: function () {
                return core$1.formatNumber(this.value[0]);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputRangeComponent.prototype, "computedValueRight", {
            get: function () {
                return core$1.formatNumber(this.value[1]);
            },
            enumerable: true,
            configurable: true
        });
        TuiInputRangeComponent.prototype.onActiveZone = function (active) {
            this.updateFocused(active);
        };
        TuiInputRangeComponent.prototype.onMouseDown = function () {
            if (this.nativeRight && !this.isMobile) {
                cdk.setNativeFocused(this.nativeRight.nativeElement);
            }
        };
        TuiInputRangeComponent.prototype.onKeyDownArrowUpLeft = function (event) {
            if (this.readOnly) {
                return;
            }
            event.preventDefault();
            this.processStep(true, false);
        };
        TuiInputRangeComponent.prototype.onKeyDownArrowDownLeft = function (event) {
            if (this.readOnly) {
                return;
            }
            event.preventDefault();
            this.processStep(false, false);
        };
        TuiInputRangeComponent.prototype.onKeyDownArrowUpRight = function (event) {
            if (this.readOnly) {
                return;
            }
            event.preventDefault();
            this.processStep(true, true);
        };
        TuiInputRangeComponent.prototype.onKeyDownArrowDownRight = function (event) {
            if (this.readOnly) {
                return;
            }
            event.preventDefault();
            this.processStep(false, true);
        };
        TuiInputRangeComponent.prototype.onInputLeft = function () {
            var value = this.inputValueLeft;
            var capped = this.capInputValue(value, this.value[1]);
            var postfix = value.slice(-1)[0] === ',' ? ',' : '';
            if (core$1.maskedMoneyValueIsEmpty(value) || capped === null) {
                return;
            }
            var newValue = core$1.formatNumber(capped) + postfix;
            if (this.nativeLeft && this.inputValueLeft !== newValue) {
                this.nativeLeft.nativeElement.value = newValue;
            }
            this.updateValue([capped, this.value[1]]);
        };
        TuiInputRangeComponent.prototype.onInputRight = function () {
            var value = this.inputValueRight;
            var capped = this.capInputValue(value);
            var postfix = value.slice(-1)[0] === ',' ? ',' : '';
            if (core$1.maskedMoneyValueIsEmpty(value) || capped === null) {
                return;
            }
            var newValue = core$1.formatNumber(capped) + postfix;
            if (this.nativeRight && this.inputValueRight !== newValue) {
                this.nativeRight.nativeElement.value = newValue;
            }
            if (capped >= this.value[0]) {
                this.updateValue([this.value[0], capped]);
            }
        };
        TuiInputRangeComponent.prototype.onRangeValue = function (value) {
            var _this = this;
            var guardedValue = value.map(function (item) { return _this.valueGuard(item); });
            if (!this.nativeLeft ||
                !this.nativeRight ||
                (guardedValue[0] === this.value[0] && guardedValue[1] === this.value[1])) {
                return;
            }
            if (!this.isMobile) {
                var element = guardedValue[0] !== this.value[0]
                    ? this.nativeLeft.nativeElement
                    : this.nativeRight.nativeElement;
                cdk.setNativeFocused(element);
            }
            this.updateValue(guardedValue);
            this.nativeLeft.nativeElement.value = core$1.formatNumber(guardedValue[0]);
        };
        TuiInputRangeComponent.prototype.onLeftFocused = function (focused) {
            if (focused || !this.nativeLeft) {
                return;
            }
            var inputValue = core$1.maskedNumberStringToNumber(this.computedValueLeft);
            var value = isNaN(inputValue) ? this.min : this.valueGuard(inputValue);
            this.nativeLeft.nativeElement.value = core$1.formatNumber(value);
            if (value !== this.value[0]) {
                this.updateValue([value, this.value[1]]);
            }
        };
        TuiInputRangeComponent.prototype.onRightFocused = function (focused) {
            if (focused || !this.nativeRight) {
                return;
            }
            var inputValue = core$1.maskedNumberStringToNumber(this.computedValueRight);
            var value = isNaN(inputValue)
                ? this.value[0]
                : this.valueGuard(Math.max(inputValue, this.value[0]));
            this.nativeRight.nativeElement.value = core$1.formatNumber(value);
            if (value !== this.value[1]) {
                this.updateValue([this.value[0], value]);
            }
        };
        TuiInputRangeComponent.prototype.getFallbackValue = function () {
            return [0, 0];
        };
        TuiInputRangeComponent.prototype.processStep = function (increment, right) {
            var start = this.valueGuard(increment ? this.value[0] + this.step : this.value[0] - this.step);
            var end = this.valueGuard(increment ? this.value[1] + this.step : this.value[1] - this.step);
            var value = [
                right ? this.value[0] : Math.min(start, this.value[1]),
                right ? Math.max(end, this.value[0]) : this.value[1],
            ];
            if (value[0] !== this.value[0] || value[1] !== this.value[1]) {
                this.updateValue(value);
            }
        };
        var TuiInputRangeComponent_1;
        TuiInputRangeComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: core$1.TuiModeDirective, decorators: [{ type: core.Optional }, { type: core.Inject, args: [core$1.TuiModeDirective,] }] },
            { type: Boolean, decorators: [{ type: core.Inject, args: [cdk.TUI_IS_MOBILE,] }] }
        ]; };
        __decorate([
            core.ViewChild('nativeLeft')
        ], TuiInputRangeComponent.prototype, "nativeLeft", void 0);
        __decorate([
            core.ViewChild('nativeRight')
        ], TuiInputRangeComponent.prototype, "nativeRight", void 0);
        __decorate([
            core.HostBinding('class._min-label')
        ], TuiInputRangeComponent.prototype, "showMinLabel", null);
        __decorate([
            core.HostBinding('class._max-label')
        ], TuiInputRangeComponent.prototype, "showMaxLabel", null);
        TuiInputRangeComponent = TuiInputRangeComponent_1 = __decorate([
            core.Component({
                selector: 'tui-input-range',
                template: "<div class=\"zone\" (tuiActiveZoneChange)=\"onActiveZone($event)\">\n    <tui-wrapper\n        appearance=\"textfield\"\n        [readOnly]=\"readOnly\"\n        [disabled]=\"disabled\"\n        [focused]=\"computedFocused\"\n        [hovered]=\"computedHovered\"\n        [invalid]=\"computedInvalid\"\n    >\n        <span\n            *ngIf=\"hasPlaceholder\"\n            class=\"placeholder\"\n            automation-id=\"tui-input-range__placeholder\"\n        >\n            <ng-content></ng-content>\n        </span>\n        <div class=\"content\">\n            <span *ngIf=\"!showMinLabel\" class=\"label\">\n                <span class=\"value\">{{inputValueLeft}}</span>\n                <span\n                    *ngIf=\"isPluralized(pluralize)\"\n                    class=\"pluralize\"\n                    automation-id=\"tui-input-range__pluralize-left\"\n                >\n                    {{value[0] | tuiPluralize: pluralize}}\n                </span>\n            </span>\n            <span\n                *ngIf=\"showMinLabel\"\n                class=\"label\"\n                automation-id=\"tui-input-range__min-label\"\n            >\n                {{minLabel}}\n            </span>\n            <span\n                *ngIf=\"showMaxLabel\"\n                automation-id=\"tui-input-range__max-label\"\n                class=\"max\"\n            >\n                {{maxLabel}}\n            </span>\n        </div>\n        <tui-range\n            class=\"slider\"\n            [min]=\"min\"\n            [max]=\"max\"\n            [steps]=\"computedSteps\"\n            [segments]=\"segments\"\n            [keySteps]=\"keySteps\"\n            [pluralize]=\"segmentsPluralize\"\n            [focusable]=\"false\"\n            [disabled]=\"readOnly || disabled\"\n            [ngModel]=\"value\"\n            (ngModelChange)=\"onRangeValue($event)\"\n            (mousedown.prevent)=\"onMouseDown()\"\n        ></tui-range>\n    </tui-wrapper>\n\n    <div class=\"native-wrapper\" (tuiHoveredChange)=\"onHovered($event)\">\n        <input\n            #nativeLeft\n            class=\"native\"\n            automation-id=\"tui-input-range__native-left\"\n            [disabled]=\"disabled\"\n            [readOnly]=\"readOnly\"\n            [class.native_hidden]=\"showMinLabel\"\n            [tuiInputMode]=\"inputMode\"\n            [tuiFocusable]=\"focusable\"\n            [textMask]=\"quantum | tuiMapper: mask: min\"\n            [ngModel]=\"computedValueLeft\"\n            (ngModelChange)=\"onInputLeft()\"\n            (tuiFocusedChange)=\"onLeftFocused($event)\"\n            (keydown.arrowUp)=\"onKeyDownArrowUpLeft($event)\"\n            (keydown.arrowDown)=\"onKeyDownArrowDownLeft($event)\"\n        />\n        <input\n            #nativeRight\n            class=\"native native_right\"\n            automation-id=\"tui-input-range__native-right\"\n            [disabled]=\"disabled\"\n            [readOnly]=\"readOnly\"\n            [class.native_hidden]=\"showMaxLabel\"\n            [tuiInputMode]=\"inputMode\"\n            [tuiFocusable]=\"focusable\"\n            [textMask]=\"quantum | tuiMapper: mask: min\"\n            [ngModel]=\"computedValueRight\"\n            (ngModelChange)=\"onInputRight()\"\n            (tuiFocusedChange)=\"onRightFocused($event)\"\n            (keydown.arrowUp)=\"onKeyDownArrowUpRight($event)\"\n            (keydown.arrowDown)=\"onKeyDownArrowDownRight($event)\"\n        />\n        <span\n            *ngIf=\"isPluralized(pluralize) && !showMaxLabel\"\n            class=\"pluralize pluralize_right\"\n            automation-id=\"tui-input-range__pluralize-right\"\n            (mousedown)=\"onMouseDown()\"\n        >\n            {{value[1] | tuiPluralize: pluralize}}\n        </span>\n    </div>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [
                    {
                        provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TuiInputRangeComponent_1; }),
                    },
                ],
                styles: [":host{font:var(--tui-font-text-m);position:relative;z-index:0;display:block;border-radius:var(--tui-radius-m);color:var(--tui-text-01)}:host._segmented{border-bottom:26px solid transparent}:host[data-mode=onDark]{color:var(--tui-base-01)}.zone{border-radius:inherit}.native-wrapper{position:absolute;top:0;left:0;width:100%;height:100%;display:flex}.pluralize{display:flex;align-items:center;padding-right:12px;margin-left:-8px}.pluralize_right{margin-left:-12px}:host[data-tui-host-size='l'] .pluralize_right{padding-top:20px}:host._disabled .pluralize_right{color:var(--tui-text-03)}:host[data-mode=onDark]._disabled .pluralize_right{color:var(--tui-text-03-night)}.native{padding:0 16px;border:0;border-radius:inherit;background:0 0;font-size:inherit;line-height:inherit;font-weight:inherit;color:inherit;caret-color:currentColor;-webkit-appearance:none;-moz-appearance:none;appearance:none;word-break:keep-all;-webkit-text-fill-color:currentColor;flex:1;min-width:0;color:var(--tui-text-01);box-sizing:border-box;outline:0}.native:-webkit-autofill,.native:-webkit-autofill:focus,.native:-webkit-autofill:hover{border-radius:inherit;-webkit-text-fill-color:inherit!important;color:inherit!important;background-color:transparent!important;-webkit-box-shadow:0 0 0 1000px var(--tui-autofill) inset!important}.native_right{text-align:right}.native_hidden{opacity:0}:host._disabled .native{color:var(--tui-text-03);-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none}:host[data-tui-host-size='l'] .native{padding:20px 16px 0}:host[data-mode=onDark] .native{color:var(--tui-text-01-night)}:host[data-mode=onDark]._disabled .native{color:var(--tui-text-03-night)}.content{display:flex;height:var(--tui-height-m);padding:0 16px;justify-content:space-between;align-items:center}:host[data-tui-host-size='l'] .content{height:var(--tui-height-l);padding:0 16px}:host[data-tui-host-size='l']._disabled .content{color:var(--tui-text-03)}:host[data-tui-host-size='l'][data-mode=onDark]._disabled .content{color:var(--tui-text-03-night)}.placeholder{transition-property:transform,min-width,color,letter-spacing;transition-duration:.3s;transition-timing-function:ease-in-out;font:var(--tui-font-text-s);color:var(--tui-text-01);white-space:nowrap;overflow:hidden;text-overflow:ellipsis;display:block;font-size:13px;box-sizing:border-box;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;pointer-events:none;color:var(--tui-text-02);transform-origin:left;letter-spacing:normal;text-transform:none;position:absolute;top:0;width:100%;height:100%;left:16px;transform:translateY(-10px) scale(.87);line-height:var(--tui-height-l);max-width:calc(114% - 32px)}.placeholder_raised{width:114%;transform:translateY(-10px) scale(.87)}:host[data-tui-host-size='m'] .placeholder_raised{width:118%;transform:translateY(-8px) scale(.85);letter-spacing:.4px}:host._invalid:not(._focused) .placeholder_raised,:host._invalid:not(._focused)._hovered .placeholder_raised{color:var(--tui-error-fill)}:host[data-mode=onDark]._invalid:not(._focused) .placeholder_raised,:host[data-mode=onDark]._invalid:not(._focused)._hovered .placeholder_raised{color:var(--tui-error-fill-night)}:host._focused .placeholder,:host[data-tui-host-size='l']._focused._label-outside .placeholder,:host[data-tui-host-size='m']._focused._label-outside .placeholder{color:var(--tui-text-03)}:host[data-tui-host-size='l'] .placeholder{font-size:15px}:host[data-tui-host-size='l']._focused:not(._label-outside) .placeholder,:host[data-tui-host-size='m']._focused:not(._label-outside) .placeholder{color:var(--tui-text-01)}:host[data-mode=onDark] .placeholder{color:var(--tui-text-02-night)}:host[data-tui-host-size='l'][data-mode=onDark]._focused:not(._label-outside) .placeholder,:host[data-tui-host-size='m'][data-mode=onDark]._focused:not(._label-outside) .placeholder{color:var(--tui-text-01-night)}:host[data-mode=onDark]._focused .placeholder,:host[data-tui-host-size='l'][data-mode=onDark]._focused._label-outside .placeholder,:host[data-tui-host-size='m'][data-mode=onDark]._focused._label-outside .placeholder{color:var(--tui-text-02-night)}.value{visibility:hidden;overflow:hidden;padding-right:12px}:host._disabled .value{color:var(--tui-text-03)}.label{display:flex;width:50%}:host[data-tui-host-size='l'] .label{padding-top:20px}.max{text-align:right;flex:1}:host[data-tui-host-size='l'] .max{padding-top:20px;line-height:calc(var(--tui-height-l) - 20px)}.slider{position:absolute;top:100%;left:0;right:0;z-index:1;border-top-left-radius:0;border-top-right-radius:0;margin:-9px 0 0;color:transparent}:host._disabled .slider,:host._readonly .slider{pointer-events:none}"]
            }),
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Optional()),
            __param(2, core.Inject(core$1.TuiModeDirective)),
            __param(3, core.Inject(cdk.TUI_IS_MOBILE))
        ], TuiInputRangeComponent);
        return TuiInputRangeComponent;
    }(abstract.AbstractTuiInputSlider));

    var TuiInputRangeModule = /** @class */ (function () {
        function TuiInputRangeModule() {
        }
        TuiInputRangeModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    forms.FormsModule,
                    forms.ReactiveFormsModule,
                    angular2TextMask.TextMaskModule,
                    cdk.TuiFocusableModule,
                    cdk.TuiFocusedModule,
                    cdk.TuiHoveredModule,
                    cdk.TuiPressedModule,
                    cdk.TuiMapperPipeModule,
                    cdk.TuiInputModeModule,
                    cdk.TuiActiveZoneModule,
                    core$1.TuiPluralizePipeModule,
                    range.TuiRangeModule,
                    core$1.TuiWrapperModule,
                ],
                declarations: [TuiInputRangeComponent],
                exports: [TuiInputRangeComponent],
            })
        ], TuiInputRangeModule);
        return TuiInputRangeModule;
    }());

    exports.TuiInputRangeComponent = TuiInputRangeComponent;
    exports.TuiInputRangeModule = TuiInputRangeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
