// @taiga-ui/kit/bundles/taiga-ui-kit-directives-dropdown-selection.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/utils/dom'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/directives/dropdown-selection', ['exports', '@angular/common', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/utils/dom', 'rxjs', 'rxjs/operators'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.directives = global['taiga-ui'].kit.directives || {}, global['taiga-ui'].kit.directives['dropdown-selection'] = {}), global.ng.common, global.ng.core, global.cdk, global.core$1, global['taiga-ui'].kit.utils.dom, global.rxjs, global.rxjs.operators));
}(this, (function (exports, common, core, cdk, core$1, dom, rxjs, operators) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    // @dynamic
    var TuiDropdownSelectionDirective = /** @class */ (function (_super) {
        __extends(TuiDropdownSelectionDirective, _super);
        function TuiDropdownSelectionDirective(documentRef, componentFactoryResolver, injector, portalService, elementRef, activeZone, shadowRootRef, customElementRef, destroy$, refresh$, changeDetectorRef, ngZone, renderer, viewContainerRef) {
            var _this = _super.call(this, componentFactoryResolver, injector, portalService, customElementRef || elementRef, activeZone) || this;
            _this.refresh$ = refresh$;
            _this.changeDetectorRef = changeDetectorRef;
            _this.ngZone = ngZone;
            _this.renderer = renderer;
            _this.viewContainerRef = viewContainerRef;
            _this.position = "selection" /* Selection */;
            _this.visibilityHandler = cdk.ALWAYS_TRUE_HANDLER;
            _this.documentRef = shadowRootRef || documentRef;
            _this.range = _this.documentRef.createRange();
            var nativeElement = _this.elementRef.nativeElement;
            rxjs.merge(cdk.typedFromEvent(_this.documentRef, 'mouseup'), cdk.typedFromEvent(nativeElement, 'mousedown').pipe(operators.switchMapTo(cdk.typedFromEvent(nativeElement, 'mousemove').pipe(operators.takeUntil(cdk.typedFromEvent(_this.documentRef, 'mouseup'))))), cdk.typedFromEvent(nativeElement, 'keyup'))
                .pipe(operators.map(function () {
                var active = cdk.getNativeFocused(_this.documentRef);
                var selection = _this.documentRef.getSelection();
                if ((active instanceof HTMLInputElement ||
                    active instanceof HTMLTextAreaElement) &&
                    nativeElement.contains(active)) {
                    return _this.veryVerySadInputFix(active);
                }
                return selection && selection.rangeCount
                    ? selection.getRangeAt(0)
                    : _this.range;
            }), operators.takeUntil(destroy$))
                .subscribe(function (range) {
                var contained = nativeElement.contains(range.commonAncestorContainer);
                _this.range = contained ? range : _this.range;
                var valid = contained &&
                    (!_this.visibilityHandler || _this.visibilityHandler(_this.range));
                _this.toggleDropdownBox(valid || _this.inDropdown(range));
            });
            return _this;
        }
        TuiDropdownSelectionDirective_1 = TuiDropdownSelectionDirective;
        Object.defineProperty(TuiDropdownSelectionDirective.prototype, "tuiDropdownSelection", {
            set: function (handler) {
                if (!handler) {
                    return;
                }
                var inHostAndValid = this.elementRef.nativeElement.contains(this.range.commonAncestorContainer) &&
                    handler(this.range);
                this.visibilityHandler = handler;
                this.toggleDropdownBox(inHostAndValid);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiDropdownSelectionDirective.prototype, "clientRect", {
            get: function () {
                var defaultView = this.documentRef.defaultView;
                var rangeRect = this.rangeRect;
                var frameElement = defaultView ? defaultView.frameElement : null;
                if (!frameElement) {
                    return rangeRect;
                }
                var documentRect = frameElement.getBoundingClientRect();
                return {
                    top: rangeRect.top + documentRect.top,
                    left: rangeRect.left + documentRect.left,
                    right: rangeRect.left + documentRect.left + rangeRect.width,
                    bottom: rangeRect.top + documentRect.top + rangeRect.height,
                    width: rangeRect.width,
                    height: rangeRect.height,
                };
            },
            enumerable: true,
            configurable: true
        });
        TuiDropdownSelectionDirective.prototype.ngOnDestroy = function () {
            this.closeDropdownBox();
            if (this.ghost) {
                this.renderer.removeChild(this.viewContainerRef.element.nativeElement, this.ghost);
            }
        };
        Object.defineProperty(TuiDropdownSelectionDirective.prototype, "rangeRect", {
            /**
             * get ClientRect of current Range according to provided position
             */
            get: function () {
                switch (this.position) {
                    case "tag" /* Tag */:
                        var commonAncestorContainer = this.range.commonAncestorContainer;
                        var element = commonAncestorContainer.nodeType === Node.ELEMENT_NODE
                            ? commonAncestorContainer
                            : commonAncestorContainer.parentNode;
                        return element.getBoundingClientRect();
                    case "word" /* Word */:
                        return dom.getWordRange(this.range).getBoundingClientRect();
                    default:
                        return this.range.getBoundingClientRect();
                }
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Toggle dropdown visibility (has to be in ngZone.run because it could be initiated inside iframe in Editor)
         */
        TuiDropdownSelectionDirective.prototype.toggleDropdownBox = function (visible) {
            var _this = this;
            this.ngZone.run(function () {
                if (visible) {
                    _this.openDropdownBox();
                }
                else {
                    _this.closeDropdownBox();
                }
                _this.changeDetectorRef.markForCheck();
            });
        };
        /**
         * Check if Node is inside dropdown
         */
        TuiDropdownSelectionDirective.prototype.boxContains = function (node) {
            return (!!this.dropdownBoxRef &&
                this.dropdownBoxRef.location.nativeElement.contains(node));
        };
        /**
         * Check if given range is at leaset partially inside dropdown
         */
        TuiDropdownSelectionDirective.prototype.inDropdown = function (range) {
            var startContainer = range.startContainer, endContainer = range.endContainer;
            var inDropdown = this.boxContains(range.commonAncestorContainer);
            var hostToDropdown = this.boxContains(endContainer) &&
                this.elementRef.nativeElement.contains(startContainer);
            var dropdownToHost = this.boxContains(startContainer) &&
                this.elementRef.nativeElement.contains(endContainer);
            return inDropdown || hostToDropdown || dropdownToHost;
        };
        /**
         * Position invisible DIV and create Range similar to selected range inside input/textarea
         */
        TuiDropdownSelectionDirective.prototype.veryVerySadInputFix = function (element) {
            var _a = this.ghost, ghost = _a === void 0 ? this.initGhost(element) : _a;
            var _b = element.getBoundingClientRect(), top = _b.top, left = _b.left, width = _b.width, height = _b.height;
            var selectionStart = element.selectionStart, selectionEnd = element.selectionEnd;
            var range = this.documentRef.createRange();
            var hostRect = this.elementRef.nativeElement.getBoundingClientRect();
            ghost.style.top = cdk.px(top - hostRect.top);
            ghost.style.left = cdk.px(left - hostRect.left);
            ghost.style.width = cdk.px(width);
            ghost.style.height = cdk.px(height);
            ghost.textContent = cdk.CHAR_ZERO_WIDTH_SPACE + element.value + cdk.CHAR_NO_BREAK_SPACE;
            range.setStart(ghost.firstChild, selectionStart || 0);
            range.setEnd(ghost.firstChild, selectionEnd || 0);
            return range;
        };
        /**
         * Create an invisible DIV styled exactly like input/textarea element inside directive
         */
        TuiDropdownSelectionDirective.prototype.initGhost = function (element) {
            var ghost = this.renderer.createElement('div');
            var nativeElement = this.viewContainerRef.element.nativeElement;
            var _a = getComputedStyle(element), font = _a.font, letterSpacing = _a.letterSpacing, textTransform = _a.textTransform, padding = _a.padding;
            ghost.style.position = 'absolute';
            ghost.style.pointerEvents = 'none';
            ghost.style.opacity = '0';
            ghost.style.whiteSpace = 'pre-wrap';
            ghost.style.font = font;
            ghost.style.letterSpacing = letterSpacing;
            ghost.style.textTransform = textTransform;
            ghost.style.padding = padding;
            this.renderer.appendChild(nativeElement, ghost);
            this.ghost = ghost;
            return ghost;
        };
        var TuiDropdownSelectionDirective_1;
        TuiDropdownSelectionDirective.ctorParameters = function () { return [
            { type: Document, decorators: [{ type: core.Inject, args: [common.DOCUMENT,] }] },
            { type: core.ComponentFactoryResolver, decorators: [{ type: core.Inject, args: [core.ComponentFactoryResolver,] }] },
            { type: core.Injector, decorators: [{ type: core.Inject, args: [core.Injector,] }] },
            { type: cdk.TuiPortalService, decorators: [{ type: core.Inject, args: [cdk.TuiPortalService,] }] },
            { type: core.ElementRef, decorators: [{ type: core.Host }, { type: core.Inject, args: [core.ElementRef,] }] },
            { type: cdk.TuiActiveZoneDirective, decorators: [{ type: core.Inject, args: [cdk.TuiActiveZoneDirective,] }, { type: core.Optional }] },
            { type: Document, decorators: [{ type: core.Inject, args: [core$1.TUI_DOCUMENT_OR_SHADOW_ROOT,] }, { type: core.Optional }] },
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core$1.TUI_ELEMENT_REF,] }, { type: core.Optional }] },
            { type: cdk.TuiDestroyService, decorators: [{ type: core.Inject, args: [cdk.TuiDestroyService,] }] },
            { type: cdk.TuiParentsScrollService, decorators: [{ type: core.Inject, args: [cdk.TuiParentsScrollService,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: core.NgZone, decorators: [{ type: core.Inject, args: [core.NgZone,] }] },
            { type: core.Renderer2, decorators: [{ type: core.Inject, args: [core.Renderer2,] }] },
            { type: core.ViewContainerRef, decorators: [{ type: core.Inject, args: [core.ViewContainerRef,] }] }
        ]; };
        __decorate([
            core.Input()
        ], TuiDropdownSelectionDirective.prototype, "tuiDropdownSelection", null);
        __decorate([
            core.Input('tuiDropdownSelectionPosition')
        ], TuiDropdownSelectionDirective.prototype, "position", void 0);
        TuiDropdownSelectionDirective = TuiDropdownSelectionDirective_1 = __decorate([
            core.Directive({
                selector: '[tuiDropdownSelection]:not(ng-container)',
                providers: [
                    {
                        provide: core$1.TUI_DROPDOWN_DIRECTIVE,
                        useExisting: core.forwardRef(function () { return TuiDropdownSelectionDirective_1; }),
                    },
                    cdk.TuiParentsScrollService,
                    cdk.TuiDestroyService,
                ],
            }),
            __param(0, core.Inject(common.DOCUMENT)),
            __param(1, core.Inject(core.ComponentFactoryResolver)),
            __param(2, core.Inject(core.Injector)),
            __param(3, core.Inject(cdk.TuiPortalService)),
            __param(4, core.Host()), __param(4, core.Inject(core.ElementRef)),
            __param(5, core.Inject(cdk.TuiActiveZoneDirective)),
            __param(5, core.Optional()),
            __param(6, core.Inject(core$1.TUI_DOCUMENT_OR_SHADOW_ROOT)),
            __param(6, core.Optional()),
            __param(7, core.Inject(core$1.TUI_ELEMENT_REF)),
            __param(7, core.Optional()),
            __param(8, core.Inject(cdk.TuiDestroyService)),
            __param(9, core.Inject(cdk.TuiParentsScrollService)),
            __param(10, core.Inject(core.ChangeDetectorRef)),
            __param(11, core.Inject(core.NgZone)),
            __param(12, core.Inject(core.Renderer2)),
            __param(13, core.Inject(core.ViewContainerRef))
        ], TuiDropdownSelectionDirective);
        return TuiDropdownSelectionDirective;
    }(core$1.AbstractTuiDropdown));

    var TuiDropdownSelectionModule = /** @class */ (function () {
        function TuiDropdownSelectionModule() {
        }
        TuiDropdownSelectionModule = __decorate([
            core.NgModule({
                imports: [core$1.TuiDropdownBoxModule],
                declarations: [TuiDropdownSelectionDirective],
                exports: [TuiDropdownSelectionDirective],
            })
        ], TuiDropdownSelectionModule);
        return TuiDropdownSelectionModule;
    }());

    exports.TuiDropdownSelectionDirective = TuiDropdownSelectionDirective;
    exports.TuiDropdownSelectionModule = TuiDropdownSelectionModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
