// @taiga-ui/kit/bundles/taiga-ui-kit-components-pagination.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/tokens'), require('@taiga-ui/kit/utils/math'), require('rxjs'), require('@angular/common'), require('@tinkoff/ng-polymorpheus')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/pagination', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/tokens', '@taiga-ui/kit/utils/math', 'rxjs', '@angular/common', '@tinkoff/ng-polymorpheus'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components.pagination = {}), global.ng.core, global.cdk, global.core$1, global['taiga-ui'].kit.tokens, global['taiga-ui'].kit.utils.math, global.rxjs, global.ng.common, global.ngPolymorpheus));
}(this, (function (exports, core, cdk, core$1, tokens, math, rxjs, common, ngPolymorpheus) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var DOTS_LENGTH = 1;
    var ACTIVE_ITEM_LENGTH = 1;
    function nonNegativeInteger(length) {
        return Number.isInteger(length) && length >= 0;
    }
    // @dynamic
    var TuiPaginationComponent = /** @class */ (function (_super) {
        __extends(TuiPaginationComponent, _super);
        function TuiPaginationComponent(elementRef, modeDirective, texts$) {
            var _this = _super.call(this) || this;
            _this.elementRef = elementRef;
            _this.modeDirective = modeDirective;
            _this.texts$ = texts$;
            _this.length = 1;
            _this.size = 'm';
            _this.disabled = false;
            /**
             * Amount of visible pages around active page
             */
            _this.activePadding = 1;
            /**
             * Amount of visible pages at the edges
             */
            _this.sidePadding = 1;
            /**
             * Customization for page number display.
             */
            _this.content = null;
            /**
             * Active page index
             */
            _this.index = 0;
            _this.indexChange = new core.EventEmitter();
            _this.elements = cdk.EMPTY_QUERY;
            return _this;
        }
        TuiPaginationComponent_1 = TuiPaginationComponent;
        Object.defineProperty(TuiPaginationComponent.prototype, "nativeFocusableElement", {
            get: function () {
                if (this.disabled) {
                    return null;
                }
                var activeElementIndex = 0;
                var elementsLength = this.elementsLength;
                for (var i = 0; i < elementsLength; i++) {
                    var itemIndex = this.getItemIndexByElementIndex(i);
                    if (itemIndex) {
                        activeElementIndex++;
                    }
                    if (itemIndex === this.index) {
                        break;
                    }
                }
                var activeElement = this.elements.find(function (_, index) { return index === activeElementIndex; });
                return activeElement ? activeElement.nativeFocusableElement : null;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "focused", {
            get: function () {
                return cdk.isNativeFocusedIn(this.elementRef.nativeElement);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "elementsLength", {
            /**
             * Number of items in a container.
             */
            get: function () {
                return this.itemsFit ? this.length : this.maxElementsLength;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "sizeM", {
            get: function () {
                return this.size === 'm';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "mode", {
            get: function () {
                return this.modeDirective ? this.modeDirective.mode : null;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "arrowIsDisabledLeft", {
            get: function () {
                return this.index === 0;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "arrowIsDisabledRight", {
            get: function () {
                return this.reverseIndex === 0;
            },
            enumerable: true,
            configurable: true
        });
        TuiPaginationComponent.prototype.elementIsFocusable = function (index) {
            return this.index === index && !this.focused;
        };
        TuiPaginationComponent.prototype.getItemContext = function (index) {
            return { $implicit: index };
        };
        /**
         * Get index by element index
         * @param elementIndex
         * @returns index or null (for '…')
         */
        TuiPaginationComponent.prototype.getItemIndexByElementIndex = function (elementIndex) {
            if (!this.sizeM) {
                return elementIndex;
            }
            if (elementIndex < this.sidePadding) {
                return elementIndex;
            }
            if (elementIndex === this.sidePadding && this.hasCollapsedItems(this.index)) {
                return null;
            }
            var reverseElementIndex = this.lastElementIndex - elementIndex;
            if (reverseElementIndex === this.sidePadding &&
                this.hasCollapsedItems(this.reverseIndex)) {
                return null;
            }
            if (reverseElementIndex < this.sidePadding) {
                return this.lastIndex - reverseElementIndex;
            }
            var computedIndex = this.index - this.maxHalfLength + elementIndex;
            return cdk.clamp(computedIndex, elementIndex, this.lastIndex - reverseElementIndex);
        };
        TuiPaginationComponent.prototype.getElementMode = function (index) {
            return this.index === index ? "primary" /* Primary */ : "flat" /* Flat */;
        };
        TuiPaginationComponent.prototype.getSmallElementMode = function (index, mode) {
            return this.index === index && mode !== 'onLight'
                ? "primary" /* Primary */
                : "secondary" /* Secondary */;
        };
        TuiPaginationComponent.prototype.onElementClick = function (index) {
            this.updateIndex(index);
        };
        TuiPaginationComponent.prototype.onElementKeyDownArrowLeft = function (element) {
            if (element === this.elements.first) {
                return;
            }
            var previous = this.elements.find(function (_, index, array) { return array[index + 1] === element; });
            if (previous && previous.nativeFocusableElement) {
                cdk.setNativeFocused(previous.nativeFocusableElement);
            }
        };
        TuiPaginationComponent.prototype.onElementKeyDownArrowRight = function (element) {
            if (element === this.elements.last) {
                return;
            }
            var next = this.elements.find(function (_, index, array) { return array[index - 1] === element; });
            if (next && next.nativeFocusableElement) {
                cdk.setNativeFocused(next.nativeFocusableElement);
            }
        };
        TuiPaginationComponent.prototype.onArrowClick = function (direction) {
            this.tryChangeTo(direction);
            this.focusActive();
        };
        TuiPaginationComponent.prototype.onActiveZone = function (focused) {
            this.updateFocused(focused);
        };
        Object.defineProperty(TuiPaginationComponent.prototype, "reverseIndex", {
            /**
             * Active index from the end
             */
            get: function () {
                return this.lastIndex - this.index;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "maxHalfLength", {
            /**
             * Max number of elements in half (not counting the middle one).
             */
            get: function () {
                return this.sidePadding + DOTS_LENGTH + this.activePadding;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "itemsFit", {
            /**
             * Is there '...' anywhere
             */
            get: function () {
                return this.length <= this.maxElementsLength;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "maxElementsLength", {
            /**
             * Max number of elements
             */
            get: function () {
                return this.maxHalfLength * 2 + ACTIVE_ITEM_LENGTH;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "lastIndex", {
            get: function () {
                return this.length - 1;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiPaginationComponent.prototype, "lastElementIndex", {
            get: function () {
                return this.elementsLength - 1;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Are there collapsed items at that index
         * @param index
         * @returns there are collapsed items
         */
        TuiPaginationComponent.prototype.hasCollapsedItems = function (index) {
            return !this.itemsFit && index > this.maxHalfLength;
        };
        TuiPaginationComponent.prototype.tryChangeTo = function (direction) {
            this.updateIndex(cdk.clamp(this.index + math.horizontalDirectionToNumber(direction), 0, this.lastIndex));
        };
        TuiPaginationComponent.prototype.focusActive = function () {
            var nativeFocusableElement = this.nativeFocusableElement;
            if (nativeFocusableElement) {
                cdk.setNativeFocused(nativeFocusableElement);
            }
        };
        TuiPaginationComponent.prototype.updateIndex = function (index) {
            if (this.index === index) {
                return;
            }
            this.index = index;
            this.indexChange.emit(index);
        };
        var TuiPaginationComponent_1;
        TuiPaginationComponent.ctorParameters = function () { return [
            { type: core.ElementRef, decorators: [{ type: core.Inject, args: [core.ElementRef,] }] },
            { type: core$1.TuiModeDirective, decorators: [{ type: core.Optional }, { type: core.Inject, args: [core$1.TuiModeDirective,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_PAGINATION_TEXTS,] }] }
        ]; };
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp(nonNegativeInteger, 'Must be non-negative integer')
        ], TuiPaginationComponent.prototype, "length", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPaginationComponent.prototype, "size", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPaginationComponent.prototype, "disabled", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPaginationComponent.prototype, "activePadding", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPaginationComponent.prototype, "sidePadding", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiPaginationComponent.prototype, "content", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp(nonNegativeInteger, 'Must be non-negative integer')
        ], TuiPaginationComponent.prototype, "index", void 0);
        __decorate([
            core.Output()
        ], TuiPaginationComponent.prototype, "indexChange", void 0);
        __decorate([
            core.ViewChildren('element', { read: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR })
        ], TuiPaginationComponent.prototype, "elements", void 0);
        TuiPaginationComponent = TuiPaginationComponent_1 = __decorate([
            core.Component({
                selector: 'tui-pagination',
                template: "<div class=\"content\" (tuiActiveZoneChange)=\"onActiveZone($event)\">\n    <ng-container *ngIf=\"sizeM && (texts$ | async) as texts; else smallButtons\">\n        <button\n            tuiIconButton\n            type=\"button\"\n            tuiPreventDefault=\"mousedown\"\n            class=\"button\"\n            size=\"s\"\n            appearance=\"flat\"\n            icon=\"tuiIconChevronLeft\"\n            [title]=\"texts[0]\"\n            [disabled]=\"arrowIsDisabledLeft\"\n            [focusable]=\"false\"\n            (click)=\"onArrowClick('left')\"\n        ></button>\n        <ng-container *tuiRepeatTimes=\"let elementIndex of elementsLength\">\n            <ng-container\n                *tuiLet=\"getItemIndexByElementIndex(elementIndex) as index\"\n            >\n                <button\n                    *ngIf=\"index !== null; else dotsTemplate\"\n                    #element\n                    tuiButton\n                    type=\"button\"\n                    automation-id=\"tui-pagination__element\"\n                    class=\"button\"\n                    shape=\"square\"\n                    size=\"s\"\n                    [disabled]=\"disabled\"\n                    [focusable]=\"elementIsFocusable(index)\"\n                    [appearance]=\"getElementMode(index)\"\n                    (click)=\"onElementClick(index)\"\n                    (keydown.arrowLeft.prevent)=\"onElementKeyDownArrowLeft(element)\"\n                    (keydown.arrowRight.prevent)=\"onElementKeyDownArrowRight(element)\"\n                >\n                    <div\n                        polymorpheus-outlet\n                        [content]=\"content || index + 1\"\n                        [context]=\"getItemContext(index)\"\n                    ></div>\n                </button>\n            </ng-container>\n        </ng-container>\n        <button\n            tuiIconButton\n            type=\"button\"\n            tuiPreventDefault=\"mousedown\"\n            class=\"button\"\n            size=\"s\"\n            appearance=\"flat\"\n            icon=\"tuiIconChevronRight\"\n            [title]=\"texts[1]\"\n            [disabled]=\"arrowIsDisabledRight\"\n            [focusable]=\"false\"\n            (click)=\"onArrowClick('right')\"\n        ></button>\n    </ng-container>\n    <ng-template #smallButtons>\n        <button\n            *tuiRepeatTimes=\"let indexItem of length\"\n            #element\n            tuiButton\n            type=\"button\"\n            class=\"button button_small\"\n            shape=\"square\"\n            [class.button_active]=\"indexItem === index\"\n            [disabled]=\"disabled\"\n            [focusable]=\"elementIsFocusable(indexItem)\"\n            [appearance]=\"getSmallElementMode(indexItem, mode)\"\n            (click)=\"onElementClick(indexItem)\"\n            (keydown.arrowLeft.prevent)=\"onElementKeyDownArrowLeft(element)\"\n            (keydown.arrowRight.prevent)=\"onElementKeyDownArrowRight(element)\"\n        ></button>\n    </ng-template>\n    <ng-template #dotsTemplate>\n        <div automation-id=\"tui-pagination__element\" class=\"dots\"></div>\n    </ng-template>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [
                    {
                        provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TuiPaginationComponent_1; }),
                    },
                ],
                styles: [":host{font:var(--tui-font-text-s);color:var(--tui-text-01);display:block;text-align:center}.content{display:flex;justify-content:center}.button{margin:0 2px;flex-shrink:0}.button_active{background:currentColor}.button.button.button_small{width:8px;height:8px;margin:0}.button.button.button_small:not(:first-child){margin-left:8px}.dots{width:var(--tui-height-s);height:var(--tui-height-s);line-height:var(--tui-height-s);margin:0 2px;flex-shrink:0;color:var(--tui-text-03);text-align:center;cursor:default}.dots:before{content:'\\2026'}"]
            }),
            __param(0, core.Inject(core.ElementRef)),
            __param(1, core.Optional()),
            __param(1, core.Inject(core$1.TuiModeDirective)),
            __param(2, core.Inject(tokens.TUI_PAGINATION_TEXTS))
        ], TuiPaginationComponent);
        return TuiPaginationComponent;
    }(cdk.AbstractTuiInteractive));

    var TuiPaginationModule = /** @class */ (function () {
        function TuiPaginationModule() {
        }
        TuiPaginationModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    cdk.TuiRepeatTimesModule,
                    cdk.TuiLetModule,
                    ngPolymorpheus.PolymorpheusModule,
                    cdk.TuiPreventDefaultModule,
                    cdk.TuiActiveZoneModule,
                    core$1.TuiButtonModule,
                ],
                declarations: [TuiPaginationComponent],
                exports: [TuiPaginationComponent],
            })
        ], TuiPaginationModule);
        return TuiPaginationModule;
    }());

    exports.TuiPaginationComponent = TuiPaginationComponent;
    exports.TuiPaginationModule = TuiPaginationModule;
    exports.nonNegativeInteger = nonNegativeInteger;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
