// @taiga-ui/kit/bundles/taiga-ui-kit-components-accordion.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('rxjs'), require('rxjs/operators'), require('@taiga-ui/core'), require('@tinkoff/ng-polymorpheus'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/accordion', ['exports', '@angular/core', '@taiga-ui/cdk', 'rxjs', 'rxjs/operators', '@taiga-ui/core', '@tinkoff/ng-polymorpheus', '@angular/common'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components.accordion = {}), global.ng.core, global.cdk, global.rxjs, global.rxjs.operators, global.core$1, global.ngPolymorpheus, global.ng.common));
}(this, (function (exports, core, cdk, rxjs, operators, core$1, ngPolymorpheus, common) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    // TODO: do not rebuild the template, think about "lazy" flag
    var TuiAccordionItemContentDirective = /** @class */ (function (_super) {
        __extends(TuiAccordionItemContentDirective, _super);
        function TuiAccordionItemContentDirective(templateRef, changeDetectorRef) {
            return _super.call(this, templateRef, changeDetectorRef) || this;
        }
        TuiAccordionItemContentDirective.ctorParameters = function () { return [
            { type: core.TemplateRef, decorators: [{ type: core.Inject, args: [core.TemplateRef,] }, { type: core.Self }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] }
        ]; };
        TuiAccordionItemContentDirective = __decorate([
            core.Directive({
                selector: '[tuiAccordionItemContent]',
            }),
            __param(0, core.Inject(core.TemplateRef)),
            __param(0, core.Self()),
            __param(1, core.Inject(core.ChangeDetectorRef))
        ], TuiAccordionItemContentDirective);
        return TuiAccordionItemContentDirective;
    }(ngPolymorpheus.PolymorpheusTemplate));

    var TuiAccordionItemComponent = /** @class */ (function (_super) {
        __extends(TuiAccordionItemComponent, _super);
        function TuiAccordionItemComponent(changeDetectorRef, mode$) {
            var _this = _super.call(this) || this;
            _this.changeDetectorRef = changeDetectorRef;
            _this.mode$ = mode$;
            _this.noPadding = false;
            _this.showArrow = true;
            _this.borders = "all" /* All */;
            _this.size = 'm';
            _this.disabled = false;
            _this.disableHover = false;
            _this.open = false;
            _this.async = false;
            _this.openChange = new core.EventEmitter();
            return _this;
        }
        TuiAccordionItemComponent_1 = TuiAccordionItemComponent;
        Object.defineProperty(TuiAccordionItemComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return this.disabled || !this.focusableElement
                    ? null
                    : this.focusableElement.nativeElement;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiAccordionItemComponent.prototype, "focused", {
            get: function () {
                return cdk.isNativeFocused(this.nativeFocusableElement);
            },
            enumerable: true,
            configurable: true
        });
        TuiAccordionItemComponent.prototype.onHovered = function (hovered) {
            if (!this.disableHover) {
                this.updateHovered(hovered);
            }
        };
        TuiAccordionItemComponent.prototype.onFocused = function (focused) {
            this.updateFocused(focused);
        };
        TuiAccordionItemComponent.prototype.onFocusVisible = function (focusVisible) {
            this.updateFocusVisible(focusVisible);
        };
        TuiAccordionItemComponent.prototype.onRowToggle = function () {
            if (this.disabled) {
                return;
            }
            this.updateOpen(!this.open);
        };
        TuiAccordionItemComponent.prototype.onItemKeyDownEsc = function (event) {
            if (!this.focused || !this.open) {
                return;
            }
            event.stopPropagation();
            this.updateOpen(false);
        };
        TuiAccordionItemComponent.prototype.onItemKeyDownSpace = function (event) {
            if (!this.focused) {
                return;
            }
            event.preventDefault();
            this.onRowToggle();
        };
        TuiAccordionItemComponent.prototype.close = function () {
            this.updateOpen(false);
            this.changeDetectorRef.markForCheck();
        };
        TuiAccordionItemComponent.prototype.updateOpen = function (open) {
            if (this.open === open) {
                return;
            }
            this.open = open;
            this.openChange.emit(open);
        };
        var TuiAccordionItemComponent_1;
        TuiAccordionItemComponent.ctorParameters = function () { return [
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [core$1.TUI_MODE,] }] }
        ]; };
        __decorate([
            core.Input(),
            core.HostBinding('class._no-padding'),
            cdk.tuiDefaultProp()
        ], TuiAccordionItemComponent.prototype, "noPadding", void 0);
        __decorate([
            core.Input(),
            core.HostBinding('class._has-arrow'),
            cdk.tuiDefaultProp()
        ], TuiAccordionItemComponent.prototype, "showArrow", void 0);
        __decorate([
            core.Input(),
            core.HostBinding('attr.data-tui-host-borders'),
            cdk.tuiDefaultProp()
        ], TuiAccordionItemComponent.prototype, "borders", void 0);
        __decorate([
            core.Input(),
            core.HostBinding('attr.data-tui-host-size'),
            cdk.tuiDefaultProp()
        ], TuiAccordionItemComponent.prototype, "size", void 0);
        __decorate([
            core.Input(),
            core.HostBinding('class._disabled'),
            cdk.tuiDefaultProp()
        ], TuiAccordionItemComponent.prototype, "disabled", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiAccordionItemComponent.prototype, "disableHover", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiAccordionItemComponent.prototype, "open", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiAccordionItemComponent.prototype, "async", void 0);
        __decorate([
            core.Output()
        ], TuiAccordionItemComponent.prototype, "openChange", void 0);
        __decorate([
            core.ContentChild(TuiAccordionItemContentDirective)
        ], TuiAccordionItemComponent.prototype, "content", void 0);
        __decorate([
            core.ViewChild('focusableElement')
        ], TuiAccordionItemComponent.prototype, "focusableElement", void 0);
        TuiAccordionItemComponent = TuiAccordionItemComponent_1 = __decorate([
            core.Component({
                selector: 'tui-accordion-item',
                template: "<div\n    #focusableElement\n    automation-id=\"tui-accordion__item-wrapper\"\n    class=\"wrapper\"\n    [tuiFocusable]=\"computedFocusable\"\n    (tuiFocusVisibleChange)=\"onFocusVisible($event)\"\n    (tuiFocusedChange)=\"onFocused($event)\"\n    (keydown.space)=\"onItemKeyDownSpace($event)\"\n    (keydown.enter)=\"onItemKeyDownSpace($event)\"\n    (keydown.esc)=\"onItemKeyDownEsc($event)\"\n>\n    <div\n        automation-id=\"tui-accordion__item-header\"\n        class=\"header\"\n        [class.header_open]=\"open\"\n        (tuiHoveredChange)=\"onHovered($event)\"\n        (click)=\"onRowToggle()\"\n    >\n        <div class=\"title\" automation-id=\"tui-accordion__item-title\">\n            <ng-content></ng-content>\n        </div>\n        <tui-svg\n            *ngIf=\"showArrow\"\n            automation-id=\"tui-accordion__item-arrow\"\n            class=\"icon\"\n            src=\"tuiIconChevronDownLarge\"\n            [class.icon_open]=\"open\"\n        ></tui-svg>\n    </div>\n    <tui-expand [async]=\"async\" [expanded]=\"open\">\n        <ng-template tuiExpandContent>\n            <div\n                polymorpheus-outlet\n                automation-id=\"tui-accordion__item-content\"\n                class=\"content\"\n                [content]=\"content\"\n            ></div>\n        </ng-template>\n    </tui-expand>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [
                    {
                        provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TuiAccordionItemComponent_1; }),
                    },
                    core$1.MODE_PROVIDER,
                ],
                host: {
                    '($.data-mode.attr)': 'mode$',
                },
                styles: [":host{position:relative;display:block;overflow:hidden;border-radius:var(--tui-radius-l)}:host[data-tui-host-borders=top-bottom]{border-radius:0!important}.wrapper{position:relative;border-radius:inherit}.wrapper:focus{outline:0}.wrapper:after{position:absolute;top:0;left:0;width:100%;height:100%;content:'';box-sizing:border-box;border-radius:inherit;border:1px solid var(--tui-base-04);pointer-events:none}:host:not([data-tui-host-borders]) .wrapper:after{border-width:0}:host[data-tui-host-borders=all] .wrapper:after{border-width:1px}:host[data-tui-host-borders=top-bottom] .wrapper:after{border-left-width:0;border-right-width:0}:host[data-tui-host-borders=top] .wrapper:after{border-left-width:0;border-right-width:0;border-bottom-width:0}:host[data-tui-host-borders=bottom] .wrapper:after{border-left-width:0;border-right-width:0;border-top-width:0}:host[data-mode=onDark] .wrapper:after{border-color:var(--tui-base-03)}:host[data-mode=onLight] .wrapper:after{border-color:var(--tui-text-01)}:host._focus-visible .wrapper:after{border:2px solid var(--tui-focus)}:host:not([data-mode]) .wrapper{background:var(--tui-base-01)}.header{font:var(--tui-font-text-l);transition-property:background;transition-duration:.3s;transition-timing-function:ease-in-out;display:flex;align-items:center;box-sizing:border-box;border-bottom:1px solid var(--tui-base-04);min-height:var(--tui-height-l);padding:12px 20px;color:var(--tui-text-01);cursor:pointer}:host[data-mode=onDark] .header{color:var(--tui-text-01-night);border-color:var(--tui-base-03)}:host[data-mode=onLight] .header{border-color:var(--tui-text-01)}:host:not([data-tui-host-borders]) .header{border-bottom-width:0;box-shadow:none}:host._has-arrow .header{padding-right:12px}:host._hovered:not([data-mode]) .header{background:var(--tui-base-02)}:host._hovered[data-mode=onDark] .header,:host[data-mode=onDark] .header_open{background:var(--tui-clear-inverse)}:host._hovered[data-mode=onLight] .header,:host[data-mode=onLight] .header_open{background:var(--tui-clear)}:host[data-tui-host-size='s'] .header{font:var(--tui-font-text-m);min-height:var(--tui-height-m);padding:10px 12px 10px 16px}:host._no-padding .header{padding-left:0;padding-right:0}:host._disabled .header{cursor:default}.title{white-space:nowrap;overflow:hidden;text-overflow:ellipsis;margin-right:0;flex-grow:1}:host._has-arrow .title{margin-right:8px}.icon{transition-property:transform,color;transition-duration:.3s;transition-timing-function:ease-in-out;width:24px;height:24px;margin-left:auto;color:var(--tui-base-06)}.icon_open{transform:rotate(180deg)}:host[data-mode] .icon{opacity:var(--tui-disabled-opacity)}:host._hovered[data-mode] .icon{opacity:.8}:host[data-mode=onDark] .icon{color:var(--tui-text-01-night)}:host[data-mode=onLight] .icon{color:var(--tui-text-01)}:host._hovered:not([data-mode]) .icon{color:var(--tui-base-07)}.content{font:var(--tui-font-text-m);padding:20px;word-wrap:break-word}:host[data-tui-host-size='s'] .content{font:var(--tui-font-text-s);padding:16px}"]
            }),
            __param(0, core.Inject(core.ChangeDetectorRef)),
            __param(1, core.Inject(core$1.TUI_MODE))
        ], TuiAccordionItemComponent);
        return TuiAccordionItemComponent;
    }(cdk.AbstractTuiInteractive));

    var TuiAccordionComponent = /** @class */ (function () {
        function TuiAccordionComponent(destroy$) {
            this.destroy$ = destroy$;
            this.closeOthers = true;
            this.rounded = true;
            this.accordionItems = cdk.EMPTY_QUERY;
        }
        TuiAccordionComponent.prototype.ngAfterContentInit = function () {
            var _this = this;
            var accordionItems = this.accordionItems;
            var rows$ = cdk.itemsQueryListObservable(accordionItems);
            var newOpenRow$ = rows$.pipe(operators.pairwise(), operators.map(function (_a) {
                var _b = __read(_a, 2), previous = _b[0], current = _b[1];
                return current.find(function (item) { return previous.indexOf(item) === -1 && item.open; });
            }), operators.filter(cdk.isPresent));
            var rowsOpen$ = rxjs.merge(rows$.pipe(operators.switchMap(function (rows) {
                return rxjs.merge.apply(void 0, __spread(rows.map(function (row) {
                    return row.openChange.pipe(operators.filter(cdk.identity), operators.mapTo(row));
                })));
            })), newOpenRow$).pipe(operators.filter(function () { return _this.closeOthers; }), operators.takeUntil(this.destroy$));
            rowsOpen$.subscribe(function (currentRow) {
                accordionItems.forEach(function (row) {
                    if (currentRow !== row) {
                        row.close();
                    }
                });
            });
        };
        TuiAccordionComponent.ctorParameters = function () { return [
            { type: cdk.TuiDestroyService, decorators: [{ type: core.Inject, args: [cdk.TuiDestroyService,] }] }
        ]; };
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiAccordionComponent.prototype, "closeOthers", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiAccordionComponent.prototype, "rounded", void 0);
        __decorate([
            core.ContentChildren(TuiAccordionItemComponent)
        ], TuiAccordionComponent.prototype, "accordionItems", void 0);
        TuiAccordionComponent = __decorate([
            core.Component({
                selector: 'tui-accordion',
                template: "<!--TODO: Unnecessary nesting, probably accordion can be a directive-->\n<div\n    tuiGroup\n    automation-id=\"tui-accordion__group\"\n    class=\"group\"\n    orientation=\"vertical\"\n    size=\"l\"\n    [collapsed]=\"true\"\n    [rounded]=\"rounded\"\n>\n    <ng-content select=\"tui-accordion-item\"></ng-content>\n</div>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [cdk.TuiDestroyService],
                styles: [":host{display:block}.group{display:flex}"]
            }),
            __param(0, core.Inject(cdk.TuiDestroyService))
        ], TuiAccordionComponent);
        return TuiAccordionComponent;
    }());

    var TuiAccordionModule = /** @class */ (function () {
        function TuiAccordionModule() {
        }
        TuiAccordionModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    cdk.TuiFocusedModule,
                    cdk.TuiFocusVisibleModule,
                    cdk.TuiFocusableModule,
                    cdk.TuiHoveredModule,
                    ngPolymorpheus.PolymorpheusModule,
                    core$1.TuiSvgModule,
                    core$1.TuiGroupModule,
                    core$1.TuiExpandModule,
                ],
                declarations: [
                    TuiAccordionComponent,
                    TuiAccordionItemComponent,
                    TuiAccordionItemContentDirective,
                ],
                exports: [
                    TuiAccordionComponent,
                    TuiAccordionItemComponent,
                    TuiAccordionItemContentDirective,
                ],
            })
        ], TuiAccordionModule);
        return TuiAccordionModule;
    }());

    exports.TuiAccordionComponent = TuiAccordionComponent;
    exports.TuiAccordionItemComponent = TuiAccordionItemComponent;
    exports.TuiAccordionItemContentDirective = TuiAccordionItemContentDirective;
    exports.TuiAccordionModule = TuiAccordionModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
