// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-phone.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('rxjs'), require('@angular/common'), require('@taiga-ui/kit/providers'), require('rxjs/operators'), require('angular2-text-mask')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-phone', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', 'rxjs', '@angular/common', '@taiga-ui/kit/providers', 'rxjs/operators', 'angular2-text-mask'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['input-phone'] = {}), global.ng.core, global.ng.forms, global.cdk, global.core$1, global.rxjs, global.ng.common, global['taiga-ui'].kit.providers, global.rxjs.operators, global.angular2TextMask));
}(this, (function (exports, core, forms, cdk, core$1, rxjs, common, providers, operators, angular2TextMask) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var SELECTION_STREAM = new core.InjectionToken('A stream of selection changes');
    var INPUT_PHONE_PROVIDERS = [
        cdk.TuiDestroyService,
        providers.FIXED_DROPDOWN_CONTROLLER_PROVIDER,
        {
            provide: SELECTION_STREAM,
            deps: [cdk.TuiDestroyService, common.DOCUMENT],
            useFactory: selectionStreamFactory,
        },
    ];
    function selectionStreamFactory(destroy$, documentRef) {
        return rxjs.fromEvent(documentRef, 'selectionchange').pipe(operators.windowToggle(rxjs.merge(rxjs.fromEvent(documentRef, 'mouseup'), rxjs.fromEvent(documentRef, 'keydown')), function () { return rxjs.fromEvent(documentRef, 'mousedown'); }), operators.flatMap(function (windowed$) { return windowed$.pipe(operators.startWith(null)); }), operators.takeUntil(destroy$));
    }

    var NON_PLUS_AND_DIGITS_REGEX = /[ \-_\(\)]/g;
    // @dynamic
    var TuiInputPhoneComponent = /** @class */ (function (_super) {
        __extends(TuiInputPhoneComponent, _super);
        function TuiInputPhoneComponent(control, changeDetectorRef, selection$) {
            var _this = _super.call(this, control, changeDetectorRef) || this;
            _this.phoneMaskAfterCountryCode = '(###) ###-##-##';
            _this.allowText = false;
            _this.search = '';
            _this.searchChange = new core.EventEmitter();
            _this.textMaskOptions = {
                mask: function (value) {
                    return _this.allowText && !_this.value && isText(value) && value !== '+'
                        ? false
                        : __spread(_this.countryCode.split(''), [
                            ' '
                        ], _this.phoneMaskAfterCountryCode
                            .replace(/[^#\- \(\)]+/g, '')
                            .split('')
                            .map(function (item) { return (item === '#' ? /\d/ : item); }));
                },
                pipe: function (value) {
                    if (_this.allowText) {
                        return value;
                    }
                    return value === '' && _this.focused && !_this.readOnly
                        ? _this.countryCode + " "
                        : value.replace(/-$/, '');
                },
                guide: false,
            };
            _this.countryCode = '+7';
            _this.open = false;
            selection$.subscribe(function () {
                _this.setCaretPosition();
            });
            return _this;
        }
        TuiInputPhoneComponent_1 = TuiInputPhoneComponent;
        Object.defineProperty(TuiInputPhoneComponent.prototype, "countryCodeSetter", {
            set: function (countryCode) {
                this.updateValueWithNewContryCode(countryCode);
                this.countryCode = countryCode;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return !this.textfield || this.computedDisabled
                    ? null
                    : this.textfield.nativeFocusableElement;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneComponent.prototype, "focused", {
            get: function () {
                return (cdk.isNativeFocused(this.nativeFocusableElement) ||
                    (!!this.dropdown && this.dropdown.focused));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneComponent.prototype, "nativeValue", {
            get: function () {
                return this.value
                    ? core$1.formatPhone(this.value, this.countryCode, this.phoneMaskAfterCountryCode)
                    : this.search || '';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneComponent.prototype, "inputMode", {
            get: function () {
                return this.allowText ? "text" /* Text */ : "numeric" /* Numeric */;
            },
            enumerable: true,
            configurable: true
        });
        TuiInputPhoneComponent.prototype.onHovered = function (hovered) {
            this.updateHovered(hovered);
        };
        TuiInputPhoneComponent.prototype.onDrop = function (event) {
            if (!event.dataTransfer) {
                return;
            }
            this.setValueWithoutPrefix(event.dataTransfer.getData('text'));
            event.preventDefault();
        };
        TuiInputPhoneComponent.prototype.onPaste = function (event) {
            this.setValueWithoutPrefix(cdk.getClipboardDataText(event));
        };
        TuiInputPhoneComponent.prototype.onActiveZone = function (active) {
            this.updateFocused(active);
            if (active && !this.nativeValue && !this.readOnly && !this.allowText) {
                this.updateSearch(this.countryCode);
                return;
            }
            if (this.nativeValue === this.countryCode ||
                (this.search !== null &&
                    isNaN(parseInt(this.search.replace(NON_PLUS_AND_DIGITS_REGEX, ''), 10)))) {
                this.updateSearch('');
            }
        };
        TuiInputPhoneComponent.prototype.onBackspace = function (event) {
            if ((event.target.selectionStart || 0) <= this.nonRemovableLength &&
                event.target.selectionStart === event.target.selectionEnd) {
                event.preventDefault();
            }
        };
        TuiInputPhoneComponent.prototype.onValueChange = function (value) {
            var parsed = isText(value)
                ? value
                : value.replace(NON_PLUS_AND_DIGITS_REGEX, '');
            this.updateSearch(parsed);
            this.updateValue(parsed === this.countryCode || isText(parsed) ? '' : parsed);
            this.open = true;
        };
        TuiInputPhoneComponent.prototype.handleOption = function (item) {
            this.focusInput();
            this.updateValue(item);
            this.updateSearch('');
            this.open = false;
        };
        TuiInputPhoneComponent.prototype.setDisabledState = function () {
            _super.prototype.setDisabledState.call(this);
            this.open = false;
        };
        TuiInputPhoneComponent.prototype.writeValue = function (value) {
            _super.prototype.writeValue.call(this, value);
            this.updateSearch('');
        };
        TuiInputPhoneComponent.prototype.getFallbackValue = function () {
            return '';
        };
        Object.defineProperty(TuiInputPhoneComponent.prototype, "caretIsInForbiddenArea", {
            get: function () {
                var nativeFocusableElement = this.nativeFocusableElement;
                if (!nativeFocusableElement) {
                    return false;
                }
                var selectionStart = nativeFocusableElement.selectionStart, selectionEnd = nativeFocusableElement.selectionEnd;
                return (cdk.isNativeFocused(nativeFocusableElement) &&
                    selectionStart !== null &&
                    selectionStart < this.nonRemovableLength &&
                    selectionStart === selectionEnd);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneComponent.prototype, "nonRemovableLength", {
            get: function () {
                return this.isTextValue ? 0 : this.countryCode.length + 1;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneComponent.prototype, "maxPhoneLength", {
            get: function () {
                return (this.countryCode.length +
                    this.phoneMaskAfterCountryCode.replace(/[^#]+/g, '').length);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputPhoneComponent.prototype, "isTextValue", {
            get: function () {
                return !!this.search && isText(this.search);
            },
            enumerable: true,
            configurable: true
        });
        TuiInputPhoneComponent.prototype.setCaretPosition = function () {
            if (this.caretIsInForbiddenArea && !!this.nativeFocusableElement) {
                this.nativeFocusableElement.setSelectionRange(this.nonRemovableLength, this.nonRemovableLength);
            }
        };
        TuiInputPhoneComponent.prototype.setValueWithoutPrefix = function (value) {
            if (this.readOnly) {
                return;
            }
            this.open = true;
            this.updateValue(this.cleanValue(value));
            this.updateSearch(this.allowText && isText(value)
                ? value
                : value.replace(NON_PLUS_AND_DIGITS_REGEX, ''));
        };
        TuiInputPhoneComponent.prototype.cleanValue = function (value) {
            var reg = this.countryCode === '+7'
                ? /^7|^8/
                : new RegExp(this.countryCode.substring(1));
            var oldValueExist = this.value.length > this.countryCode.length &&
                this.value.length < this.maxPhoneLength;
            var newValueLength = value.replace(NON_PLUS_AND_DIGITS_REGEX, '').length;
            var cleanNewValue = value.replace(/[^0-9]+/g, '');
            var selectionLength = getSelection().toString().length;
            if (oldValueExist && selectionLength === 0) {
                return ("" + this.value + cleanNewValue).slice(0, this.maxPhoneLength);
            }
            if (newValueLength < this.maxPhoneLength - 1) {
                return ("" + this.countryCode + cleanNewValue).slice(0, this.maxPhoneLength);
            }
            return ("" + this.countryCode + cleanNewValue.replace(reg, '')).slice(0, this.maxPhoneLength);
        };
        TuiInputPhoneComponent.prototype.focusInput = function () {
            if (this.nativeFocusableElement) {
                cdk.setNativeFocused(this.nativeFocusableElement, true, true);
            }
        };
        TuiInputPhoneComponent.prototype.updateSearch = function (search) {
            if (this.search === search) {
                return;
            }
            this.search = search;
            this.searchChange.emit(search);
        };
        TuiInputPhoneComponent.prototype.updateValueWithNewContryCode = function (newCountryCode) {
            if (!this.isTextValue) {
                this.updateValue(this.value.replace(this.countryCode, newCountryCode));
            }
        };
        var TuiInputPhoneComponent_1;
        TuiInputPhoneComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [SELECTION_STREAM,] }] }
        ]; };
        __decorate([
            core.Input('countryCode'),
            cdk.tuiRequiredSetter()
        ], TuiInputPhoneComponent.prototype, "countryCodeSetter", null);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputPhoneComponent.prototype, "phoneMaskAfterCountryCode", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputPhoneComponent.prototype, "allowText", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputPhoneComponent.prototype, "search", void 0);
        __decorate([
            core.Output()
        ], TuiInputPhoneComponent.prototype, "searchChange", void 0);
        __decorate([
            core.ContentChild(core$1.TuiDataListDirective, { read: core.TemplateRef })
        ], TuiInputPhoneComponent.prototype, "datalist", void 0);
        __decorate([
            core.ViewChild(core$1.TuiHostedDropdownComponent)
        ], TuiInputPhoneComponent.prototype, "dropdown", void 0);
        __decorate([
            core.ViewChild(core$1.TuiPrimitiveTextfieldComponent)
        ], TuiInputPhoneComponent.prototype, "textfield", void 0);
        TuiInputPhoneComponent = TuiInputPhoneComponent_1 = __decorate([
            core.Component({
                selector: 'tui-input-phone',
                template: "<tui-hosted-dropdown\n    class=\"hosted\"\n    [content]=\"datalist || ''\"\n    [(open)]=\"open\"\n    (tuiActiveZoneChange)=\"onActiveZone($event)\"\n>\n    <tui-primitive-textfield\n        class=\"textfield\"\n        tuiTextfieldType=\"tel\"\n        [tuiTextfieldInputMode]=\"inputMode\"\n        [pseudoHovered]=\"pseudoHovered\"\n        [pseudoFocused]=\"computedFocused\"\n        [invalid]=\"computedInvalid\"\n        [nativeId]=\"nativeId\"\n        [readOnly]=\"readOnly\"\n        [disabled]=\"computedDisabled\"\n        [focusable]=\"focusable\"\n        [textMask]=\"textMaskOptions\"\n        [value]=\"nativeValue\"\n        (valueChange)=\"onValueChange($event)\"\n        (hoveredChange)=\"onHovered($event)\"\n        (drop)=\"onDrop($event)\"\n        (keydown.backspace)=\"onBackspace($event)\"\n        (paste.prevent)=\"onPaste($event)\"\n    >\n        <ng-content></ng-content>\n    </tui-primitive-textfield>\n</tui-hosted-dropdown>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [
                    {
                        provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TuiInputPhoneComponent_1; }),
                    },
                    {
                        provide: core$1.TUI_DATA_LIST_HOST,
                        useExisting: core.forwardRef(function () { return TuiInputPhoneComponent_1; }),
                    },
                    INPUT_PHONE_PROVIDERS,
                ],
                styles: [":host{display:block;border-radius:var(--tui-radius-m)}:host._disabled{pointer-events:none}.hosted{display:block;border-radius:inherit}.textfield{border-radius:inherit}"]
            }),
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(SELECTION_STREAM))
        ], TuiInputPhoneComponent);
        return TuiInputPhoneComponent;
    }(cdk.AbstractTuiControl));
    function isText(value) {
        return isNaN(parseInt(value.replace(NON_PLUS_AND_DIGITS_REGEX, ''), 10));
    }

    var TuiInputPhoneModule = /** @class */ (function () {
        function TuiInputPhoneModule() {
        }
        TuiInputPhoneModule = __decorate([
            core.NgModule({
                imports: [
                    angular2TextMask.TextMaskModule,
                    core$1.TuiPrimitiveTextfieldModule,
                    core$1.TuiHostedDropdownModule,
                    core$1.TuiTextfieldControllerModule,
                    cdk.TuiActiveZoneModule,
                ],
                declarations: [TuiInputPhoneComponent],
                exports: [TuiInputPhoneComponent],
            })
        ], TuiInputPhoneModule);
        return TuiInputPhoneModule;
    }());

    exports.INPUT_PHONE_PROVIDERS = INPUT_PHONE_PROVIDERS;
    exports.SELECTION_STREAM = SELECTION_STREAM;
    exports.TuiInputPhoneComponent = TuiInputPhoneComponent;
    exports.TuiInputPhoneModule = TuiInputPhoneModule;
    exports.selectionStreamFactory = selectionStreamFactory;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
