// @taiga-ui/kit/bundles/taiga-ui-kit-classes.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/classes', ['exports'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.classes = {})));
}(this, (function (exports) { 'use strict';

    var TuiDayRangePeriod = /** @class */ (function () {
        function TuiDayRangePeriod(range, name) {
            this.range = range;
            this.name = name;
        }
        TuiDayRangePeriod.prototype.toString = function () {
            return this.name;
        };
        return TuiDayRangePeriod;
    }());

    var TuiNamedDay = /** @class */ (function () {
        function TuiNamedDay(day, name, displayDay) {
            if (displayDay === void 0) { displayDay = day; }
            this.day = day;
            this.name = name;
            this.displayDay = displayDay;
        }
        TuiNamedDay.prototype.toString = function () {
            return this.name;
        };
        return TuiNamedDay;
    }());

    /**
     * Wrapper around an item to add `toString()` method
     */
    var TuiStringifiableItem = /** @class */ (function () {
        function TuiStringifiableItem(item, stringify) {
            this.item = item;
            this.stringify = stringify;
        }
        TuiStringifiableItem.prototype.toString = function () {
            return this.stringify(this.item);
        };
        return TuiStringifiableItem;
    }());

    exports.TuiDayRangePeriod = TuiDayRangePeriod;
    exports.TuiNamedDay = TuiNamedDay;
    exports.TuiStringifiableItem = TuiStringifiableItem;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
