// @taiga-ui/kit/bundles/taiga-ui-kit-components-input-time.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@taiga-ui/cdk'), require('@taiga-ui/core'), require('@taiga-ui/kit/providers'), require('@taiga-ui/kit/tokens'), require('@taiga-ui/kit/utils/mask'), require('rxjs'), require('rxjs/operators'), require('@angular/common'), require('angular2-text-mask')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/input-time', ['exports', '@angular/core', '@angular/forms', '@taiga-ui/cdk', '@taiga-ui/core', '@taiga-ui/kit/providers', '@taiga-ui/kit/tokens', '@taiga-ui/kit/utils/mask', 'rxjs', 'rxjs/operators', '@angular/common', 'angular2-text-mask'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['input-time'] = {}), global.ng.core, global.ng.forms, global.cdk, global.core$1, global['taiga-ui'].kit.providers, global['taiga-ui'].kit.tokens, global['taiga-ui'].kit.utils.mask, global.rxjs, global.rxjs.operators, global.ng.common, global.angular2TextMask));
}(this, (function (exports, core, forms, cdk, core$1, providers, tokens, mask, rxjs, operators, common, angular2TextMask) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    // @dynamic
    var TuiInputTimeComponent = /** @class */ (function (_super) {
        __extends(TuiInputTimeComponent, _super);
        function TuiInputTimeComponent(control, changeDetectorRef, textfieldSize, timeTexts$) {
            var _this = _super.call(this, control, changeDetectorRef) || this;
            _this.textfieldSize = textfieldSize;
            _this.timeTexts$ = timeTexts$;
            _this.disabledItemHandler = cdk.ALWAYS_FALSE_HANDLER;
            _this.items = [];
            _this.itemSize = 'm';
            _this.strict = false;
            _this.mode = 'HH:MM';
            _this.open = false;
            return _this;
        }
        TuiInputTimeComponent_1 = TuiInputTimeComponent;
        Object.defineProperty(TuiInputTimeComponent.prototype, "nativeFocusableElement", {
            get: function () {
                return this.textfield ? this.textfield.nativeFocusableElement : null;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "focused", {
            get: function () {
                return cdk.isNativeFocused(this.nativeFocusableElement);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "filtered", {
            get: function () {
                return this.filter(this.items, this.mode, this.computedSearch);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "textMaskOptions", {
            get: function () {
                return this.calculateMask(this.mode);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "computedValue", {
            get: function () {
                return this.value ? this.value.toString(this.mode) : this.nativeValue;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "computedSearch", {
            get: function () {
                return this.computedValue.length !== this.mode.length ? this.computedValue : '';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "interactive", {
            get: function () {
                return !this.disabled && !this.readOnly;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "innerPseudoFocused", {
            get: function () {
                if (this.pseudoFocused === false) {
                    return false;
                }
                if (this.open || this.computedFocused) {
                    return true;
                }
                return null;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "icon", {
            get: function () {
                return core$1.sizeBigger(this.textfieldSize.size) ? 'tuiIconTimeLarge' : 'tuiIconTime';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiInputTimeComponent.prototype, "nativeValue", {
            get: function () {
                return this.nativeFocusableElement ? this.nativeFocusableElement.value : '';
            },
            set: function (value) {
                if (!this.nativeFocusableElement) {
                    return;
                }
                this.nativeFocusableElement.value = value;
            },
            enumerable: true,
            configurable: true
        });
        TuiInputTimeComponent.prototype.getFiller$ = function (mode) {
            return this.timeTexts$.pipe(operators.map(function (texts) { return texts[mode]; }));
        };
        TuiInputTimeComponent.prototype.onValueChange = function (value) {
            this.open = !!this.items.length;
            if (value && this.control) {
                this.control.updateValueAndValidity();
            }
            var match = this.getMatch(value);
            if (match !== undefined) {
                this.updateValue(match);
                return;
            }
            if (value.length !== this.mode.length) {
                this.updateValue(null);
                return;
            }
            var time = cdk.TuiTime.fromString(value);
            this.updateValue(this.strict ? this.findNearestTimeFromItems(time) : time);
        };
        TuiInputTimeComponent.prototype.onFocused = function (focused) {
            var _this = this;
            this.updateFocused(focused);
            if (focused ||
                this.value !== null ||
                this.nativeValue === '' ||
                this.mode === 'HH:MM') {
                return;
            }
            var parsedTime = cdk.TuiTime.fromString(this.nativeValue);
            this.updateValue(parsedTime);
            setTimeout(function () {
                if (_this.nativeValue.endsWith('.') || _this.nativeValue.endsWith(':')) {
                    _this.nativeValue = _this.nativeValue.slice(0, -1);
                }
            });
        };
        TuiInputTimeComponent.prototype.onHovered = function (hovered) {
            this.updateHovered(hovered);
        };
        TuiInputTimeComponent.prototype.onArrowUp = function (event) {
            if (this.items.length) {
                return;
            }
            this.processArrow(event, 1);
        };
        TuiInputTimeComponent.prototype.onArrowDown = function (event) {
            if (this.items.length) {
                return;
            }
            this.processArrow(event, -1);
        };
        TuiInputTimeComponent.prototype.onMenuClick = function (item) {
            this.focusInput();
            this.updateValue(item);
        };
        TuiInputTimeComponent.prototype.onOpen = function (open) {
            this.open = open;
        };
        TuiInputTimeComponent.prototype.writeValue = function (value) {
            _super.prototype.writeValue.call(this, value);
            this.nativeValue = value ? this.computedValue : '';
        };
        TuiInputTimeComponent.prototype.findNearestTimeFromItems = function (value) {
            return this.items.reduce(function (previous, current) {
                return Math.abs(current.toAbsoluteMilliseconds() - value.toAbsoluteMilliseconds()) <
                    Math.abs(previous.toAbsoluteMilliseconds() - value.toAbsoluteMilliseconds())
                    ? current
                    : previous;
            });
        };
        TuiInputTimeComponent.prototype.getMatch = function (value) {
            return this.items.find(function (item) { return cdk.TUI_STRICT_MATCHER(item, value); });
        };
        TuiInputTimeComponent.prototype.close = function () {
            this.open = false;
        };
        TuiInputTimeComponent.prototype.processArrow = function (event, shift) {
            var target = event.target;
            if (this.readOnly || !(target instanceof HTMLInputElement)) {
                return;
            }
            var selectionStart = target.selectionStart || 0;
            this.shiftTime(this.calculateShift(selectionStart, shift));
            target.setSelectionRange(selectionStart, selectionStart);
            event.preventDefault();
        };
        TuiInputTimeComponent.prototype.calculateShift = function (selectionStart, shift) {
            if (selectionStart <= 2) {
                return { hours: shift };
            }
            if (selectionStart <= 5) {
                return { minutes: shift };
            }
            if (selectionStart <= 8) {
                return { seconds: shift };
            }
            return { ms: shift };
        };
        TuiInputTimeComponent.prototype.shiftTime = function (shift) {
            if (this.value === null) {
                return;
            }
            var increasedTime = this.value.shift(shift);
            // Manual update so we can set caret position properly
            this.nativeValue = increasedTime.toString(this.mode);
            this.updateValue(increasedTime);
        };
        TuiInputTimeComponent.prototype.focusInput = function (preventScroll) {
            if (preventScroll === void 0) { preventScroll = false; }
            if (this.nativeFocusableElement) {
                cdk.setNativeFocused(this.nativeFocusableElement, true, preventScroll);
                this.close();
            }
        };
        TuiInputTimeComponent.prototype.calculateMask = function (mode) {
            return {
                mask: mask.tuiCreateTimeMask(mode),
                pipe: mask.tuiCreateAutoCorrectedTimePipe(mode),
                guide: false,
            };
        };
        TuiInputTimeComponent.prototype.filter = function (items, mode, search) {
            return items.filter(function (item) { return item.toString(mode).includes(search); });
        };
        var TuiInputTimeComponent_1;
        TuiInputTimeComponent.ctorParameters = function () { return [
            { type: forms.NgControl, decorators: [{ type: core.Optional }, { type: core.Self }, { type: core.Inject, args: [forms.NgControl,] }] },
            { type: core.ChangeDetectorRef, decorators: [{ type: core.Inject, args: [core.ChangeDetectorRef,] }] },
            { type: core$1.TuiTextfieldSizeDirective, decorators: [{ type: core.Inject, args: [core$1.TUI_TEXTFIELD_SIZE,] }] },
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_TIME_TEXTS,] }] }
        ]; };
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTimeComponent.prototype, "disabledItemHandler", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTimeComponent.prototype, "items", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTimeComponent.prototype, "itemSize", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTimeComponent.prototype, "strict", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiInputTimeComponent.prototype, "mode", void 0);
        __decorate([
            core.ViewChild(core$1.TuiPrimitiveTextfieldComponent)
        ], TuiInputTimeComponent.prototype, "textfield", void 0);
        __decorate([
            cdk.tuiPure
        ], TuiInputTimeComponent.prototype, "getFiller$", null);
        __decorate([
            cdk.tuiPure
        ], TuiInputTimeComponent.prototype, "calculateMask", null);
        __decorate([
            cdk.tuiPure
        ], TuiInputTimeComponent.prototype, "filter", null);
        TuiInputTimeComponent = TuiInputTimeComponent_1 = __decorate([
            core.Component({
                selector: 'tui-input-time',
                template: "<tui-hosted-dropdown\n    class=\"wrapper\"\n    [canOpen]=\"interactive && !!filtered.length\"\n    [content]=\"dropdownContent\"\n    [open]=\"open && !!filtered.length\"\n    (openChange)=\"onOpen($event)\"\n    (focusedChange)=\"onFocused($event)\"\n>\n    <tui-primitive-textfield\n        class=\"textfield\"\n        [filler]=\"getFiller$(mode) | async\"\n        [nativeId]=\"nativeId\"\n        [pseudoFocused]=\"innerPseudoFocused\"\n        [pseudoHovered]=\"pseudoHovered\"\n        [pseudoPressed]=\"pseudoPressed\"\n        [invalid]=\"computedInvalid\"\n        [focusable]=\"focusable\"\n        [disabled]=\"disabled\"\n        [readOnly]=\"readOnly\"\n        [textMask]=\"textMaskOptions\"\n        [iconContent]=\"icon\"\n        [value]=\"computedValue\"\n        (valueChange)=\"onValueChange($event)\"\n        (hoveredChange)=\"onHovered($event)\"\n        (keydown.arrowUp)=\"onArrowUp($event)\"\n        (keydown.arrowDown)=\"onArrowDown($event)\"\n    >\n        <ng-content></ng-content>\n    </tui-primitive-textfield>\n</tui-hosted-dropdown>\n<ng-template #dropdownContent>\n    <tui-data-list automation-id=\"tui-input-time__dropdown\">\n        <button\n            *ngFor=\"let item of filtered\"\n            tuiOption\n            automation-id=\"tui-input-time__item\"\n            [size]=\"itemSize\"\n            [disabled]=\"disabledItemHandler(item)\"\n            (click)=\"onMenuClick(item)\"\n        >\n            {{item}}\n        </button>\n    </tui-data-list>\n</ng-template>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                providers: [
                    {
                        provide: cdk.TUI_FOCUSABLE_ITEM_ACCESSOR,
                        useExisting: core.forwardRef(function () { return TuiInputTimeComponent_1; }),
                    },
                    providers.FIXED_DROPDOWN_CONTROLLER_PROVIDER,
                ],
                styles: [":host{display:block;border-radius:var(--tui-radius-m)}:host._disabled{pointer-events:none}.wrapper{display:block;border-radius:inherit}.textfield{border-radius:inherit}"]
            }),
            __param(0, core.Optional()),
            __param(0, core.Self()),
            __param(0, core.Inject(forms.NgControl)),
            __param(1, core.Inject(core.ChangeDetectorRef)),
            __param(2, core.Inject(core$1.TUI_TEXTFIELD_SIZE)),
            __param(3, core.Inject(tokens.TUI_TIME_TEXTS))
        ], TuiInputTimeComponent);
        return TuiInputTimeComponent;
    }(cdk.AbstractTuiNullableControl));

    var TuiInputTimeModule = /** @class */ (function () {
        function TuiInputTimeModule() {
        }
        TuiInputTimeModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    angular2TextMask.TextMaskModule,
                    core$1.TuiDataListModule,
                    core$1.TuiHostedDropdownModule,
                    core$1.TuiPrimitiveTextfieldModule,
                ],
                declarations: [TuiInputTimeComponent],
                exports: [TuiInputTimeComponent],
            })
        ], TuiInputTimeModule);
        return TuiInputTimeModule;
    }());

    exports.TuiInputTimeComponent = TuiInputTimeComponent;
    exports.TuiInputTimeModule = TuiInputTimeModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
