// @taiga-ui/kit/bundles/taiga-ui-kit-components-calendar-month.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@taiga-ui/cdk'), require('@taiga-ui/kit/tokens'), require('rxjs'), require('@angular/common'), require('@taiga-ui/core')) :
    typeof define === 'function' && define.amd ? define('@taiga-ui/kit/components/calendar-month', ['exports', '@angular/core', '@taiga-ui/cdk', '@taiga-ui/kit/tokens', 'rxjs', '@angular/common', '@taiga-ui/core'], factory) :
    (global = global || self, factory((global['taiga-ui'] = global['taiga-ui'] || {}, global['taiga-ui'].kit = global['taiga-ui'].kit || {}, global['taiga-ui'].kit.components = global['taiga-ui'].kit.components || {}, global['taiga-ui'].kit.components['calendar-month'] = {}), global.ng.core, global.cdk, global['taiga-ui'].kit.tokens, global.rxjs, global.ng.common, global.core$1));
}(this, (function (exports, core, cdk, tokens, rxjs, common, core$1) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation.

    Permission to use, copy, modify, and/or distribute this software for any
    purpose with or without fee is hereby granted.

    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
    REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
    AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
    INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
    LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
    OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
    PERFORMANCE OF THIS SOFTWARE.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __createBinding(o, m, k, k2) {
        if (k2 === undefined) k2 = k;
        o[k2] = m[k];
    }

    function __exportStar(m, exports) {
        for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
        if (m) return m.call(o);
        if (o && typeof o.length === "number") return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
        throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    function __classPrivateFieldGet(receiver, privateMap) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to get private field on non-instance");
        }
        return privateMap.get(receiver);
    }

    function __classPrivateFieldSet(receiver, privateMap, value) {
        if (!privateMap.has(receiver)) {
            throw new TypeError("attempted to set private field on non-instance");
        }
        privateMap.set(receiver, value);
        return value;
    }

    var TODAY = cdk.TuiDay.currentLocal();
    // @dynamic
    var TuiCalendarMonthComponent = /** @class */ (function () {
        function TuiCalendarMonthComponent(months$) {
            this.months$ = months$;
            this.value = null;
            this.year = TODAY;
            this.disabledItemHandler = cdk.ALWAYS_FALSE_HANDLER;
            this.min = cdk.TUI_FIRST_DAY;
            this.max = cdk.TUI_LAST_DAY;
            this.monthClick = new core.EventEmitter();
            this.hoveredItemChange = new core.EventEmitter();
            this.yearChange = new core.EventEmitter();
            this.isYearPickerShown = false;
            this.hoveredItem = null;
            this.pressedItem = null;
        }
        Object.defineProperty(TuiCalendarMonthComponent.prototype, "isSingle", {
            get: function () {
                return (this.value !== null &&
                    (this.value instanceof cdk.TuiMonth || this.value.isSingleMonth));
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiCalendarMonthComponent.prototype, "previousYearDisabled", {
            get: function () {
                return this.year.yearSameOrBefore(this.min);
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(TuiCalendarMonthComponent.prototype, "nextYearDisabled", {
            get: function () {
                return this.year.yearSameOrAfter(this.max);
            },
            enumerable: true,
            configurable: true
        });
        TuiCalendarMonthComponent.prototype.getItemState = function (item) {
            var _a = this, disabledItemHandlerWithMinMax = _a.disabledItemHandlerWithMinMax, pressedItem = _a.pressedItem, hoveredItem = _a.hoveredItem;
            if (disabledItemHandlerWithMinMax(item)) {
                return "disabled" /* Disabled */;
            }
            if (pressedItem !== null && pressedItem.monthSame(item)) {
                return "pressed" /* Pressed */;
            }
            if (hoveredItem !== null && hoveredItem.monthSame(item)) {
                return "hovered" /* Hovered */;
            }
            return null;
        };
        TuiCalendarMonthComponent.prototype.getItemRange = function (item) {
            var _a = this, value = _a.value, hoveredItem = _a.hoveredItem;
            if (value === null) {
                return null;
            }
            if (value instanceof cdk.TuiMonth) {
                return value.monthSame(item) ? "single" /* Single */ : null;
            }
            var theFirstOfRange = value.from.monthSame(item) && !value.isSingleMonth;
            var hoveredItemAfterFrom = hoveredItem !== null &&
                hoveredItem.monthAfter(value.from) &&
                value.from.monthSame(item) &&
                value.isSingleMonth;
            var hoveredItemIsCandidatToBeFrom = hoveredItem !== null &&
                hoveredItem.monthSame(item) &&
                hoveredItem.monthBefore(value.from) &&
                value.isSingleMonth;
            if (theFirstOfRange || hoveredItemAfterFrom || hoveredItemIsCandidatToBeFrom) {
                return "start" /* Start */;
            }
            var theLastOfRange = value.to.monthSame(item) && !value.isSingleMonth;
            var hoveredItemBeforeTo = value.to.monthSame(item) && !value.isSingleMonth;
            var hoveredItemIsCandidatToBeTo = hoveredItem !== null &&
                hoveredItem.monthSame(item) &&
                hoveredItem.monthAfter(value.from) &&
                value.isSingleMonth;
            if (theLastOfRange || hoveredItemBeforeTo || hoveredItemIsCandidatToBeTo) {
                return "end" /* End */;
            }
            return value.isSingleMonth && value.from.monthSame(item)
                ? "single" /* Single */
                : null;
        };
        TuiCalendarMonthComponent.prototype.getTuiMonth = function (monthNumber, yearNumber) {
            return new cdk.TuiMonth(yearNumber, monthNumber);
        };
        TuiCalendarMonthComponent.prototype.isItemToday = function (item) {
            return TODAY.monthSame(item);
        };
        TuiCalendarMonthComponent.prototype.isItemInsideRange = function (month) {
            var _a = this, value = _a.value, hoveredItem = _a.hoveredItem;
            if (value === null || value instanceof cdk.TuiMonth) {
                return false;
            }
            if (!value.isSingleMonth) {
                return value.from.monthSameOrBefore(month) && value.to.monthAfter(month);
            }
            if (hoveredItem === null) {
                return false;
            }
            var range = cdk.TuiMonthRange.sort(value.from, hoveredItem);
            return range.from.monthSameOrBefore(month) && range.to.monthAfter(month);
        };
        TuiCalendarMonthComponent.prototype.onPickerYearClick = function (year) {
            this.isYearPickerShown = false;
            if (this.year.yearSame(year)) {
                return;
            }
            this.year = year;
            this.yearChange.emit(year);
        };
        TuiCalendarMonthComponent.prototype.onItemClick = function (month) {
            if (this.disabledItemHandlerWithMinMax(month)) {
                return;
            }
            this.monthClick.emit(month);
        };
        TuiCalendarMonthComponent.prototype.onYearClick = function () {
            this.isYearPickerShown = true;
        };
        TuiCalendarMonthComponent.prototype.onNextYear = function () {
            this.year = this.year.append({ year: 1 });
        };
        TuiCalendarMonthComponent.prototype.onPreviousYear = function () {
            this.year = this.year.append({ year: -1 });
        };
        TuiCalendarMonthComponent.prototype.onItemHovered = function (hovered, item) {
            this.updateHoveredItem(hovered ? item : null);
        };
        TuiCalendarMonthComponent.prototype.onItemPressed = function (pressed, item) {
            this.updatePressedItem(pressed ? item : null);
        };
        Object.defineProperty(TuiCalendarMonthComponent.prototype, "disabledItemHandlerWithMinMax", {
            get: function () {
                return this.calculateDisabledItemHandlerWithMinMax(this.disabledItemHandler, this.value, this.min, this.max);
            },
            enumerable: true,
            configurable: true
        });
        TuiCalendarMonthComponent.prototype.updateHoveredItem = function (month) {
            if (cdk.nullableSame(this.hoveredItem, month, function (a, b) { return a.monthSame(b); })) {
                return;
            }
            this.hoveredItem = month;
            this.hoveredItemChange.emit(month);
        };
        TuiCalendarMonthComponent.prototype.updatePressedItem = function (item) {
            this.pressedItem = item;
        };
        TuiCalendarMonthComponent.prototype.calculateDisabledItemHandlerWithMinMax = function (disabledItemHandler, value, min, max) {
            return function (item) {
                return item.monthBefore(min) ||
                    item.monthAfter(max) ||
                    disabledItemHandler(item, { value: value });
            };
        };
        TuiCalendarMonthComponent.ctorParameters = function () { return [
            { type: rxjs.Observable, decorators: [{ type: core.Inject, args: [tokens.TUI_CALENDAR_MONTHS,] }] }
        ]; };
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiCalendarMonthComponent.prototype, "value", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiCalendarMonthComponent.prototype, "year", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiCalendarMonthComponent.prototype, "disabledItemHandler", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiCalendarMonthComponent.prototype, "min", void 0);
        __decorate([
            core.Input(),
            cdk.tuiDefaultProp()
        ], TuiCalendarMonthComponent.prototype, "max", void 0);
        __decorate([
            core.Output()
        ], TuiCalendarMonthComponent.prototype, "monthClick", void 0);
        __decorate([
            core.Output()
        ], TuiCalendarMonthComponent.prototype, "hoveredItemChange", void 0);
        __decorate([
            core.Output()
        ], TuiCalendarMonthComponent.prototype, "yearChange", void 0);
        __decorate([
            core.HostBinding('class._single')
        ], TuiCalendarMonthComponent.prototype, "isSingle", null);
        __decorate([
            cdk.tuiPure
        ], TuiCalendarMonthComponent.prototype, "calculateDisabledItemHandlerWithMinMax", null);
        TuiCalendarMonthComponent = __decorate([
            core.Component({
                selector: 'tui-calendar-month',
                template: "<tui-scrollbar *ngIf=\"isYearPickerShown; else monthSelect\" class=\"scrollbar\">\n    <tui-primitive-year-picker\n        [min]=\"min\"\n        [max]=\"max\"\n        [initialItem]=\"year\"\n        [value]=\"value\"\n        (yearClick)=\"onPickerYearClick($event)\"\n    ></tui-primitive-year-picker>\n</tui-scrollbar>\n<ng-template #monthSelect>\n    <tui-primitive-spin-button\n        [focusable]=\"false\"\n        [leftDisabled]=\"previousYearDisabled\"\n        [rightDisabled]=\"nextYearDisabled\"\n        (leftClick)=\"onPreviousYear()\"\n        (rightClick)=\"onNextYear()\"\n    >\n        <button tuiLink [tuiFocusable]=\"false\" (click)=\"onYearClick()\">\n            {{year.formattedYear}}\n        </button>\n    </tui-primitive-spin-button>\n    <div class=\"row\">\n        <ng-container\n            *ngFor=\"let month of (months$ | async); let index = index\"\n        >\n            <div\n                *tuiLet=\"getTuiMonth(index, year.year) as item\"\n                class=\"cell\"\n                [class.cell_today]=\"isItemToday(item)\"\n                [class.cell_interval]=\"isItemInsideRange(item)\"\n                [attr.data-tui-element-range]=\"getItemRange(item)\"\n                [attr.data-tui-element-state]=\"getItemState(item)\"\n                (tuiHoveredChange)=\"onItemHovered($event, item)\"\n                (tuiPressedChange)=\"onItemHovered($event, item)\"\n                (click)=\"onItemClick(item)\"\n            >\n                <div class=\"item\">{{month}}</div>\n            </div>\n        </ng-container>\n    </div>\n</ng-template>\n",
                changeDetection: core.ChangeDetectionStrategy.OnPush,
                styles: [":host{font:var(--tui-font-text-m)}.row{position:relative;z-index:0;display:flex;justify-content:space-between;height:36px}.item{position:relative;flex:1;line-height:32px;border-radius:var(--tui-radius-m)}.item:after,.item:before{content:'';position:absolute;top:0;left:0;right:0;bottom:0;z-index:-1;border-radius:var(--tui-radius-m)}.cell{position:relative;display:flex;align-items:center;justify-content:center;width:59px;text-align:center;outline:0;cursor:pointer;background-clip:content-box;box-sizing:border-box;border:2px solid transparent;box-sizing:content-box}.cell:before{content:'';position:absolute;top:0;left:0;right:0;bottom:0;z-index:-1;border-radius:var(--tui-radius-m)}.cell_today:after{position:absolute;left:50%;transform:translate(-50%,0);content:'';bottom:5px;height:2px;width:12px;border-radius:6px;background-color:var(--tui-text-02)}.cell_interval:before{background:var(--tui-base-02)}:host._single .cell_interval:before{background:var(--tui-secondary-hover)}.cell_interval:not(:last-child):before{right:-59px}.cell_interval:last-child:first-child:before{right:0}.cell_interval:first-child>.item{border-top-left-radius:var(--tui-radius-m);border-bottom-left-radius:var(--tui-radius-m)}.cell_interval:last-child>.item{border-top-right-radius:var(--tui-radius-m);border-bottom-right-radius:var(--tui-radius-m)}.cell_interval>.item{border-radius:0}.cell[data-tui-element-range]:after{background-color:var(--tui-primary-text)}.cell[data-tui-element-range]>.item{color:var(--tui-primary-text)}.cell[data-tui-element-range]>.item:after,.cell[data-tui-element-range]>.item:before{background-color:var(--tui-primary)}.cell[data-tui-element-range][data-tui-element-state=hovered]>.item:after,.cell[data-tui-element-range][data-tui-element-state=hovered]>.item:before{background-color:var(--tui-primary-hover)}.cell[data-tui-element-range][data-tui-element-state=pressed]>.item:after,.cell[data-tui-element-range][data-tui-element-state=pressed]>.item:before{background-color:var(--tui-primary-active)}.cell[data-tui-element-range=end]>.item:before{left:4px}.cell[data-tui-element-range=end]>.item:after{left:-32px;right:100%;transform:translateX(23px) scaleY(.6) scaleX(.4) rotate(45deg)}.cell[data-tui-element-range=start]>.item:before{right:4px}.cell[data-tui-element-range=start]>.item:after{left:100%;right:-32px;transform:translateX(-23px) scaleY(.6) scaleX(.4) rotate(45deg)}.cell[data-tui-element-state=disabled]{pointer-events:none}.cell[data-tui-element-state=disabled]>.item{opacity:.36}.cell[data-tui-element-state=hovered]:hover:not([data-tui-element-range])>.item{background-color:var(--tui-secondary-hover)}.cell[data-tui-element-state=pressed]:hover:not([data-tui-element-range])>.item{background-color:var(--tui-secondary-active)}:host{display:block;height:218px;width:252px;padding:18px;box-sizing:content-box}.row{flex-wrap:wrap;margin-top:23px}.cell:nth-child(n+5){margin-top:28px}.cell_interval:nth-child(4n):before{right:0}.scrollbar{height:inherit;width:inherit}"]
            }),
            __param(0, core.Inject(tokens.TUI_CALENDAR_MONTHS))
        ], TuiCalendarMonthComponent);
        return TuiCalendarMonthComponent;
    }());

    var TuiCalendarMonthModule = /** @class */ (function () {
        function TuiCalendarMonthModule() {
        }
        TuiCalendarMonthModule = __decorate([
            core.NgModule({
                imports: [
                    common.CommonModule,
                    core$1.TuiPrimitiveYearPickerModule,
                    core$1.TuiPrimitiveSpinButtonModule,
                    core$1.TuiScrollbarModule,
                    core$1.TuiLinkModule,
                    cdk.TuiLetModule,
                    cdk.TuiHoveredModule,
                    cdk.TuiPressedModule,
                    cdk.TuiFocusableModule,
                ],
                declarations: [TuiCalendarMonthComponent],
                exports: [TuiCalendarMonthComponent],
            })
        ], TuiCalendarMonthModule);
        return TuiCalendarMonthModule;
    }());

    exports.TuiCalendarMonthComponent = TuiCalendarMonthComponent;
    exports.TuiCalendarMonthModule = TuiCalendarMonthModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
