// @taiga-ui/cdk/esm2015/directives/obscured/obscured.directive.js
import { __decorate, __param } from "tslib";
import { Directive, Inject, Input, Optional, Output, Self } from '@angular/core';
import { tuiRequiredSetter } from '@taiga-ui/cdk/decorators';
import { TuiActiveZoneDirective } from '@taiga-ui/cdk/directives/active-zone';
import { TuiDestroyService, TuiObscuredService, TuiParentsScrollService, } from '@taiga-ui/cdk/services';
import { EMPTY, Subject } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
/**
 * Directive that monitors element visibility
 */
let TuiObscuredDirective = class TuiObscuredDirective {
    constructor(activeZone, obscured$) {
        this.enabled$ = new Subject();
        const mapped$ = obscured$.pipe(map(obscuredBy => !!obscuredBy &&
            (!activeZone ||
                !obscuredBy.length ||
                obscuredBy.every(element => !activeZone.contains(element)))));
        this.tuiObscured = this.enabled$.pipe(switchMap(enabled => (enabled ? mapped$ : EMPTY)));
    }
    set tuiObscuredEnabled(enabled) {
        this.enabled$.next(enabled);
    }
};
TuiObscuredDirective.ctorParameters = () => [
    { type: TuiActiveZoneDirective, decorators: [{ type: Optional }, { type: Inject, args: [TuiActiveZoneDirective,] }] },
    { type: TuiObscuredService, decorators: [{ type: Self }, { type: Inject, args: [TuiObscuredService,] }] }
];
__decorate([
    Input(),
    tuiRequiredSetter()
], TuiObscuredDirective.prototype, "tuiObscuredEnabled", null);
__decorate([
    Output()
], TuiObscuredDirective.prototype, "tuiObscured", void 0);
TuiObscuredDirective = __decorate([
    Directive({
        selector: '[tuiObscured]',
        providers: [TuiObscuredService, TuiParentsScrollService, TuiDestroyService],
    }),
    __param(0, Optional()),
    __param(0, Inject(TuiActiveZoneDirective)),
    __param(1, Self()),
    __param(1, Inject(TuiObscuredService))
], TuiObscuredDirective);
export { TuiObscuredDirective };
