// @taiga-ui/cdk/fesm5/taiga-ui-cdk-directives-focusable.js
import { __decorate } from 'tslib';
import { Input, HostBinding, Directive, NgModule } from '@angular/core';

/**
 * Abstraction over `tabindex`
 */
var TuiFocusableDirective = /** @class */ (function () {
    function TuiFocusableDirective() {
        /**
         * Element can be focused with keyboard
         */
        this.focusable = true;
    }
    Object.defineProperty(TuiFocusableDirective.prototype, "tabIndex", {
        get: function () {
            return this.focusable ? 0 : -1;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        Input('tuiFocusable')
    ], TuiFocusableDirective.prototype, "focusable", void 0);
    __decorate([
        HostBinding('tabIndex')
    ], TuiFocusableDirective.prototype, "tabIndex", null);
    TuiFocusableDirective = __decorate([
        Directive({
            selector: '[tuiFocusable]',
        })
    ], TuiFocusableDirective);
    return TuiFocusableDirective;
}());

var TuiFocusableModule = /** @class */ (function () {
    function TuiFocusableModule() {
    }
    TuiFocusableModule = __decorate([
        NgModule({
            declarations: [TuiFocusableDirective],
            exports: [TuiFocusableDirective],
        })
    ], TuiFocusableModule);
    return TuiFocusableModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiFocusableDirective, TuiFocusableModule };
