// @taiga-ui/cdk/fesm5/taiga-ui-cdk-directives-drag.js
import { __decorate, __param } from 'tslib';
import { ElementRef, Inject, Output, Directive, NgModule } from '@angular/core';
import { dragAndDropFrom } from '@taiga-ui/cdk/observables';
import { filter, map } from 'rxjs/operators';

var TuiDragDirective = /** @class */ (function () {
    function TuiDragDirective(elementRef) {
        this.elementRef = elementRef;
        this.dragAndDropFrom$ = dragAndDropFrom(this.elementRef.nativeElement);
        this.start = this.dragAndDropFrom$.pipe(filter(function (_a) {
            var stage = _a.stage;
            return stage === 0 /* Start */;
        }), map(function (_a) {
            var event = _a.event;
            return event;
        }));
        this.continues = this.dragAndDropFrom$.pipe(filter(function (_a) {
            var stage = _a.stage;
            return stage === 1 /* Continues */;
        }), map(function (_a) {
            var event = _a.event;
            return event;
        }));
        this.end = this.dragAndDropFrom$.pipe(filter(function (_a) {
            var stage = _a.stage;
            return stage === 2 /* End */;
        }), map(function (_a) {
            var event = _a.event;
            return event;
        }));
    }
    TuiDragDirective.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] }
    ]; };
    __decorate([
        Output('tuiDragStart')
    ], TuiDragDirective.prototype, "start", void 0);
    __decorate([
        Output('tuiDragContinues')
    ], TuiDragDirective.prototype, "continues", void 0);
    __decorate([
        Output('tuiDragEnd')
    ], TuiDragDirective.prototype, "end", void 0);
    TuiDragDirective = __decorate([
        Directive({
            selector: '[tuiDragStart], [tuiDragContinues], [tuiDragEnd]',
        }),
        __param(0, Inject(ElementRef))
    ], TuiDragDirective);
    return TuiDragDirective;
}());

var TuiDragModule = /** @class */ (function () {
    function TuiDragModule() {
    }
    TuiDragModule = __decorate([
        NgModule({
            declarations: [TuiDragDirective],
            exports: [TuiDragDirective],
        })
    ], TuiDragModule);
    return TuiDragModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiDragDirective, TuiDragModule };
