// @taiga-ui/cdk/fesm5/taiga-ui-cdk-directives-resize.js
import { __decorate, __param } from 'tslib';
import { Inject, Directive, NgModule } from '@angular/core';
import { TuiResizeService, TuiDestroyService } from '@taiga-ui/cdk/services';
import { Observable } from 'rxjs';

// @dynamic
var TuiResizeDirective = /** @class */ (function () {
    function TuiResizeDirective(tuiResize) {
        this.tuiResize = tuiResize;
    }
    TuiResizeDirective.ctorParameters = function () { return [
        { type: Observable, decorators: [{ type: Inject, args: [TuiResizeService,] }] }
    ]; };
    TuiResizeDirective = __decorate([
        Directive({
            selector: '[tuiResize]',
            outputs: ['tuiResize'],
            providers: [TuiDestroyService, TuiResizeService],
        }),
        __param(0, Inject(TuiResizeService))
    ], TuiResizeDirective);
    return TuiResizeDirective;
}());

/**
 * @experimental
 */
var TuiResizeModule = /** @class */ (function () {
    function TuiResizeModule() {
    }
    TuiResizeModule = __decorate([
        NgModule({
            declarations: [TuiResizeDirective],
            exports: [TuiResizeDirective],
        })
    ], TuiResizeModule);
    return TuiResizeModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiResizeDirective, TuiResizeModule };
