// @taiga-ui/cdk/fesm5/taiga-ui-cdk-pipes-filter.js
import { __spread, __decorate } from 'tslib';
import { Pipe, NgModule } from '@angular/core';

var TuiFilterPipe = /** @class */ (function () {
    function TuiFilterPipe() {
    }
    /**
     * Filters an array through a matcher function using additional arguments
     *
     * @param items array
     * @param matcher method for filtering
     * @param args arbitrary number of additional arguments
     */
    TuiFilterPipe.prototype.transform = function (items, matcher) {
        var args = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            args[_i - 2] = arguments[_i];
        }
        return items.filter(function (item) { return matcher.apply(void 0, __spread([item], args)); });
    };
    TuiFilterPipe = __decorate([
        Pipe({ name: 'tuiFilter' })
    ], TuiFilterPipe);
    return TuiFilterPipe;
}());

var TuiFilterPipeModule = /** @class */ (function () {
    function TuiFilterPipeModule() {
    }
    TuiFilterPipeModule = __decorate([
        NgModule({
            exports: [TuiFilterPipe],
            declarations: [TuiFilterPipe],
        })
    ], TuiFilterPipeModule);
    return TuiFilterPipeModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiFilterPipe, TuiFilterPipeModule };
