// @taiga-ui/cdk/fesm5/taiga-ui-cdk-directives-pressed.js
import { __decorate, __param } from 'tslib';
import { ElementRef, Inject, Output, Directive, NgModule } from '@angular/core';
import { pressedObservable } from '@taiga-ui/cdk/observables';

var TuiPressedDirective = /** @class */ (function () {
    function TuiPressedDirective(elementRef) {
        this.elementRef = elementRef;
        this.tuiPressedChange = pressedObservable(this.elementRef.nativeElement);
    }
    TuiPressedDirective.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] }
    ]; };
    __decorate([
        Output()
    ], TuiPressedDirective.prototype, "tuiPressedChange", void 0);
    TuiPressedDirective = __decorate([
        Directive({
            selector: '[tuiPressedChange]',
        }),
        __param(0, Inject(ElementRef))
    ], TuiPressedDirective);
    return TuiPressedDirective;
}());

var TuiPressedModule = /** @class */ (function () {
    function TuiPressedModule() {
    }
    TuiPressedModule = __decorate([
        NgModule({
            declarations: [TuiPressedDirective],
            exports: [TuiPressedDirective],
        })
    ], TuiPressedModule);
    return TuiPressedModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiPressedDirective, TuiPressedModule };
