// @taiga-ui/cdk/fesm5/taiga-ui-cdk-directives-focused.js
import { __decorate, __param } from 'tslib';
import { ElementRef, Inject, NgZone, Output, Directive, NgModule } from '@angular/core';
import { typedFromEvent, tuiZoneOptimized } from '@taiga-ui/cdk/observables';
import { isNativeFocused } from '@taiga-ui/cdk/utils/focus';
import { merge } from 'rxjs';
import { map, startWith, distinctUntilChanged, skip } from 'rxjs/operators';

/**
 * Directive to monitor focus/blur status, works with focusIn/focus-out
 * instead of focus/blur to sync events order with Internet Explorer and
 * other focus related directives that require bubbling
 */
var TuiFocusedDirective = /** @class */ (function () {
    function TuiFocusedDirective(_a, ngZone) {
        var nativeElement = _a.nativeElement;
        this.tuiFocusedChange = merge(typedFromEvent(nativeElement, 'focusin'), typedFromEvent(nativeElement, 'focusout')).pipe(map(function () { return isNativeFocused(nativeElement); }), startWith(false), distinctUntilChanged(), skip(1), tuiZoneOptimized(ngZone));
    }
    TuiFocusedDirective.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: NgZone, decorators: [{ type: Inject, args: [NgZone,] }] }
    ]; };
    __decorate([
        Output()
    ], TuiFocusedDirective.prototype, "tuiFocusedChange", void 0);
    TuiFocusedDirective = __decorate([
        Directive({
            selector: '[tuiFocusedChange]',
        }),
        __param(0, Inject(ElementRef)),
        __param(1, Inject(NgZone))
    ], TuiFocusedDirective);
    return TuiFocusedDirective;
}());

var TuiFocusedModule = /** @class */ (function () {
    function TuiFocusedModule() {
    }
    TuiFocusedModule = __decorate([
        NgModule({
            declarations: [TuiFocusedDirective],
            exports: [TuiFocusedDirective],
        })
    ], TuiFocusedModule);
    return TuiFocusedModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiFocusedDirective, TuiFocusedModule };
