// @taiga-ui/kit/esm5/components/input-phone/input-phone.providers.js
import { DOCUMENT } from '@angular/common';
import { InjectionToken } from '@angular/core';
import { TuiDestroyService } from '@taiga-ui/cdk';
import { FIXED_DROPDOWN_CONTROLLER_PROVIDER } from '@taiga-ui/kit/providers';
import { fromEvent, merge } from 'rxjs';
import { flatMap, startWith, takeUntil, windowToggle } from 'rxjs/operators';
export var SELECTION_STREAM = new InjectionToken('A stream of selection changes');
export var INPUT_PHONE_PROVIDERS = [
    TuiDestroyService,
    FIXED_DROPDOWN_CONTROLLER_PROVIDER,
    {
        provide: SELECTION_STREAM,
        deps: [TuiDestroyService, DOCUMENT],
        useFactory: selectionStreamFactory,
    },
];
export function selectionStreamFactory(destroy$, documentRef) {
    return fromEvent(documentRef, 'selectionchange').pipe(windowToggle(merge(fromEvent(documentRef, 'mouseup'), fromEvent(documentRef, 'keydown')), function () { return fromEvent(documentRef, 'mousedown'); }), flatMap(function (windowed$) { return windowed$.pipe(startWith(null)); }), takeUntil(destroy$));
}
