// @taiga-ui/core/esm2015/components/dialog/dialog.service.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { AbstractTuiDialogService } from '@taiga-ui/cdk';
import { PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { TuiDialogComponent } from './dialog.component';
import * as i0 from "@angular/core";
import * as i1 from "@taiga-ui/cdk/services";
const DIALOG = new PolymorpheusComponent(TuiDialogComponent);
const ɵ0 = undefined;
const DEFAULT_OPTIONS = {
    size: 'm',
    required: false,
    closeable: true,
    dismissible: true,
    label: '',
    header: '',
    data: ɵ0,
};
let TuiDialogService = class TuiDialogService extends AbstractTuiDialogService {
    constructor() {
        super(...arguments);
        this.component = DIALOG;
        this.defaultOptions = DEFAULT_OPTIONS;
    }
};
TuiDialogService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TuiDialogService_Factory() { return new TuiDialogService(i0.ɵɵinject(i1.TuiIdService)); }, token: TuiDialogService, providedIn: "root" });
TuiDialogService = __decorate([
    Injectable({
        providedIn: 'root',
    })
], TuiDialogService);
export { TuiDialogService };
export { ɵ0 };
