// @taiga-ui/core/esm2015/components/dialog/dialog.providers.js
import { DOCUMENT } from '@angular/common';
import { ElementRef, InjectionToken } from '@angular/core';
import { WINDOW } from '@ng-web-apis/common';
import { containsOrAfter, isCurrentTarget, TuiDestroyService, typedFromEvent, } from '@taiga-ui/cdk';
import { POLYMORPHEUS_CONTEXT } from '@tinkoff/ng-polymorpheus';
import { EMPTY, merge } from 'rxjs';
import { filter, switchMapTo, take, takeUntil } from 'rxjs/operators';
export const TUI_DIALOGS_CLOSE = new InjectionToken('A stream to close dialogs', {
    factory: () => EMPTY,
});
const SCROLLBAR_PLACEHOLDER = 17;
export function dialogCloseStreamFactory(documentRef, windowRef, { nativeElement }, close$, destroy$, { dismissible }) {
    return dismissible
        ? merge(typedFromEvent(nativeElement, 'click').pipe(filter(isCurrentTarget)), typedFromEvent(documentRef, 'keydown').pipe(filter(({ key, target }) => key === 'Escape' &&
            target instanceof Element &&
            (!containsOrAfter(nativeElement, target) ||
                nativeElement.contains(target)))), typedFromEvent(documentRef, 'mousedown').pipe(filter(({ target, clientX }) => target instanceof Element &&
            windowRef.innerWidth - clientX > SCROLLBAR_PLACEHOLDER &&
            !containsOrAfter(nativeElement, target)), switchMapTo(typedFromEvent(documentRef, 'mouseup').pipe(take(1), filter(({ target }) => target instanceof Element &&
            !containsOrAfter(nativeElement, target))))), close$).pipe(takeUntil(destroy$))
        : close$;
}
export const TUI_DIALOG_CLOSE_STREAM = new InjectionToken('Dialogs closing stream');
export const TUI_DIALOG_PROVIDERS = [
    TuiDestroyService,
    {
        provide: TUI_DIALOG_CLOSE_STREAM,
        deps: [
            DOCUMENT,
            WINDOW,
            ElementRef,
            TUI_DIALOGS_CLOSE,
            TuiDestroyService,
            POLYMORPHEUS_CONTEXT,
        ],
        useFactory: dialogCloseStreamFactory,
    },
];
