// @taiga-ui/core/esm5/modules/notifications/notifications-host/notifications-host.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, Inject } from '@angular/core';
import { tuiFadeInList, tuiHeightCollapseList, tuiSlideInRightList, } from '@taiga-ui/core/animations';
import { TuiNotificationsService } from '../notifications.service';
var TuiNotificationsHostComponent = /** @class */ (function () {
    function TuiNotificationsHostComponent(service) {
        this.service = service;
    }
    TuiNotificationsHostComponent.ctorParameters = function () { return [
        { type: TuiNotificationsService, decorators: [{ type: Inject, args: [TuiNotificationsService,] }] }
    ]; };
    TuiNotificationsHostComponent = __decorate([
        Component({
            selector: 'tui-notifications-host',
            changeDetection: ChangeDetectionStrategy.OnPush,
            template: "<ng-container *tuiLet=\"service.items$ | async as items\">\n    <div\n        *ngIf=\"items.length\"\n        class=\"wrapper\"\n        [@tuiFadeInList]=\"items.length\"\n        [@tuiHeightCollapseList]=\"items.length\"\n        [@tuiSlideInRightList]=\"items.length\"\n    >\n        <tui-notification-alert\n            *ngFor=\"let item of items\"\n            class=\"alert\"\n            role=\"alert\"\n            [item]=\"item\"\n        ></tui-notification-alert>\n    </div>\n</ng-container>\n",
            animations: [tuiFadeInList, tuiSlideInRightList, tuiHeightCollapseList],
            styles: [":host{position:fixed;top:0;left:0;width:100%;height:0}.wrapper{width:288px;margin-left:auto;padding:32px 48px 0 0}@media screen and (max-width:1023px){.wrapper{padding:16px 16px 0 0}}@media screen and (min-width:1360px){.wrapper{width:368px}}.alert:not(:first-child){margin-top:12px}"]
        }),
        __param(0, Inject(TuiNotificationsService))
    ], TuiNotificationsHostComponent);
    return TuiNotificationsHostComponent;
}());
export { TuiNotificationsHostComponent };
