// @taiga-ui/core/esm5/components/scroll-controls/scroll-controls.component.js
import { __decorate, __param } from "tslib";
import { DOCUMENT } from '@angular/common';
import { ChangeDetectionStrategy, Component, ElementRef, Inject, NgZone, Optional, } from '@angular/core';
import { tuiZoneOptimized } from '@taiga-ui/cdk';
import { tuiFadeIn } from '@taiga-ui/core/animations';
import { MODE_PROVIDER } from '@taiga-ui/core/providers';
import { TUI_MODE, TUI_SCROLL_REF } from '@taiga-ui/core/tokens';
import { interval, Observable } from 'rxjs';
import { distinctUntilChanged, map, startWith } from 'rxjs/operators';
// @bad TODO: handle click on bar to scroll to that position
// @dynamic
var TuiScrollControlsComponent = /** @class */ (function () {
    function TuiScrollControlsComponent(ngZone, documentRef, scrollRef, mode$) {
        var _this = this;
        this.ngZone = ngZone;
        this.documentRef = documentRef;
        this.scrollRef = scrollRef;
        this.mode$ = mode$;
        this.refresh$ = interval(300).pipe(map(function () { return _this.scrollbars; }), startWith([false, false]), distinctUntilChanged(function (a, b) { return a[0] === b[0] && a[1] === b[1]; }), tuiZoneOptimized(this.ngZone));
    }
    Object.defineProperty(TuiScrollControlsComponent.prototype, "scrollbars", {
        get: function () {
            var _a = this.scrollRef
                ? this.scrollRef.nativeElement
                : this.documentRef.documentElement, clientHeight = _a.clientHeight, scrollHeight = _a.scrollHeight, clientWidth = _a.clientWidth, scrollWidth = _a.scrollWidth;
            return [
                Math.ceil((clientHeight / scrollHeight) * 100) < 100,
                Math.ceil((clientWidth / scrollWidth) * 100) < 100,
            ];
        },
        enumerable: true,
        configurable: true
    });
    TuiScrollControlsComponent.ctorParameters = function () { return [
        { type: NgZone, decorators: [{ type: Inject, args: [NgZone,] }] },
        { type: Document, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: ElementRef, decorators: [{ type: Optional }, { type: Inject, args: [TUI_SCROLL_REF,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TUI_MODE,] }] }
    ]; };
    TuiScrollControlsComponent = __decorate([
        Component({
            selector: 'tui-scroll-controls',
            template: "<ng-container *tuiLet=\"refresh$ | async as bars\">\n    <div\n        *ngIf=\"bars[0]\"\n        class=\"bar bar_vertical\"\n        [class.bar_has-horizontal]=\"bars[1]\"\n        [@tuiFadeIn]\n    >\n        <div tuiScrollbar=\"vertical\" class=\"thumb\"></div>\n    </div>\n    <div\n        *ngIf=\"bars[1]\"\n        class=\"bar bar_horizontal\"\n        [class.bar_has-vertical]=\"bars[0]\"\n        [@tuiFadeIn]\n    >\n        <div tuiScrollbar=\"horizontal\" class=\"thumb\"></div>\n    </div>\n</ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            animations: [tuiFadeIn],
            providers: [MODE_PROVIDER],
            host: {
                '($.data-mode.attr)': 'mode$',
            },
            styles: [":host{position:-webkit-sticky;position:sticky;top:0;left:0;z-index:100;min-width:calc(100% - 1px);min-height:calc(100% - 1px);float:left;margin-right:calc(-100% + 1px);pointer-events:none}.bar{position:absolute;right:0;bottom:0}.bar_vertical{top:0}.bar_horizontal{left:0}.bar_has-horizontal{bottom:8px}.bar_has-vertical{right:8px}.thumb{transition-duration:.15s;transition-timing-function:ease-in-out;position:absolute;border-radius:100px;border:4px solid transparent;cursor:pointer;pointer-events:auto;background:var(--tui-text-01);background-clip:content-box;box-sizing:border-box;transition-property:width,height;opacity:.2}.thumb:hover{opacity:.24}.thumb:active{opacity:.48}:host[data-mode=onDark] .thumb{background-color:var(--tui-text-01-night)}.bar_vertical .thumb{right:0;width:12px;min-height:24px}.bar_vertical .thumb:active,.bar_vertical:hover .thumb{width:14px}.bar_horizontal .thumb{bottom:0;height:12px;min-width:24px}.bar_horizontal .thumb:active,.bar_horizontal:hover .thumb{height:14px}"]
        }),
        __param(0, Inject(NgZone)),
        __param(1, Inject(DOCUMENT)),
        __param(2, Optional()),
        __param(2, Inject(TUI_SCROLL_REF)),
        __param(3, Inject(TUI_MODE))
    ], TuiScrollControlsComponent);
    return TuiScrollControlsComponent;
}());
export { TuiScrollControlsComponent };
