// @taiga-ui/kit/esm5/components/breadcrumbs/breadcrumbs.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, HostBinding, Input } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import { DEFAULT_ROUTER_LINK_OPTIONS } from '@taiga-ui/kit/constants';
var TuiBreadcrumbsComponent = /** @class */ (function () {
    function TuiBreadcrumbsComponent() {
        this.items = [];
        this.size = 'm';
        this.routerLinkOptions = function (_a) {
            var routerLinkOptions = _a.routerLinkOptions;
            return routerLinkOptions || DEFAULT_ROUTER_LINK_OPTIONS;
        };
    }
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiBreadcrumbsComponent.prototype, "items", void 0);
    __decorate([
        Input(),
        HostBinding('attr.data-tui-host-size'),
        tuiDefaultProp()
    ], TuiBreadcrumbsComponent.prototype, "size", void 0);
    TuiBreadcrumbsComponent = __decorate([
        Component({
            selector: 'tui-breadcrumbs',
            changeDetection: ChangeDetectionStrategy.OnPush,
            template: "<ng-container *ngFor=\"let item of items; last as lastItem\">\n    <ng-container *tuiLet=\"item | tuiMapper: routerLinkOptions as options\">\n        <a\n            tuiLink\n            tuiMode=\"onLight\"\n            routerLinkActive=\"active-link\"\n            #rla=\"routerLinkActive\"\n            [attr.tabindex]=\"rla.isActive ? -1 : null\"\n            [fragment]=\"options.fragment\"\n            [preserveFragment]=\"options.preserveFragment\"\n            [replaceUrl]=\"options.replaceUrl\"\n            [routerLink]=\"item.routerLink\"\n            [routerLinkActiveOptions]=\"item.routerLinkActiveOptions || {exact: false}\"\n            [skipLocationChange]=\"options.skipLocationChange\"\n            [queryParams]=\"options.queryParams\"\n            [target]=\"options.target\"\n        >\n            {{item.caption}}\n        </a>\n        <tui-svg\n            *ngIf=\"!lastItem\"\n            class=\"icon\"\n            src=\"tuiIconChevronRight\"\n        ></tui-svg>\n    </ng-container>\n</ng-container>\n",
            styles: [":host{display:flex;align-items:center;height:40px}:host[data-tui-host-size='m']{font:var(--tui-font-text-s)}:host[data-tui-host-size='l']{font:var(--tui-font-text-m)}.icon{margin:0 4px;color:var(--tui-text-03)}a.active-link[data-mode=onLight]{color:var(--tui-text-03);pointer-events:none}"]
        })
    ], TuiBreadcrumbsComponent);
    return TuiBreadcrumbsComponent;
}());
export { TuiBreadcrumbsComponent };
