// @taiga-ui/kit/fesm5/taiga-ui-kit-classes.js
var TuiDayRangePeriod = /** @class */ (function () {
    function TuiDayRangePeriod(range, name) {
        this.range = range;
        this.name = name;
    }
    TuiDayRangePeriod.prototype.toString = function () {
        return this.name;
    };
    return TuiDayRangePeriod;
}());

var TuiNamedDay = /** @class */ (function () {
    function TuiNamedDay(day, name, displayDay) {
        if (displayDay === void 0) { displayDay = day; }
        this.day = day;
        this.name = name;
        this.displayDay = displayDay;
    }
    TuiNamedDay.prototype.toString = function () {
        return this.name;
    };
    return TuiNamedDay;
}());

/**
 * Wrapper around an item to add `toString()` method
 */
var TuiStringifiableItem = /** @class */ (function () {
    function TuiStringifiableItem(item, stringify) {
        this.item = item;
        this.stringify = stringify;
    }
    TuiStringifiableItem.prototype.toString = function () {
        return this.stringify(this.item);
    };
    return TuiStringifiableItem;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiDayRangePeriod, TuiNamedDay, TuiStringifiableItem };
