// @taiga-ui/kit/fesm5/taiga-ui-kit-components-radio-group.js
import { __decorate } from 'tslib';
import { Input, Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';

var TuiRadioGroupComponent = /** @class */ (function () {
    function TuiRadioGroupComponent() {
        this.name = "tui-radio-group-" + TuiRadioGroupComponent_1.index++;
    }
    TuiRadioGroupComponent_1 = TuiRadioGroupComponent;
    var TuiRadioGroupComponent_1;
    TuiRadioGroupComponent.index = 0;
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiRadioGroupComponent.prototype, "name", void 0);
    TuiRadioGroupComponent = TuiRadioGroupComponent_1 = __decorate([
        Component({
            selector: 'tui-radio-group',
            template: "<ng-content></ng-content>\n",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], TuiRadioGroupComponent);
    return TuiRadioGroupComponent;
}());

var TuiRadioGroupModule = /** @class */ (function () {
    function TuiRadioGroupModule() {
    }
    TuiRadioGroupModule = __decorate([
        NgModule({
            declarations: [TuiRadioGroupComponent],
            exports: [TuiRadioGroupComponent],
        })
    ], TuiRadioGroupModule);
    return TuiRadioGroupModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiRadioGroupComponent, TuiRadioGroupModule };
