// @taiga-ui/core/esm5/components/loader/loader.directive.js
import { __decorate, __param, __read } from "tslib";
import { Directive, ElementRef, Inject, NgZone } from '@angular/core';
import { ANIMATION_FRAME, USER_AGENT } from '@ng-web-apis/common';
import { isEdgeOlderThan, isIE, TuiDestroyService, tuiZonefree } from '@taiga-ui/cdk';
import { Observable } from 'rxjs';
import { pairwise, takeUntil } from 'rxjs/operators';
var PERIMETER = 314;
var COEFFICIENT = 1.5;
var TuiLoaderDirective = /** @class */ (function () {
    function TuiLoaderDirective(animationFrame$, ngZone, destroy$, elementRef, userAgent) {
        var _this = this;
        this.elementRef = elementRef;
        this.strokeDasharray = PERIMETER;
        this.strokeDashoffset = 0;
        if (!isEdgeOlderThan(17, userAgent) && !isIE(userAgent)) {
            return;
        }
        animationFrame$
            .pipe(tuiZonefree(ngZone), pairwise(), takeUntil(destroy$))
            .subscribe(function (_a) {
            var _b = __read(_a, 2), cur = _b[0], prev = _b[1];
            _this.animate(prev - cur);
        });
    }
    TuiLoaderDirective.prototype.animate = function (delta) {
        if (this.strokeDasharray < 0) {
            this.strokeDasharray = PERIMETER * 2;
            this.strokeDashoffset = 0;
        }
        var strokeDasharray = Math.abs(this.strokeDasharray - PERIMETER);
        var fps = 1000 / delta;
        var offsetStep = 1 + Math.floor(this.strokeDasharray / PERIMETER);
        var style = this.elementRef.nativeElement.style;
        style.strokeDashoffset = this.strokeDashoffset.toString();
        style.strokeDasharray = PERIMETER - strokeDasharray + ' ' + strokeDasharray;
        this.strokeDasharray -= PERIMETER / COEFFICIENT / fps;
        this.strokeDashoffset += (offsetStep * PERIMETER) / COEFFICIENT / fps;
        // this.elementRef.nativeElement.parentElement!.style.animation = 'none';
    };
    TuiLoaderDirective.ctorParameters = function () { return [
        { type: Observable, decorators: [{ type: Inject, args: [ANIMATION_FRAME,] }] },
        { type: NgZone, decorators: [{ type: Inject, args: [NgZone,] }] },
        { type: Observable, decorators: [{ type: Inject, args: [TuiDestroyService,] }] },
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: String, decorators: [{ type: Inject, args: [USER_AGENT,] }] }
    ]; };
    TuiLoaderDirective = __decorate([
        Directive({
            selector: 'circle',
            providers: [TuiDestroyService],
        }),
        __param(0, Inject(ANIMATION_FRAME)),
        __param(1, Inject(NgZone)),
        __param(2, Inject(TuiDestroyService)),
        __param(3, Inject(ElementRef)),
        __param(4, Inject(USER_AGENT))
    ], TuiLoaderDirective);
    return TuiLoaderDirective;
}());
export { TuiLoaderDirective };
