// @taiga-ui/core/esm2015/components/error/error.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, Inject, Input } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import { tuiFadeIn, tuiHeightCollapse } from '@taiga-ui/core/animations';
import { MODE_PROVIDER } from '@taiga-ui/core/providers';
import { TUI_DEFAULT_ERROR_MESSAGE, TUI_MODE } from '@taiga-ui/core/tokens';
import { Observable } from 'rxjs';
let TuiErrorComponent = class TuiErrorComponent {
    constructor(mode$, defauleErrorMessage$) {
        this.mode$ = mode$;
        this.defauleErrorMessage$ = defauleErrorMessage$;
        this.error = null;
    }
};
TuiErrorComponent.ctorParameters = () => [
    { type: Observable, decorators: [{ type: Inject, args: [TUI_MODE,] }] },
    { type: Observable, decorators: [{ type: Inject, args: [TUI_DEFAULT_ERROR_MESSAGE,] }] }
];
__decorate([
    Input(),
    tuiDefaultProp()
], TuiErrorComponent.prototype, "error", void 0);
TuiErrorComponent = __decorate([
    Component({
        selector: 'tui-error',
        template: "<div\n    polymorpheus-outlet\n    *ngIf=\"!!error\"\n    automation-id=\"tui-error__text\"\n    class=\"message-text\"\n    [class.message-text_light]=\"(mode$ | async) === 'onDark'\"\n    [content]=\"error.message || (defauleErrorMessage$ | async)\"\n    [context]=\"error.context || {}\"\n    [@tuiHeightCollapse]\n    [@tuiFadeIn]\n></div>\n",
        providers: [MODE_PROVIDER],
        changeDetection: ChangeDetectionStrategy.OnPush,
        animations: [tuiHeightCollapse, tuiFadeIn],
        styles: [":host{font:var(--tui-font-text-s);color:var(--tui-text-01);display:block;color:var(--tui-negative);word-wrap:break-word}.message-text{margin-top:4px;white-space:pre-wrap}.message-text_light{color:var(--tui-negative-night)}"]
    }),
    __param(0, Inject(TUI_MODE)),
    __param(1, Inject(TUI_DEFAULT_ERROR_MESSAGE))
], TuiErrorComponent);
export { TuiErrorComponent };
