// @taiga-ui/cdk/esm5/directives/checked/checked.directive.js
import { __decorate, __param } from "tslib";
import { Directive, ElementRef, EventEmitter, HostListener, Inject, Input, Output, Renderer2, } from '@angular/core';
var TuiCheckedDirective = /** @class */ (function () {
    function TuiCheckedDirective(element, renderer) {
        this.element = element;
        this.renderer = renderer;
        this.tuiCheckedChange = new EventEmitter();
        this.updateProperty('checked', false);
    }
    Object.defineProperty(TuiCheckedDirective.prototype, "tuiChecked", {
        set: function (checked) {
            this.updateProperty('checked', checked || false);
            this.updateProperty('indeterminate', checked === null);
        },
        enumerable: true,
        configurable: true
    });
    TuiCheckedDirective.prototype.onChange = function (_a) {
        var checked = _a.checked;
        this.updateProperty('indeterminate', false);
        this.tuiCheckedChange.emit(checked);
    };
    TuiCheckedDirective.prototype.updateProperty = function (property, value) {
        this.renderer.setProperty(this.element.nativeElement, property, value);
    };
    TuiCheckedDirective.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: Renderer2, decorators: [{ type: Inject, args: [Renderer2,] }] }
    ]; };
    __decorate([
        Input()
    ], TuiCheckedDirective.prototype, "tuiChecked", null);
    __decorate([
        Output()
    ], TuiCheckedDirective.prototype, "tuiCheckedChange", void 0);
    __decorate([
        HostListener('change', ['$event.target'])
    ], TuiCheckedDirective.prototype, "onChange", null);
    TuiCheckedDirective = __decorate([
        Directive({
            selector: 'input[tuiChecked], input[tuiCheckedChange]',
        }),
        __param(0, Inject(ElementRef)),
        __param(1, Inject(Renderer2))
    ], TuiCheckedDirective);
    return TuiCheckedDirective;
}());
export { TuiCheckedDirective };
