// @taiga-ui/cdk/esm5/components/dialog-host/dialog-host.component.js
import { __decorate, __param, __read, __spread } from "tslib";
import { ChangeDetectionStrategy, Component, Inject } from '@angular/core';
import { TUI_PARENT_ANIMATION } from '@taiga-ui/cdk/constants';
import { TUI_DIALOGS } from '@taiga-ui/cdk/tokens';
import { merge } from 'rxjs';
// @dynamic
var TuiDialogHostComponent = /** @class */ (function () {
    function TuiDialogHostComponent(dialogs) {
        this.dialogs = dialogs;
        this.dialogs$ = merge.apply(void 0, __spread(this.dialogs));
    }
    TuiDialogHostComponent.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [TUI_DIALOGS,] }] }
    ]; };
    TuiDialogHostComponent = __decorate([
        Component({
            selector: 'tui-dialog-host',
            template: "<section\n    *ngFor=\"let item of dialogs$ | async\"\n    polymorpheus-outlet\n    tuiFocusTrap\n    tuiOverscroll=\"all\"\n    class=\"dialog\"\n    role=\"dialog\"\n    aria-modal=\"true\"\n    [attr.aria-labelledby]=\"item.id\"\n    [content]=\"item.component\"\n    [context]=\"item\"\n    [@tuiParentAnimation]\n></section>\n<div class=\"overlay\"></div>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            animations: [TUI_PARENT_ANIMATION],
            styles: [":host{position:fixed;left:0;bottom:0;width:100%;height:0}.dialog,.overlay{position:fixed;top:0;left:0;right:0;bottom:0;display:flex;align-items:flex-start;outline:0;overflow:auto;overflow:overlay}.overlay{pointer-events:none;opacity:0;transition:opacity .3s;background:rgba(0,0,0,.5)}.dialog+.overlay{opacity:1}.dialog:last-of-type{z-index:1}"]
        }),
        __param(0, Inject(TUI_DIALOGS))
    ], TuiDialogHostComponent);
    return TuiDialogHostComponent;
}());
export { TuiDialogHostComponent };
