// @taiga-ui/core/esm5/components/root/root.component.js
import { __decorate, __param, __read, __spread } from "tslib";
import { ChangeDetectionStrategy, Component, ElementRef, Inject, Optional, } from '@angular/core';
import { EVENT_MANAGER_PLUGINS } from '@angular/platform-browser';
import { TUI_DIALOGS, TUI_IS_MOBILE, tuiAssert } from '@taiga-ui/cdk';
import { tuiFadeIn } from '@taiga-ui/core/animations';
import { VERSION } from '@taiga-ui/core/constants';
import { TuiNotificationsHostComponent } from '@taiga-ui/core/modules/notifications';
import { TUI_ASSERT_ENABLED } from '@taiga-ui/core/tokens';
import { SilentEventPlugin } from '@tinkoff/ng-event-plugins';
import { merge, of } from 'rxjs';
import { map } from 'rxjs/operators';
// @dynamic
var TuiRootComponent = /** @class */ (function () {
    function TuiRootComponent(elementRef, dialogs, notificationsHost, isMobile, plugins, enabled) {
        this.elementRef = elementRef;
        this.dialogs = dialogs;
        this.notificationsHost = notificationsHost;
        this.isMobile = isMobile;
        this.scrollbars$ = this.dialogs && !this.isMobile
            ? merge.apply(void 0, __spread(this.dialogs)).pipe(map(function (_a) {
                var length = _a.length;
                return !length;
            }))
            : of(!this.isMobile);
        tuiAssert.enabled = enabled;
        tuiAssert.assert(!(plugins[0] instanceof SilentEventPlugin), 'PlatformBrowser or PlatformServer modules must come before TuiRootModule in your main module');
    }
    TuiRootComponent.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: undefined, decorators: [{ type: Optional }, { type: Inject, args: [TUI_DIALOGS,] }] },
        { type: TuiNotificationsHostComponent, decorators: [{ type: Optional }, { type: Inject, args: [TuiNotificationsHostComponent,] }] },
        { type: Boolean, decorators: [{ type: Inject, args: [TUI_IS_MOBILE,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [EVENT_MANAGER_PLUGINS,] }] },
        { type: Boolean, decorators: [{ type: Inject, args: [TUI_ASSERT_ENABLED,] }] }
    ]; };
    TuiRootComponent = __decorate([
        Component({
            selector: 'tui-root',
            template: "<tui-scroll-controls\n    *ngIf=\"scrollbars$ | async\"\n    class=\"scrollbar\"\n></tui-scroll-controls>\n<tui-svg-defs-host></tui-svg-defs-host>\n<tui-portal-host>\n    <div class=\"content\">\n        <ng-content></ng-content>\n    </div>\n    <tui-dialog-host *ngIf=\"dialogs && dialogs.length\"></tui-dialog-host>\n    <ng-content select=\"tuiOverDialogs\"></ng-content>\n    <ng-container\n        *ngIf=\"notificationsHost\"\n        [ngComponentOutlet]=\"notificationsHost\"\n    ></ng-container>\n    <ng-content select=\"tuiOverNotifications\"></ng-content>\n</tui-portal-host>\n<ng-content select=\"tuiOverPortals\"></ng-content>\n<tui-hints-host></tui-hints-host>\n<ng-content select=\"tuiOverHints\"></ng-content>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            host: {
                'data-tui-version': VERSION,
            },
            animations: [tuiFadeIn],
            styles: [":host{position:relative;display:block;flex:1}.scrollbar{position:fixed;top:0;left:0;bottom:0;right:0;margin:0}.content{position:relative;z-index:0;height:100%}"]
        }),
        __param(0, Inject(ElementRef)),
        __param(1, Optional()),
        __param(1, Inject(TUI_DIALOGS)),
        __param(2, Optional()),
        __param(2, Inject(TuiNotificationsHostComponent)),
        __param(3, Inject(TUI_IS_MOBILE)),
        __param(4, Inject(EVENT_MANAGER_PLUGINS)),
        __param(5, Inject(TUI_ASSERT_ENABLED))
    ], TuiRootComponent);
    return TuiRootComponent;
}());
export { TuiRootComponent };
