// @taiga-ui/core/fesm2015/taiga-ui-core-directives-described-by.js
import { __decorate, __param } from 'tslib';
import { InjectionToken, ElementRef, Inject, Input, HostBinding, Directive, NgModule } from '@angular/core';
import { TuiDestroyService, TuiFocusVisibleService, typedFromEvent, tuiDefaultProp } from '@taiga-ui/cdk';
import { TuiHintService } from '@taiga-ui/core/services';
import { merge, timer, Observable } from 'rxjs';
import { filter, switchMapTo, mapTo, takeUntil, take, tap, startWith, distinctUntilChanged } from 'rxjs/operators';

const DELAY = 1000;
const TUI_DESCRIBED_BY_SHOW = new InjectionToken('Accessible tooltip visibility stream');
const TUI_DESCRIBED_BY_PROVIDERS = [
    TuiDestroyService,
    TuiFocusVisibleService,
    {
        provide: TUI_DESCRIBED_BY_SHOW,
        deps: [TuiFocusVisibleService, ElementRef],
        useFactory: describedByFactory,
    },
];
function describedByFactory(focusVisible$, { nativeElement }) {
    return merge(focusVisible$.pipe(filter(v => v), switchMapTo(timer(DELAY).pipe(mapTo(true), takeUntil(typedFromEvent(nativeElement, 'keydown')))), switchMapTo(typedFromEvent(nativeElement, 'keydown').pipe(filter(({ key }) => key === 'Escape'), take(1), tap(event => {
        event.stopPropagation();
    }), mapTo(false), startWith(true)))), typedFromEvent(nativeElement, 'blur').pipe(mapTo(false))).pipe(distinctUntilChanged());
}

const DESCRIBED_BY = '_described-by';
/**
 * A directive linking focusable elements and hints for accessibility
 */
let TuiDescribedByDirective = class TuiDescribedByDirective {
    constructor(hintService, visibility$) {
        this.tuiDescribedBy = '';
        visibility$.subscribe(visible => {
            if (!this.tuiDescribedBy) {
                return;
            }
            if (visible) {
                hintService.showHintForId(this.tuiDescribedBy);
            }
            else {
                hintService.hideHintForId(this.tuiDescribedBy);
            }
        });
    }
    get computedDescribedBy() {
        return this.tuiDescribedBy ? this.tuiDescribedBy + DESCRIBED_BY : null;
    }
};
TuiDescribedByDirective.ctorParameters = () => [
    { type: TuiHintService, decorators: [{ type: Inject, args: [TuiHintService,] }] },
    { type: Observable, decorators: [{ type: Inject, args: [TUI_DESCRIBED_BY_SHOW,] }] }
];
__decorate([
    Input(),
    tuiDefaultProp()
], TuiDescribedByDirective.prototype, "tuiDescribedBy", void 0);
__decorate([
    HostBinding('attr.aria-describedby')
], TuiDescribedByDirective.prototype, "computedDescribedBy", null);
TuiDescribedByDirective = __decorate([
    Directive({
        selector: '[tuiDescribedBy]:not(ng-container)',
        providers: TUI_DESCRIBED_BY_PROVIDERS,
    }),
    __param(0, Inject(TuiHintService)),
    __param(1, Inject(TUI_DESCRIBED_BY_SHOW))
], TuiDescribedByDirective);

let TuiDescribedByModule = class TuiDescribedByModule {
};
TuiDescribedByModule = __decorate([
    NgModule({
        declarations: [TuiDescribedByDirective],
        exports: [TuiDescribedByDirective],
    })
], TuiDescribedByModule);

/**
 * Generated bundle index. Do not edit.
 */

export { DESCRIBED_BY, TUI_DESCRIBED_BY_PROVIDERS, TUI_DESCRIBED_BY_SHOW, TuiDescribedByDirective, TuiDescribedByModule, describedByFactory };
