// @taiga-ui/core/fesm2015/taiga-ui-core-components-group.js
import { __decorate } from 'tslib';
import { Input, HostBinding, Directive, Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import { CommonModule } from '@angular/common';

let TuiGroupDirective = class TuiGroupDirective {
    constructor() {
        this.orientation = "horizontal" /* Horizontal */;
        this.adaptive = false;
        this.collapsed = false;
        this.rounded = true;
        this.size = 'm';
    }
    get orientationHorizontal() {
        return this.orientation === "horizontal" /* Horizontal */;
    }
    get orientationVertical() {
        return this.orientation === "vertical" /* Vertical */;
    }
    get sizeLarge() {
        return this.size === 'l';
    }
};
__decorate([
    Input(),
    tuiDefaultProp()
], TuiGroupDirective.prototype, "orientation", void 0);
__decorate([
    Input(),
    HostBinding('class.tui-group_adaptive'),
    tuiDefaultProp()
], TuiGroupDirective.prototype, "adaptive", void 0);
__decorate([
    Input(),
    HostBinding('class.tui-group_collapsed'),
    tuiDefaultProp()
], TuiGroupDirective.prototype, "collapsed", void 0);
__decorate([
    Input(),
    HostBinding('class.tui-group_rounded'),
    tuiDefaultProp()
], TuiGroupDirective.prototype, "rounded", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], TuiGroupDirective.prototype, "size", void 0);
__decorate([
    HostBinding('class.tui-group_orientation_horizontal')
], TuiGroupDirective.prototype, "orientationHorizontal", null);
__decorate([
    HostBinding('class.tui-group_orientation_vertical')
], TuiGroupDirective.prototype, "orientationVertical", null);
__decorate([
    HostBinding('class.tui-group_radius_large')
], TuiGroupDirective.prototype, "sizeLarge", null);
TuiGroupDirective = __decorate([
    Directive({
        selector: '[tuiGroup]:not(ng-container)',
        host: {
            class: 'tui-group',
            role: 'group',
        },
    })
], TuiGroupDirective);

/** @deprecated */
let TuiGroupComponent = class TuiGroupComponent extends TuiGroupDirective {
};
TuiGroupComponent = __decorate([
    Component({
        selector: 'tui-group',
        changeDetection: ChangeDetectionStrategy.OnPush,
        template: '<ng-content></ng-content>',
        host: {
            class: 'tui-group',
            role: 'group',
        }
    })
], TuiGroupComponent);

let TuiGroupModule = class TuiGroupModule {
};
TuiGroupModule = __decorate([
    NgModule({
        imports: [CommonModule],
        declarations: [TuiGroupComponent, TuiGroupDirective],
        exports: [TuiGroupComponent, TuiGroupDirective],
    })
], TuiGroupModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiGroupComponent, TuiGroupDirective, TuiGroupModule };
