// @taiga-ui/core/fesm2015/taiga-ui-core-components-label.js
import { __decorate, __param } from 'tslib';
import { Inject, Input, ContentChild, HostBinding, Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { NgControl } from '@angular/forms';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import { MODE_PROVIDER } from '@taiga-ui/core/providers';
import { TUI_MODE } from '@taiga-ui/core/tokens';
import { Observable } from 'rxjs';
import { PolymorpheusModule } from '@tinkoff/ng-polymorpheus';

let TuiLabelComponent = class TuiLabelComponent {
    constructor(mode$) {
        this.mode$ = mode$;
        this.label = '';
        this.context = {
            $implicit: null,
        };
    }
};
TuiLabelComponent.ctorParameters = () => [
    { type: Observable, decorators: [{ type: Inject, args: [TUI_MODE,] }] }
];
__decorate([
    Input(),
    tuiDefaultProp()
], TuiLabelComponent.prototype, "label", void 0);
__decorate([
    Input(),
    tuiDefaultProp()
], TuiLabelComponent.prototype, "context", void 0);
__decorate([
    ContentChild(NgControl),
    HostBinding('class._control')
], TuiLabelComponent.prototype, "control", void 0);
TuiLabelComponent = __decorate([
    Component({
        selector: 'label[tuiLabel]',
        template: "<span polymorpheus-outlet class=\"label\" [content]=\"label\" [context]=\"context\">\n    <ng-template let-text>\n        <span class=\"text\">{{text}}</span>\n    </ng-template>\n</span>\n<span class=\"content\"><ng-content></ng-content></span>\n",
        changeDetection: ChangeDetectionStrategy.OnPush,
        providers: [MODE_PROVIDER],
        host: {
            '($.data-mode.attr)': 'mode$',
        },
        styles: [":host{display:flex;min-height:44px;justify-content:space-between;flex-direction:column;pointer-events:none}.label{font:var(--tui-font-text-s);max-width:100%;-ms-grid-row-align:inherit;align-self:inherit;overflow:inherit;pointer-events:auto}.text{display:block;overflow:inherit;text-overflow:ellipsis}:host:not(._control) .text{color:var(--tui-text-02)}:host([data-mode=onDark]) .text{color:var(--tui-text-02-night)}.content{line-height:24px;overflow:inherit;text-overflow:ellipsis;pointer-events:auto}"]
    }),
    __param(0, Inject(TUI_MODE))
], TuiLabelComponent);

let TuiLabelModule = class TuiLabelModule {
};
TuiLabelModule = __decorate([
    NgModule({
        imports: [PolymorpheusModule],
        declarations: [TuiLabelComponent],
        exports: [TuiLabelComponent],
    })
], TuiLabelModule);

/**
 * Generated bundle index. Do not edit.
 */

export { TuiLabelComponent, TuiLabelModule };
