// @taiga-ui/core/fesm2015/taiga-ui-core-directives-dropdown-controller.js
import { __decorate } from 'tslib';
import { InjectionToken, Input, Directive, forwardRef, NgModule, ChangeDetectorRef } from '@angular/core';
import { TuiController, tuiDefaultProp, TuiDestroyService } from '@taiga-ui/cdk';
import { DEFAULT_MIN_HEIGHT, DEFAULT_MAX_HEIGHT } from '@taiga-ui/core/constants';
import { watchedControllerFactory } from '@taiga-ui/core/providers';

const TUI_DROPDOWN_CONTROLLER = new InjectionToken('Controls configuration of hints', {
    factory: () => new TuiDropdownControllerDirective(),
});

var TuiDropdownControllerDirective_1;
let TuiDropdownControllerDirective = TuiDropdownControllerDirective_1 = class TuiDropdownControllerDirective extends TuiController {
    constructor() {
        super(...arguments);
        this.align = 'right';
        this.direction = null;
        this.limitWidth = "auto" /* Auto */;
        this.minHeight = DEFAULT_MIN_HEIGHT;
        this.maxHeight = DEFAULT_MAX_HEIGHT;
        this.sided = false;
    }
};
__decorate([
    Input('tuiDropdownAlign'),
    tuiDefaultProp()
], TuiDropdownControllerDirective.prototype, "align", void 0);
__decorate([
    Input('tuiDropdownDirection'),
    tuiDefaultProp()
], TuiDropdownControllerDirective.prototype, "direction", void 0);
__decorate([
    Input('tuiDropdownLimitWidth'),
    tuiDefaultProp()
], TuiDropdownControllerDirective.prototype, "limitWidth", void 0);
__decorate([
    Input('tuiDropdownMinHeight'),
    tuiDefaultProp()
], TuiDropdownControllerDirective.prototype, "minHeight", void 0);
__decorate([
    Input('tuiDropdownMaxHeight'),
    tuiDefaultProp()
], TuiDropdownControllerDirective.prototype, "maxHeight", void 0);
__decorate([
    Input('tuiDropdownSided'),
    tuiDefaultProp()
], TuiDropdownControllerDirective.prototype, "sided", void 0);
TuiDropdownControllerDirective = TuiDropdownControllerDirective_1 = __decorate([
    Directive({
        selector: '[tuiDropdownAlign], [tuiDropdownDirection], [tuiDropdownLimitWidth], [tuiDropdownMinHeight]' +
            '[tuiDropdownMaxHeight], [tuiDropdownSided]',
        providers: [
            {
                provide: TUI_DROPDOWN_CONTROLLER,
                useExisting: forwardRef(() => TuiDropdownControllerDirective_1),
            },
        ],
    })
], TuiDropdownControllerDirective);

let TuiDropdownControllerModule = class TuiDropdownControllerModule {
};
TuiDropdownControllerModule = __decorate([
    NgModule({
        declarations: [TuiDropdownControllerDirective],
        exports: [TuiDropdownControllerDirective],
    })
], TuiDropdownControllerModule);

const TUI_DROPDOWN_WATCHED_CONTROLLER = new InjectionToken('watched dropdown controller');
const ɵ0 = watchedControllerFactory;
const DROPDOWN_CONTROLLER_PROVIDER = [
    TuiDestroyService,
    {
        provide: TUI_DROPDOWN_WATCHED_CONTROLLER,
        deps: [TUI_DROPDOWN_CONTROLLER, ChangeDetectorRef, TuiDestroyService],
        useFactory: ɵ0,
    },
];

/**
 * Generated bundle index. Do not edit.
 */

export { DROPDOWN_CONTROLLER_PROVIDER, TUI_DROPDOWN_CONTROLLER, TUI_DROPDOWN_WATCHED_CONTROLLER, TuiDropdownControllerDirective, TuiDropdownControllerModule, ɵ0 };
