// @taiga-ui/core/esm2015/directives/dropdown-controller/dropdown-controller.directive.js
var TuiDropdownControllerDirective_1;
import { __decorate } from "tslib";
import { Directive, forwardRef, Input } from '@angular/core';
import { TuiController, tuiDefaultProp } from '@taiga-ui/cdk';
import { DEFAULT_MAX_HEIGHT, DEFAULT_MIN_HEIGHT } from '@taiga-ui/core/constants';
import { TUI_DROPDOWN_CONTROLLER } from './dropdown-controller.token';
let TuiDropdownControllerDirective = TuiDropdownControllerDirective_1 = class TuiDropdownControllerDirective extends TuiController {
    constructor() {
        super(...arguments);
        this.align = 'right';
        this.direction = null;
        this.limitWidth = "auto" /* Auto */;
        this.minHeight = DEFAULT_MIN_HEIGHT;
        this.maxHeight = DEFAULT_MAX_HEIGHT;
        this.sided = false;
    }
};
__decorate([
    Input('tuiDropdownAlign'),
    tuiDefaultProp()
], TuiDropdownControllerDirective.prototype, "align", void 0);
__decorate([
    Input('tuiDropdownDirection'),
    tuiDefaultProp()
], TuiDropdownControllerDirective.prototype, "direction", void 0);
__decorate([
    Input('tuiDropdownLimitWidth'),
    tuiDefaultProp()
], TuiDropdownControllerDirective.prototype, "limitWidth", void 0);
__decorate([
    Input('tuiDropdownMinHeight'),
    tuiDefaultProp()
], TuiDropdownControllerDirective.prototype, "minHeight", void 0);
__decorate([
    Input('tuiDropdownMaxHeight'),
    tuiDefaultProp()
], TuiDropdownControllerDirective.prototype, "maxHeight", void 0);
__decorate([
    Input('tuiDropdownSided'),
    tuiDefaultProp()
], TuiDropdownControllerDirective.prototype, "sided", void 0);
TuiDropdownControllerDirective = TuiDropdownControllerDirective_1 = __decorate([
    Directive({
        selector: '[tuiDropdownAlign], [tuiDropdownDirection], [tuiDropdownLimitWidth], [tuiDropdownMinHeight]' +
            '[tuiDropdownMaxHeight], [tuiDropdownSided]',
        providers: [
            {
                provide: TUI_DROPDOWN_CONTROLLER,
                useExisting: forwardRef(() => TuiDropdownControllerDirective_1),
            },
        ],
    })
], TuiDropdownControllerDirective);
export { TuiDropdownControllerDirective };
