// @taiga-ui/core/esm5/components/group/group.directive.js
import { __decorate } from "tslib";
import { Directive, HostBinding, Input } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
var TuiGroupDirective = /** @class */ (function () {
    function TuiGroupDirective() {
        this.orientation = "horizontal" /* Horizontal */;
        this.adaptive = false;
        this.collapsed = false;
        this.rounded = true;
        this.size = 'm';
    }
    Object.defineProperty(TuiGroupDirective.prototype, "orientationHorizontal", {
        get: function () {
            return this.orientation === "horizontal" /* Horizontal */;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiGroupDirective.prototype, "orientationVertical", {
        get: function () {
            return this.orientation === "vertical" /* Vertical */;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiGroupDirective.prototype, "sizeLarge", {
        get: function () {
            return this.size === 'l';
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "orientation", void 0);
    __decorate([
        Input(),
        HostBinding('class.tui-group_adaptive'),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "adaptive", void 0);
    __decorate([
        Input(),
        HostBinding('class.tui-group_collapsed'),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "collapsed", void 0);
    __decorate([
        Input(),
        HostBinding('class.tui-group_rounded'),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "rounded", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "size", void 0);
    __decorate([
        HostBinding('class.tui-group_orientation_horizontal')
    ], TuiGroupDirective.prototype, "orientationHorizontal", null);
    __decorate([
        HostBinding('class.tui-group_orientation_vertical')
    ], TuiGroupDirective.prototype, "orientationVertical", null);
    __decorate([
        HostBinding('class.tui-group_radius_large')
    ], TuiGroupDirective.prototype, "sizeLarge", null);
    TuiGroupDirective = __decorate([
        Directive({
            selector: '[tuiGroup]:not(ng-container)',
            host: {
                class: 'tui-group',
                role: 'group',
            },
        })
    ], TuiGroupDirective);
    return TuiGroupDirective;
}());
export { TuiGroupDirective };
