// @taiga-ui/kit/esm5/components/multi-select-option/multi-select-option.component.js
import { __decorate, __extends } from "tslib";
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { isPresent, tuiReplayedValueChangesFrom } from '@taiga-ui/cdk';
import { sizeBigger } from '@taiga-ui/core';
import { TuiSelectOptionComponent } from '@taiga-ui/kit/components/select-option';
import { PolymorpheusComponent } from '@tinkoff/ng-polymorpheus';
import { map } from 'rxjs/operators';
var TuiMultiSelectOptionComponent = /** @class */ (function (_super) {
    __extends(TuiMultiSelectOptionComponent, _super);
    function TuiMultiSelectOptionComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.selected$ = tuiReplayedValueChangesFrom(_this.control).pipe(map(function (value) {
            return isPresent(_this.option.value) &&
                !!value &&
                !!value.find(function (item) { return _this.matcher(item, _this.option.value); });
        }));
        return _this;
    }
    Object.defineProperty(TuiMultiSelectOptionComponent.prototype, "size", {
        get: function () {
            return sizeBigger(this.option.size) ? 'l' : 'm';
        },
        enumerable: true,
        configurable: true
    });
    TuiMultiSelectOptionComponent = __decorate([
        Component({
            selector: 'tui-multi-select-option',
            template: "<tui-primitive-checkbox\n    class=\"tui-space_right-3\"\n    [size]=\"size\"\n    [value]=\"selected$ | async\"\n></tui-primitive-checkbox>\n<ng-container *ngTemplateOutlet=\"context.$implicit\"></ng-container>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [":host{display:flex;flex:1;align-items:center}"]
        })
    ], TuiMultiSelectOptionComponent);
    return TuiMultiSelectOptionComponent;
}(TuiSelectOptionComponent));
export { TuiMultiSelectOptionComponent };
export var TUI_MULTI_SELECT_OPTION = new PolymorpheusComponent(TuiMultiSelectOptionComponent);
