// @taiga-ui/core/fesm5/taiga-ui-core-pipes-month.js
import { __decorate, __param } from 'tslib';
import { Inject, Pipe, NgModule } from '@angular/core';
import { TUI_MONTHS } from '@taiga-ui/core/tokens';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

// @dynamic
var TuiMonthPipe = /** @class */ (function () {
    function TuiMonthPipe(months$) {
        this.months$ = months$;
    }
    TuiMonthPipe.prototype.transform = function (_a) {
        var month = _a.month;
        return this.months$.pipe(map(function (months) { return months[month]; }));
    };
    TuiMonthPipe.ctorParameters = function () { return [
        { type: Observable, decorators: [{ type: Inject, args: [TUI_MONTHS,] }] }
    ]; };
    TuiMonthPipe = __decorate([
        Pipe({ name: 'tuiMonth' }),
        __param(0, Inject(TUI_MONTHS))
    ], TuiMonthPipe);
    return TuiMonthPipe;
}());

var TuiMonthPipeModule = /** @class */ (function () {
    function TuiMonthPipeModule() {
    }
    TuiMonthPipeModule = __decorate([
        NgModule({
            exports: [TuiMonthPipe],
            declarations: [TuiMonthPipe],
        })
    ], TuiMonthPipeModule);
    return TuiMonthPipeModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiMonthPipe, TuiMonthPipeModule };
