// @taiga-ui/core/fesm5/taiga-ui-core-components-group.js
import { __decorate, __extends } from 'tslib';
import { Input, HostBinding, Directive, Component, ChangeDetectionStrategy, NgModule } from '@angular/core';
import { tuiDefaultProp } from '@taiga-ui/cdk';
import { CommonModule } from '@angular/common';

var TuiGroupDirective = /** @class */ (function () {
    function TuiGroupDirective() {
        this.orientation = "horizontal" /* Horizontal */;
        this.adaptive = false;
        this.collapsed = false;
        this.rounded = true;
        this.size = 'm';
    }
    Object.defineProperty(TuiGroupDirective.prototype, "orientationHorizontal", {
        get: function () {
            return this.orientation === "horizontal" /* Horizontal */;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiGroupDirective.prototype, "orientationVertical", {
        get: function () {
            return this.orientation === "vertical" /* Vertical */;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiGroupDirective.prototype, "sizeLarge", {
        get: function () {
            return this.size === 'l';
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "orientation", void 0);
    __decorate([
        Input(),
        HostBinding('class.tui-group_adaptive'),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "adaptive", void 0);
    __decorate([
        Input(),
        HostBinding('class.tui-group_collapsed'),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "collapsed", void 0);
    __decorate([
        Input(),
        HostBinding('class.tui-group_rounded'),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "rounded", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiGroupDirective.prototype, "size", void 0);
    __decorate([
        HostBinding('class.tui-group_orientation_horizontal')
    ], TuiGroupDirective.prototype, "orientationHorizontal", null);
    __decorate([
        HostBinding('class.tui-group_orientation_vertical')
    ], TuiGroupDirective.prototype, "orientationVertical", null);
    __decorate([
        HostBinding('class.tui-group_radius_large')
    ], TuiGroupDirective.prototype, "sizeLarge", null);
    TuiGroupDirective = __decorate([
        Directive({
            selector: '[tuiGroup]:not(ng-container)',
            host: {
                class: 'tui-group',
                role: 'group',
            },
        })
    ], TuiGroupDirective);
    return TuiGroupDirective;
}());

/** @deprecated */
var TuiGroupComponent = /** @class */ (function (_super) {
    __extends(TuiGroupComponent, _super);
    function TuiGroupComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TuiGroupComponent = __decorate([
        Component({
            selector: 'tui-group',
            changeDetection: ChangeDetectionStrategy.OnPush,
            template: '<ng-content></ng-content>',
            host: {
                class: 'tui-group',
                role: 'group',
            }
        })
    ], TuiGroupComponent);
    return TuiGroupComponent;
}(TuiGroupDirective));

var TuiGroupModule = /** @class */ (function () {
    function TuiGroupModule() {
    }
    TuiGroupModule = __decorate([
        NgModule({
            imports: [CommonModule],
            declarations: [TuiGroupComponent, TuiGroupDirective],
            exports: [TuiGroupComponent, TuiGroupDirective],
        })
    ], TuiGroupModule);
    return TuiGroupModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiGroupComponent, TuiGroupDirective, TuiGroupModule };
