// @taiga-ui/core/fesm5/taiga-ui-core-directives-hint-controller.js
import { __extends, __decorate } from 'tslib';
import { InjectionToken, Input, Directive, forwardRef, NgModule, ChangeDetectorRef } from '@angular/core';
import { tuiDefaultProp, TuiController, TuiDestroyService } from '@taiga-ui/cdk';
import { watchedControllerFactory } from '@taiga-ui/core/providers';

var TUI_HINT_CONTROLLER = new InjectionToken('Controls configuration of hints', {
    factory: function () { return new TuiHintControllerDirective(); },
});

// TODO: v2.0 use in Charts
var TuiHintControllerDirective = /** @class */ (function (_super) {
    __extends(TuiHintControllerDirective, _super);
    function TuiHintControllerDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.content = null;
        _this.direction = 'bottom-left';
        _this.mode = null;
        return _this;
    }
    TuiHintControllerDirective_1 = TuiHintControllerDirective;
    var TuiHintControllerDirective_1;
    __decorate([
        Input('tuiHintContent'),
        tuiDefaultProp()
    ], TuiHintControllerDirective.prototype, "content", void 0);
    __decorate([
        Input('tuiHintDirection'),
        tuiDefaultProp()
    ], TuiHintControllerDirective.prototype, "direction", void 0);
    __decorate([
        Input('tuiHintMode'),
        tuiDefaultProp()
    ], TuiHintControllerDirective.prototype, "mode", void 0);
    TuiHintControllerDirective = TuiHintControllerDirective_1 = __decorate([
        Directive({
            selector: '[tuiHintContent]',
            providers: [
                {
                    provide: TUI_HINT_CONTROLLER,
                    useExisting: forwardRef(function () { return TuiHintControllerDirective_1; }),
                },
            ],
        })
    ], TuiHintControllerDirective);
    return TuiHintControllerDirective;
}(TuiController));

var TuiHintControllerModule = /** @class */ (function () {
    function TuiHintControllerModule() {
    }
    TuiHintControllerModule = __decorate([
        NgModule({
            declarations: [TuiHintControllerDirective],
            exports: [TuiHintControllerDirective],
        })
    ], TuiHintControllerModule);
    return TuiHintControllerModule;
}());

var TUI_HINT_WATCHED_CONTROLLER = new InjectionToken('watched hint controller');
var ɵ0 = watchedControllerFactory;
var HINT_CONTROLLER_PROVIDER = [
    TuiDestroyService,
    {
        provide: TUI_HINT_WATCHED_CONTROLLER,
        deps: [TUI_HINT_CONTROLLER, ChangeDetectorRef, TuiDestroyService],
        useFactory: ɵ0,
    },
];

/**
 * Generated bundle index. Do not edit.
 */

export { HINT_CONTROLLER_PROVIDER, TUI_HINT_CONTROLLER, TUI_HINT_WATCHED_CONTROLLER, TuiHintControllerDirective, TuiHintControllerModule, ɵ0 };
