// @taiga-ui/core/fesm5/taiga-ui-core-components-theme-night.js
import { __extends, __decorate, __param } from 'tslib';
import { DOCUMENT } from '@angular/common';
import { Inject, Component, ViewEncapsulation, NgModule } from '@angular/core';
import { AbstractTuiThemeSwitcher } from '@taiga-ui/cdk';

// @dynamic
var TuiThemeNightComponent = /** @class */ (function (_super) {
    __extends(TuiThemeNightComponent, _super);
    function TuiThemeNightComponent(documentRef) {
        return _super.call(this, documentRef) || this;
    }
    TuiThemeNightComponent.ctorParameters = function () { return [
        { type: Document, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
    ]; };
    TuiThemeNightComponent = __decorate([
        Component({
            selector: 'tui-theme-night',
            template: '',
            encapsulation: ViewEncapsulation.None,
            styles: [":root{--tui-base-01:#222;--tui-base-02:#333;--tui-base-03:#808080;--tui-base-04:#959595;--tui-base-05:#b0b0b0;--tui-base-06:#d7d7d7;--tui-base-07:#ededed;--tui-base-08:#f6f6f6;--tui-base-09:#fff;--tui-focus:rgba(51, 51, 51, 0.64);--tui-secondary:rgba(255, 255, 255, 0.16);--tui-secondary-hover:rgba(255, 255, 255, 0.24);--tui-secondary-active:rgba(255, 255, 255, 0.4);--tui-clear:rgba(255, 255, 255, 0.16);--tui-clear-disabled:rgba(255, 255, 255, 0.08);--tui-clear-hover:rgba(255, 255, 255, 0.24);--tui-clear-active:rgba(255, 255, 255, 0.4);--tui-error-fill:#ff8c67;--tui-error-bg:rgba(244, 87, 37, 0.32);--tui-error-bg-hover:rgba(244, 87, 37, 0.4);--tui-success-fill:#4ac99b;--tui-success-bg:rgba(74, 201, 155, 0.32);--tui-success-bg-hover:rgba(74, 201, 155, 0.4);--tui-warning-fill:#ffc700;--tui-warning-bg:rgba(255, 199, 0, 0.32);--tui-warning-bg-hover:rgba(255, 199, 0, 0.4);--tui-info-fill:#70b6f6;--tui-info-bg:rgba(112, 182, 246, 0.32);--tui-info-bg-hover:rgba(112, 182, 246, 0.4);--tui-neutral-fill:#959ba4;--tui-neutral-bg:#959ba4;--tui-neutral-bg-hover:#959ba4;--tui-text-01:#ffffff;--tui-text-02:rgba(255, 255, 255, 0.72);--tui-text-03:rgba(255, 255, 255, 0.6);--tui-link:#6788ff;--tui-link-hover:#526ed3;--tui-positive:#44c596;--tui-positive-hover:#3aa981;--tui-negative:#ff8c67;--tui-negative-hover:#bb593a}"]
        }),
        __param(0, Inject(DOCUMENT))
    ], TuiThemeNightComponent);
    return TuiThemeNightComponent;
}(AbstractTuiThemeSwitcher));

var TuiThemeNightModule = /** @class */ (function () {
    function TuiThemeNightModule() {
    }
    TuiThemeNightModule = __decorate([
        NgModule({
            declarations: [TuiThemeNightComponent],
            exports: [TuiThemeNightComponent]
        })
    ], TuiThemeNightModule);
    return TuiThemeNightModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiThemeNightComponent, TuiThemeNightModule };
