// @taiga-ui/core/fesm5/taiga-ui-core-directives-mask-accessor.js
import { __decorate } from 'tslib';
import { Directive, NgModule } from '@angular/core';
import { NG_VALUE_ACCESSOR, DefaultValueAccessor } from '@angular/forms';
import { TUI_VALUE_ACCESSOR } from '@taiga-ui/core/tokens';

function fallbackAccessorFactory(accessor, fallback) {
    return accessor ? accessor[0] : fallback;
}
var TuiMaskAccessorDirective = /** @class */ (function () {
    function TuiMaskAccessorDirective() {
    }
    TuiMaskAccessorDirective = __decorate([
        Directive({
            selector: 'input[tuiMaskAccessor]',
            providers: [
                {
                    provide: NG_VALUE_ACCESSOR,
                    deps: [TUI_VALUE_ACCESSOR, DefaultValueAccessor],
                    useFactory: fallbackAccessorFactory,
                    multi: true,
                },
            ],
        })
    ], TuiMaskAccessorDirective);
    return TuiMaskAccessorDirective;
}());

/**
 * @internal
 */
var TuiMaskAccessorModule = /** @class */ (function () {
    function TuiMaskAccessorModule() {
    }
    TuiMaskAccessorModule = __decorate([
        NgModule({
            declarations: [TuiMaskAccessorDirective],
            exports: [TuiMaskAccessorDirective],
        })
    ], TuiMaskAccessorModule);
    return TuiMaskAccessorModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiMaskAccessorDirective, TuiMaskAccessorModule, fallbackAccessorFactory };
