// @taiga-ui/core/fesm5/taiga-ui-core-directives-manual-hint.js
import { __extends, __decorate, __param } from 'tslib';
import { ElementRef, Inject, Input, Directive, NgModule } from '@angular/core';
import { tuiDefaultProp, tuiRequiredSetter } from '@taiga-ui/cdk';
import { AbstractTuiHint } from '@taiga-ui/core/abstract';
import { TuiHintService } from '@taiga-ui/core/services';

var TuiManualHintDirective = /** @class */ (function (_super) {
    __extends(TuiManualHintDirective, _super);
    function TuiManualHintDirective(elementRef, hintService) {
        var _this = _super.call(this, elementRef, hintService) || this;
        _this.content = '';
        return _this;
    }
    Object.defineProperty(TuiManualHintDirective.prototype, "tuiManualHintShow", {
        set: function (show) {
            if (show) {
                this.showTooltip();
            }
            else {
                this.hideTooltip();
            }
        },
        enumerable: true,
        configurable: true
    });
    TuiManualHintDirective.prototype.getElementClientRect = function () {
        return this.elementRef.nativeElement.getBoundingClientRect();
    };
    TuiManualHintDirective.ctorParameters = function () { return [
        { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
        { type: TuiHintService, decorators: [{ type: Inject, args: [TuiHintService,] }] }
    ]; };
    __decorate([
        Input('tuiManualHint'),
        tuiDefaultProp()
    ], TuiManualHintDirective.prototype, "content", void 0);
    __decorate([
        Input(),
        tuiRequiredSetter()
    ], TuiManualHintDirective.prototype, "tuiManualHintShow", null);
    TuiManualHintDirective = __decorate([
        Directive({
            selector: '[tuiManualHint]:not(ng-container)',
        }),
        __param(0, Inject(ElementRef)),
        __param(1, Inject(TuiHintService))
    ], TuiManualHintDirective);
    return TuiManualHintDirective;
}(AbstractTuiHint));

var TuiManualHintModule = /** @class */ (function () {
    function TuiManualHintModule() {
    }
    TuiManualHintModule = __decorate([
        NgModule({
            declarations: [TuiManualHintDirective],
            exports: [TuiManualHintDirective],
        })
    ], TuiManualHintModule);
    return TuiManualHintModule;
}());

/**
 * Generated bundle index. Do not edit.
 */

export { TuiManualHintDirective, TuiManualHintModule };
