// @taiga-ui/core/esm5/modules/notifications/notifications.module.js
import { __decorate } from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { TuiLetModule } from '@taiga-ui/cdk';
import { TuiNotificationAlertModule } from './notification-alert/notification-alert.module';
import { TuiNotificationsHostComponent } from './notifications-host/notifications-host.component';
var ɵ0 = TuiNotificationsHostComponent;
var TuiNotificationsModule = /** @class */ (function () {
    function TuiNotificationsModule() {
    }
    TuiNotificationsModule = __decorate([
        NgModule({
            imports: [CommonModule, TuiNotificationAlertModule, TuiLetModule],
            declarations: [TuiNotificationsHostComponent],
            exports: [TuiNotificationsHostComponent],
            entryComponents: [TuiNotificationsHostComponent],
            providers: [
                {
                    provide: TuiNotificationsHostComponent,
                    useValue: ɵ0,
                },
            ],
        })
    ], TuiNotificationsModule);
    return TuiNotificationsModule;
}());
export { TuiNotificationsModule };
export { ɵ0 };
