// @taiga-ui/cdk/esm2015/components/portal-host/portal-host.component.js
var TuiPortalHostComponent_1;
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, ComponentFactory, ComponentRef, ElementRef, EmbeddedViewRef, Inject, Injector, TemplateRef, ViewChild, ViewContainerRef, } from '@angular/core';
import { TuiPortalService } from './portal.service';
const BLANK_CLIENT_RECT = {
    bottom: 0,
    height: 0,
    left: 0,
    right: 0,
    top: 0,
    width: 0,
};
/**
 * Host element for dynamically created portals, for example using {@link TuiDropdownDirective}.
 */
let TuiPortalHostComponent = TuiPortalHostComponent_1 = class TuiPortalHostComponent {
    constructor(viewContainerRef, elementRef, portalService) {
        this.viewContainerRef = viewContainerRef;
        this.elementRef = elementRef;
        portalService.attach(this);
    }
    get clientRect() {
        return this.elementRef.nativeElement.getBoundingClientRect();
    }
    addComponentChild(componentFactory, injector) {
        return this.viewContainerRef.createComponent(componentFactory, undefined, Injector.create({
            parent: injector,
            providers: [
                {
                    provide: TuiPortalHostComponent_1,
                    useValue: this,
                },
            ],
        }));
    }
    addTemplateChild(templateRef, context) {
        return this.viewContainerRef.createEmbeddedView(templateRef, context);
    }
    fixedPositionOffset() {
        return this.positionFixedOffsetRef
            ? this.positionFixedOffsetRef.nativeElement.getBoundingClientRect()
            : BLANK_CLIENT_RECT;
    }
};
TuiPortalHostComponent.ctorParameters = () => [
    { type: ViewContainerRef, decorators: [{ type: Inject, args: [ViewContainerRef,] }] },
    { type: ElementRef, decorators: [{ type: Inject, args: [ElementRef,] }] },
    { type: TuiPortalService, decorators: [{ type: Inject, args: [TuiPortalService,] }] }
];
__decorate([
    ViewChild('positionFixedOffset')
], TuiPortalHostComponent.prototype, "positionFixedOffsetRef", void 0);
TuiPortalHostComponent = TuiPortalHostComponent_1 = __decorate([
    Component({
        selector: 'tui-portal-host',
        template: "<div #positionFixedOffset class=\"position-fixed-offset\"></div>\n<ng-content></ng-content>\n",
        changeDetection: ChangeDetectionStrategy.OnPush,
        styles: [":host{position:relative;display:block;height:100%;z-index:0}:host:before{content:'';display:block;overflow:hidden}.position-fixed-offset{position:fixed;left:0;top:0;pointer-events:none;visibility:hidden;width:100%;height:100%}"]
    }),
    __param(0, Inject(ViewContainerRef)),
    __param(1, Inject(ElementRef)),
    __param(2, Inject(TuiPortalService))
], TuiPortalHostComponent);
export { TuiPortalHostComponent };
