// @taiga-ui/cdk/esm2015/components/portal-host/portal.service.js
import { __decorate } from "tslib";
import { Injectable, } from '@angular/core';
import * as i0 from "@angular/core";
const NO_HOST = 'Portals cannot be used without TuiPortalHostComponent';
/**
 * Service for displaying portals
 */
let TuiPortalService = class TuiPortalService {
    get safeHost() {
        if (!this.host) {
            throw new Error(NO_HOST);
        }
        return this.host;
    }
    attach(host) {
        this.host = host;
    }
    add(componentFactory, injector) {
        return this.safeHost.addComponentChild(componentFactory, injector);
    }
    remove({ hostView }) {
        hostView.destroy();
    }
    addTemplate(templateRef, context) {
        return this.safeHost.addTemplateChild(templateRef, context);
    }
    removeTemplate(viewRef) {
        viewRef.destroy();
    }
};
TuiPortalService.ɵprov = i0.ɵɵdefineInjectable({ factory: function TuiPortalService_Factory() { return new TuiPortalService(); }, token: TuiPortalService, providedIn: "root" });
TuiPortalService = __decorate([
    Injectable({
        providedIn: 'root',
    })
], TuiPortalService);
export { TuiPortalService };
