// @taiga-ui/core/esm5/modules/notifications/notification-alert/notification-alert.component.js
import { __decorate, __param } from "tslib";
import { ChangeDetectionStrategy, Component, Inject, Input } from '@angular/core';
import { TuiDestroyService, tuiPure } from '@taiga-ui/cdk';
import { timer } from 'rxjs';
import { takeUntil } from 'rxjs/operators';
export var ALERT_AUTOCLOSE_TIMEOUT = 3000;
var TuiNotificationAlertComponent = /** @class */ (function () {
    function TuiNotificationAlertComponent(destroy$) {
        var _this = this;
        timer(ALERT_AUTOCLOSE_TIMEOUT)
            .pipe(takeUntil(destroy$))
            .subscribe(function () {
            if (_this.safeItem.autoClose) {
                _this.closeDialog();
            }
        });
    }
    Object.defineProperty(TuiNotificationAlertComponent.prototype, "safeItem", {
        get: function () {
            if (!this.item) {
                throw new Error('Notification was created as undefined');
            }
            return this.item;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TuiNotificationAlertComponent.prototype, "context", {
        get: function () {
            return this.calculateContext(this.safeItem);
        },
        enumerable: true,
        configurable: true
    });
    TuiNotificationAlertComponent.prototype.closeDialog = function () {
        this.safeItem.observer.complete();
    };
    TuiNotificationAlertComponent.prototype.calculateContext = function (_a) {
        var status = _a.status, data = _a.data, label = _a.label, observer = _a.observer;
        return {
            $implicit: status,
            data: data,
            label: label,
            closeHook: function () {
                observer.complete();
            },
            emitHook: function (data) {
                observer.next(data);
            },
            emitAndCloseHook: function (data) {
                observer.next(data);
                observer.complete();
            },
        };
    };
    TuiNotificationAlertComponent.ctorParameters = function () { return [
        { type: TuiDestroyService, decorators: [{ type: Inject, args: [TuiDestroyService,] }] }
    ]; };
    __decorate([
        Input()
    ], TuiNotificationAlertComponent.prototype, "item", void 0);
    __decorate([
        tuiPure
    ], TuiNotificationAlertComponent.prototype, "calculateContext", null);
    TuiNotificationAlertComponent = __decorate([
        Component({
            selector: 'tui-notification-alert',
            changeDetection: ChangeDetectionStrategy.OnPush,
            template: "<tui-notification\n    *ngIf=\"safeItem.hasCloseButton else noClose\"\n    [status]=\"safeItem.status\"\n    [hasIcon]=\"safeItem.hasIcon\"\n    (close)=\"closeDialog()\"\n>\n    <label\n        *ngIf=\"safeItem.label\"\n        automation-id=\"tui-notification-alert__heading\"\n        class=\"heading\"\n    >\n        {{safeItem.label}}\n    </label>\n    <div\n        polymorpheus-outlet\n        automation-id=\"tui-notification-alert__content\"\n        class=\"content\"\n        [content]=\"safeItem.content\"\n        [context]=\"context\"\n    ></div>\n</tui-notification>\n<ng-template #noClose>\n    <tui-notification [status]=\"safeItem.status\" [hasIcon]=\"safeItem.hasIcon\">\n        <label\n            *ngIf=\"safeItem.label\"\n            automation-id=\"tui-notification-alert__heading\"\n            class=\"heading\"\n        >\n            {{safeItem.label}}\n        </label>\n        <div\n            polymorpheus-outlet\n            automation-id=\"tui-notification-alert__content\"\n            class=\"content\"\n            [content]=\"safeItem.content\"\n            [context]=\"context\"\n        ></div>\n    </tui-notification>\n</ng-template>\n",
            providers: [TuiDestroyService],
            styles: [":host{box-shadow:0 8px 16px rgba(51,51,51,.2);display:block;border-radius:var(--tui-radius-m)}.heading{font:var(--tui-font-text-m);line-height:20px;font-weight:700;margin:0}.content{color:var(--tui-text-01);word-wrap:break-word}.content:empty{display:none}"]
        }),
        __param(0, Inject(TuiDestroyService))
    ], TuiNotificationAlertComponent);
    return TuiNotificationAlertComponent;
}());
export { TuiNotificationAlertComponent };
