// @taiga-ui/core/esm5/directives/textfield-controller/textfield-example-text.directive.js
import { __decorate, __extends } from "tslib";
import { Directive, forwardRef, InjectionToken, Input } from '@angular/core';
import { TuiController } from '@taiga-ui/cdk';
export var TUI_TEXTFIELD_EXAMPLE_TEXT = new InjectionToken('tuiTextfieldExampleText', { factory: exampleTextDirectiveFactory });
var TuiTextfieldExampleTextDirective = /** @class */ (function (_super) {
    __extends(TuiTextfieldExampleTextDirective, _super);
    function TuiTextfieldExampleTextDirective() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.exampleText = '';
        return _this;
    }
    TuiTextfieldExampleTextDirective_1 = TuiTextfieldExampleTextDirective;
    var TuiTextfieldExampleTextDirective_1;
    __decorate([
        Input('tuiTextfieldExampleText')
    ], TuiTextfieldExampleTextDirective.prototype, "exampleText", void 0);
    TuiTextfieldExampleTextDirective = TuiTextfieldExampleTextDirective_1 = __decorate([
        Directive({
            selector: '[tuiTextfieldExampleText]',
            providers: [
                {
                    provide: TUI_TEXTFIELD_EXAMPLE_TEXT,
                    useExisting: forwardRef(function () { return TuiTextfieldExampleTextDirective_1; }),
                },
            ],
        })
    ], TuiTextfieldExampleTextDirective);
    return TuiTextfieldExampleTextDirective;
}(TuiController));
export { TuiTextfieldExampleTextDirective };
export function exampleTextDirectiveFactory() {
    return new TuiTextfieldExampleTextDirective();
}
