// @taiga-ui/kit/esm5/components/accordion/accordion.component.js
import { __decorate, __param, __read, __spread } from "tslib";
import { ChangeDetectionStrategy, Component, ContentChildren, Inject, Input, } from '@angular/core';
import { EMPTY_QUERY, identity, isPresent, itemsQueryListObservable, tuiDefaultProp, TuiDestroyService, } from '@taiga-ui/cdk';
import { merge } from 'rxjs';
import { filter, map, mapTo, pairwise, switchMap, takeUntil } from 'rxjs/operators';
import { TuiAccordionItemComponent } from './accordion-item/accordion-item.component';
var TuiAccordionComponent = /** @class */ (function () {
    function TuiAccordionComponent(destroy$) {
        this.destroy$ = destroy$;
        this.closeOthers = true;
        this.rounded = true;
        this.accordionItems = EMPTY_QUERY;
    }
    TuiAccordionComponent.prototype.ngAfterContentInit = function () {
        var _this = this;
        var accordionItems = this.accordionItems;
        var rows$ = itemsQueryListObservable(accordionItems);
        var newOpenRow$ = rows$.pipe(pairwise(), map(function (_a) {
            var _b = __read(_a, 2), previous = _b[0], current = _b[1];
            return current.find(function (item) { return previous.indexOf(item) === -1 && item.open; });
        }), filter(isPresent));
        var rowsOpen$ = merge(rows$.pipe(switchMap(function (rows) {
            return merge.apply(void 0, __spread(rows.map(function (row) {
                return row.openChange.pipe(filter(identity), mapTo(row));
            })));
        })), newOpenRow$).pipe(filter(function () { return _this.closeOthers; }), takeUntil(this.destroy$));
        rowsOpen$.subscribe(function (currentRow) {
            accordionItems.forEach(function (row) {
                if (currentRow !== row) {
                    row.close();
                }
            });
        });
    };
    TuiAccordionComponent.ctorParameters = function () { return [
        { type: TuiDestroyService, decorators: [{ type: Inject, args: [TuiDestroyService,] }] }
    ]; };
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiAccordionComponent.prototype, "closeOthers", void 0);
    __decorate([
        Input(),
        tuiDefaultProp()
    ], TuiAccordionComponent.prototype, "rounded", void 0);
    __decorate([
        ContentChildren(TuiAccordionItemComponent)
    ], TuiAccordionComponent.prototype, "accordionItems", void 0);
    TuiAccordionComponent = __decorate([
        Component({
            selector: 'tui-accordion',
            template: "<!--TODO: Unnecessary nesting, probably accordion can be a directive-->\n<div\n    tuiGroup\n    automation-id=\"tui-accordion__group\"\n    class=\"group\"\n    orientation=\"vertical\"\n    size=\"l\"\n    [collapsed]=\"true\"\n    [rounded]=\"rounded\"\n>\n    <ng-content select=\"tui-accordion-item\"></ng-content>\n</div>\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            providers: [TuiDestroyService],
            styles: [":host{display:block}.group{display:flex}"]
        }),
        __param(0, Inject(TuiDestroyService))
    ], TuiAccordionComponent);
    return TuiAccordionComponent;
}());
export { TuiAccordionComponent };
